---
title: "Dual organism analysis - pseudobulk RNA-seq - malaria"
output:
  html_document:
    df_print: paged
    code_folding: hide
  pdf_document: default
  toc: TRUE
---


```{r setup, include=T} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
knitr::opts_chunk$set(fig.width = 10, fig.height = 8) 
```

```{r}
library(Seurat)
library(patchwork)
library(readxl)
library(tidyverse)
library(ggplot2)
library(ggpubr)
library(reshape2)
library(broom)
library(stringr)
library(ggplot2)
library(SeuratData)
#library(celldex)
#install.packages("cowplot")
library(cowplot)
#devtools::install_github('cole-trapnell-lab/monocle3')
#library(monocle3)
#remotes::install_github('satijalab/seurat-wrappers')
library(SeuratWrappers)
#BiocManager::install('multtest', force = T)
library(multtest)
library(RColorBrewer)
library(rtracklayer)
library(DESeq2)
library(data.table)
library(MatrixModels)
# library(SingleCellExperiment)
library(scran)
library(org.Hs.eg.db)
library(clusterProfiler)
library(topGO)
library(ggrepel)
library(BiocParallel)
#library(EnhanceVolcano)
# install.packages('metap')
#library(metap)
#remotes::install_github("satijalab/seurat-data")

```

```{r}
my_PseudobulkExpression <- function(
  object,
  pb.method = 'aggregate',
  assays = NULL,
  features = NULL,
  return.seurat = FALSE,
  group.by = 'ident',
  add.ident = NULL,
  slot = 'data',
  verbose = TRUE,
  ...
) {
  # CheckDots(..., fxns = 'CreateSeuratObject')
  # if (!is.null(x = add.ident)) {
  #   .Deprecated(msg = "'add.ident' is a deprecated argument, please use the 'group.by' argument instead")
  #   group.by <- c('ident', add.ident)
  # }
  # if (!(pb.method %in% c('average', 'aggregate'))) {
  #   stop("'pb.method' must be either 'average' or 'aggregate'")
  # }
  object.assays <- FilterObjects(object = object, classes.keep = 'Assay')
  assays <- assays %||% object.assays
  if (!all(assays %in% object.assays)) {
    assays <- assays[assays %in% object.assays]
    if (length(x = assays) == 0) {
      stop("None of the requested assays are present in the object")
    } else {
      warning("Requested assays that do not exist in object. Proceeding with existing assays only.")
    }
  }
  if (length(x = slot) == 1) {
    slot <- rep_len(x = slot, length.out = length(x = assays))
  } else if (length(x = slot) != length(x = assays)) {
    stop("Number of slots provided does not match number of assays")
  }
  data <- FetchData(object = object, vars = rev(x = group.by))
  data <- data[which(rowSums(x = is.na(x = data)) == 0), , drop = F]
  if (nrow(x = data) < ncol(x = object)) {
    message("Removing cells with NA for 1 or more grouping variables")
    object <- subset(x = object, cells = rownames(x = data))
  }
  for (i in 1:ncol(x = data)) {
    data[, i] <- as.factor(x = data[, i])
  }
  num.levels <- sapply(
    X = 1:ncol(x = data),
    FUN = function(i) {
      length(x = levels(x = data[, i]))
    }
  )
  if (any(num.levels == 1)) {
    message(paste0("The following grouping variables have 1 value and will be ignored: ",
                   paste0(colnames(x = data)[which(num.levels <= 1)], collapse = ", ")))
    group.by <- colnames(x = data)[which(num.levels > 1)]
    data <- data[, which(num.levels > 1), drop = F]
  }
  if (ncol(x = data) == 0) {
    message("All grouping variables have 1 value only. Computing across all cells.")
    category.matrix <- matrix(
      data = 1,
      nrow = ncol(x = object),
      dimnames = list(Cells(x = object), 'all')
    )
    if (pb.method == 'average') {
      category.matrix <- category.matrix / sum(category.matrix)
    }
  } else {
    category.matrix <- model.Matrix(object = as.formula(
      object = paste0(
        '~0+',
        paste0(
          "data[,",
          1:length(x = group.by),
          "]",
          collapse = ":"
        )
      )
    ))
    colsums <- colSums(x = category.matrix)
    category.matrix <- category.matrix[, colsums > 0]
    colsums <- colsums[colsums > 0]
    if (pb.method == 'average') {
      category.matrix <- Sweep(
        x = category.matrix,
        MARGIN = 2,
        STATS = colsums,
        FUN = "/")
    }
    colnames(x = category.matrix) <- sapply(
      X = colnames(x = category.matrix),
      FUN = function(name) {
        name <- gsub(pattern = "data\\[, [1-9]*\\]", replacement = "", x = name)
        return(paste0(rev(x = unlist(x = strsplit(x = name, split = ":"))), collapse = "_"))
      })
  }
  data.return <- list()
  for (i in 1:length(x = assays)) {
    data.use <- GetAssayData(
      object = object,
      assay = assays[i],
      slot = slot[i]
    )
    features.to.avg <- features %||% rownames(x = data.use)
    if (inherits(x = features, what = "list")) {
      features.to.avg <- features[i]
    }
    if (IsMatrixEmpty(x = data.use)) {
      warning(
        "The ", slot[i], " slot for the ", assays[i],
        " assay is empty. Skipping assay.", immediate. = TRUE, call. = FALSE)
      next
    }
    bad.features <- setdiff(x = features.to.avg, y = rownames(x = data.use))
    if (length(x = bad.features) > 0) {
      warning(
        "The following ", length(x = bad.features),
        " features were not found in the ", assays[i], " assay: ",
        paste(bad.features, collapse = ", "), call. = FALSE, immediate. = TRUE)
    }
    features.assay <- intersect(x = features.to.avg, y = rownames(x = data.use))
    if (length(x = features.assay) > 0) {
      data.use <- data.use[features.assay, ]
    } else {
      warning("None of the features specified were found in the ", assays[i],
              " assay.", call. = FALSE, immediate. = TRUE)
      next
    }
    if (slot[i] == 'data') {
      data.use <- expm1(x = data.use)
      if (any(data.use == Inf)) {
        warning("Exponentiation yielded infinite values. `data` may not be log-normed.")
      }
    }
    data.return[[i]] <- as.matrix(x = (data.use %*% category.matrix))
    names(x = data.return)[i] <- assays[[i]]
  }
  if (return.seurat) {
    if (slot[1] == 'scale.data') {
      na.matrix <- data.return[[1]]
      na.matrix[1:length(x = na.matrix)] <- NA
      toRet <- CreateSeuratObject(
        counts = na.matrix,
        project = if (pb.method == "average") "Average" else "Aggregate",
        assay = names(x = data.return)[1],
        check.matrix = FALSE,
        ...
      )
      toRet <- SetAssayData(
        object = toRet,
        assay = names(x = data.return)[1],
        slot = "counts",
        new.data = matrix()
      )
      toRet <- SetAssayData(
        object = toRet,
        assay = names(x = data.return)[1],
        slot = "data",
        new.data = na.matrix
      )
      toRet <- SetAssayData(
        object = toRet,
        assay = names(x = data.return)[1],
        slot = "scale.data",
        new.data = data.return[[1]]
      )
    } else {
      toRet <- CreateSeuratObject(
        counts = data.return[[1]],
        project = if (pb.method == "average") "Average" else "Aggregate",
        assay = names(x = data.return)[1],
        check.matrix = FALSE,
        ...
      )
      toRet <- SetAssayData(
        object = toRet,
        assay = names(x = data.return)[1],
        slot = "data",
        new.data = log1p(x = as.matrix(x = data.return[[1]]))
      )
    }
    #for multimodal data
    if (length(x = data.return) > 1) {
      for (i in 2:length(x = data.return)) {
        if (slot[i] == 'scale.data') {
          na.matrix <- data.return[[i]]
          na.matrix[1:length(x = na.matrix)] <- NA
          toRet[[names(x = data.return)[i]]] <- CreateAssayObject(counts = na.matrix, check.matrix = FALSE)
          toRet <- SetAssayData(
            object = toRet,
            assay = names(x = data.return)[i],
            slot = "counts",
            new.data = matrix()
          )
          toRet <- SetAssayData(
            object = toRet,
            assay = names(x = data.return)[i],
            slot = "data",
            new.data = na.matrix
          )
          toRet <- SetAssayData(
            object = toRet,
            assay = names(x = data.return)[i],
            slot = "scale.data",
            new.data = as.matrix(x = data.return[[i]])
          )
        } else {
          toRet[[names(x = data.return)[i]]] <- CreateAssayObject(counts = data.return[[i]], check.matrix = FALSE)
          toRet <- SetAssayData(
            object = toRet,
            assay = names(x = data.return)[i],
            slot = "data",
            new.data = log1p(x = as.matrix(x = data.return[[i]]))
          )
        }

      }
    }
    if (DefaultAssay(object = object) %in% names(x = data.return)) {
      DefaultAssay(object = toRet) <- DefaultAssay(object = object)
      if (slot[which(DefaultAssay(object = object) %in% names(x = data.return))[1]] != 'scale.data') {
        toRet <- ScaleData(object = toRet, verbose = verbose)
      }
    }
    if ('ident' %in% group.by) {
      first.cells <- c()
      for (i in 1:ncol(x = category.matrix)) {
        first.cells <- c(first.cells, Position(x = category.matrix[,i], f = function(x) {x > 0}))
      }
      Idents(object = toRet) <- Idents(object = object)[first.cells]
    }
    return(toRet)
  } else {
    return(data.return)
  }
}

de_function_nothr <- function(condition1, condition2)
{
  #padj_thr_sc <- padj_thr_sc
  
  counts <- cts %>% 
    as.data.frame() %>% 
    dplyr::select(c(starts_with(condition1), starts_with(condition2)))

  # 2 Generate metadata
  colData <- data.frame(sample = colnames(counts))
  colData$condition = sapply(str_split(colData$sample, "D"), function(x) x[1])
  colData <- colData %>% 
    tibble::column_to_rownames("sample")
  
  # more info for metadata
  
  # DEseq2 object
  dds <- DESeqDataSetFromMatrix(countData = counts,
                               colData = colData,
                               design = ~condition)
  
  
  # filter dds
  keep <- rowSums(counts(dds)) >= 10
  dds <- dds[keep,]
  
  dds$condition <- relevel(dds$condition, ref = condition1)
  
  dds <- DESeq(dds)
  
  comparison <- resultsNames(dds)[2]

  res <- results(dds, name = comparison)
  
  # Turn the DESeq2 results object into a tibble for use with tidyverse functions
  res_tbl <- res %>%
    data.frame() %>%
    rownames_to_column(var = "gene") %>%
    as_tibble()
  
  
  
  #write.csv(res_tbl, "iRBC_vs_RBC_scRNAseq.csv", row.names = F, quote = F)
  ####
  
  # # Subset the significant results
  # sig_res <- dplyr::filter(res_tbl, padj <= padj_thr_sc) %>%
  #   filter(abs(log2FoldChange) >= logFC_thr) %>% 
  #   dplyr::arrange(padj)
  # 
  # # Check significant genes output
  # #sig_res
  # 
  # ## Extract normalized counts from dds object
  # normalized_counts <- counts(dds, normalized = TRUE)
  # 
  # ## Extract normalized counts for significant genes only
  # sig_counts <- normalized_counts[rownames(normalized_counts) %in% sig_res$gene, ]
  # 
  return(res_tbl)
  #return(list(dds,sig_counts, sig_res))
  
}

IDconvert <- read.delim("../Data/PlasmoDB-64_Pfalciparum3D7_GO.gaf", 
                header = F, 
                stringsAsFactors = F,
                comment = "!")

PF_readable <- function(df)
{
  df_geneID <- df %>% 
  #filter(grepl('gene-PF3D7-', gene)) %>% 
    na.omit() %>% 
    mutate(gene = gsub("gene-", "", gene)) %>% 
    mutate(gene = gsub("-", "_", gene))

  gene_name <- c()
  
  for(i in 1:nrow(df_geneID))
  {
    if(df_geneID$gene[i] %in% IDconvert$V2)
       {
         gene_name[i] <- paste(df_geneID$gene[i], 
                                unique(IDconvert[grep(pattern = df_geneID$gene[i],
                                               IDconvert$V2),"V10"]),
                                sep = "-")
    }else{
      gene_name[i] <- df_geneID$gene[i]
    }
  }
  df_geneID$gene_name <- gene_name
  
  return(df_geneID)
}


### host GO

host_GO <- function(host, host_nothr)
{
  if(host == "Hs")
  {
    org.db = org.Hs.eg.db
  }else{
    org.db = org.Mm.eg.db
  }
  
  host_nothr_geneList <- host_nothr %>% 
    filter(!grepl('PF3D7', gene)) %>% 
    na.omit() %>% 
    filter(abs(log2FoldChange) > 1.5) %>%
    filter(padj <= 0.1)
  host_nothr_geneList_ENTREZ <- mapIds(org.db, host_nothr_geneList$gene, 'ENTREZID', 'SYMBOL')
  
  universe <- host_nothr %>% 
    filter(!grepl('PF3D7', gene)) %>% 
    na.omit()
  universe_ENTREZ <- mapIds(org.db, universe$gene, 'ENTREZID', 'SYMBOL')
  
  # host_nothr_geneList_ <- host_nothr_geneList$log2FoldChange
  # names(host_nothr_geneList_) <-  host_nothr_geneList$gene
  # 
  # host_nothr_geneList_sorted <- sort(host_nothr_geneList_, decreasing = T)
  # names(host_nothr_geneList_sorted) <- mapIds(org.db, names(host_nothr_geneList_sorted), 'ENTREZID', 'SYMBOL')
  
  host_nothr_ego <- enrichGO(gene = host_nothr_geneList_ENTREZ,
                              #geneList     = host_nothr_geneList_sorted,
                             #keytype = "SYMBOL",
                             universe = universe_ENTREZ,
                OrgDb        = org.db,
                #eps = 0,
                ont          = "BP",
                minGSSize    = 10,
                maxGSSize    = 500,
                pvalueCutoff = 0.05,
                #verbose      = FALSE,
                readable = T)
  
  #host_nothr_ego <- setReadable(host_nothr_ego, OrgDb = org.db)
  
  return(host_nothr_ego)
}

### Plasmo GO

plasmo_GO <- function(plasmo_nothr)
{
  pf_nothr_pos_log2fc_geneList <- plasmo_nothr %>% 
    filter(grepl('PF3D7', gene)) %>% 
    na.omit() %>% 
    filter(log2FoldChange > 1.5) %>% 
    filter(padj <= 0.1) %>% 
    mutate(gene = gsub("gene-", "", gene)) %>% 
    mutate(gene = gsub("-", "_", gene))

  pf_nothr_neg_log2fc_geneList <- plasmo_nothr %>% 
    filter(grepl('PF3D7', gene)) %>% 
    na.omit() %>% 
    filter(log2FoldChange < -1.5) %>% 
    filter(padj <= 0.1)  %>% 
    mutate(gene = gsub("gene-", "", gene)) %>% 
    mutate(gene = gsub("-", "_", gene))
  
  pf_nothr_neg_log2fc_geneList_ <- pf_nothr_neg_log2fc_geneList$log2FoldChange
  names(pf_nothr_neg_log2fc_geneList_) <-  pf_nothr_neg_log2fc_geneList$gene
  
  # Plasmodium GO terms
  para_genes_neg <- names(pf_nothr_neg_log2fc_geneList_)
  
  
  ######## correct way to write the annot file from gaf #########
  # gaf <- read.delim("~/Downloads/PlasmoDB-64_Pfalciparum3D7_GO.gaf", 
  #                   header = F, 
  #                   stringsAsFactors = F,
  #                   comment = "!") %>% 
  #   select(V2, V5) %>% 
  #   rename(geneID = V2, GOterm = V5)
  # 
  # gaf <- aggregate(gaf$GOterm, gaf['geneID'], paste0, collape = "")
  # gaf$GO <- sapply(gaf$x, function(y) paste0(unique(unlist(y)), collapse = ","))
  # gaf <- gaf[,c(1,3)]
  # colnames(gaf) <- NULL
  # 
  # write.table(gaf, "../Data/PlasmoDB-64_Pfalciparum3D7_GO.txt", sep = '\t', row.names = F, quote = F)
  ################
  IDconvert <- read.delim("../Data/PlasmoDB-64_Pfalciparum3D7_GO.gaf", 
                header = F, 
                stringsAsFactors = F,
                comment = "!")
  
  geneID2GO <- readMappings("../Data/PlasmoDB-64_Pfalciparum3D7_GO.txt")
  # parasite genes of interest
  para_in_neg <- unique(as.character(para_genes_neg))
  # universe of parasite genes
  para_bg <-  plasmo_nothr %>% 
    filter(grepl('PF3D7', gene)) %>% 
    na.omit() %>% 
    # mutate(gene = gsub("gene-", "", gene)) %>% 
    # mutate(gene = gsub("-", "_", gene)) %>% 
    pull(gene)
  # to know which genes are interesting within the universe, we do %in% with background genes
  geneList_neg = factor(as.integer(para_bg %in% para_in_neg))
  # make it a named vector, that's what topGO requires
  names(geneList_neg)= para_bg

  if(length(para_in_neg > 0) > 0)
  {
    GOdata_neg <- new("topGOdata",
                ontology = "BP",
                allGenes = geneList_neg,
                annot = annFUN.gene2GO,
                gene2GO = geneID2GO,
                nodeSize = 1)
  # Expected: Under random chance, number of genes that would be expected 
  # to be significantly DE and annotated with that term
  # The column Expected represents the expected number of interesting genes mapped to the 
  # GO term if the interesting genes were randomly distributed over all GO terms.
  
  resultKS_neg=runTest(GOdata_neg, algorithm='weight01', statistic='KS') 
  allGO_neg=usedGO(GOdata_neg)
  all_res_neg=GenTable(GOdata_neg, KS=resultKS_neg, orderBy="KS", topNodes=length(allGO_neg), numChar = 1000)
  
  # Get genes in a particular GO term:
  
  
  GenesForGOterm_neg <- c()
  myterms_neg = all_res_neg$GO.ID
  mygenes_neg <- genesInTerm(GOdata_neg, myterms_neg)
  for (i in 1:length(myterms_neg))
  {
    myterm_neg <- mygenes_neg[myterms_neg[i]][[1]]
    mygenesforterm_neg <- myterm_neg[myterm_neg %in% para_in_neg]
    genes_neg <- sapply(mygenesforterm_neg,
                    function(x) paste(x, unique(IDconvert[grep(pattern = x, IDconvert$V2),"V10"]), sep = "-"))
    genes_neg <- paste(genes_neg, collapse=' ,')
    GenesForGOterm_neg[i] <- genes_neg
  }

  all_res_neg$GenesForGOterm_neg <- GenesForGOterm_neg 
  all_res_neg <- all_res_neg %>% 
    filter(Significant > 0)
  
  }else
  {all_res_neg <- ""}
  
  ### positive
  pf_nothr_pos_log2fc_geneList_ <- pf_nothr_pos_log2fc_geneList$log2FoldChange
  names(pf_nothr_pos_log2fc_geneList_) <-  pf_nothr_pos_log2fc_geneList$gene
  
  # Plasmodium GO terms
  para_genes_pos <- names(pf_nothr_pos_log2fc_geneList_)
  
  
  ######## correct way to write the annot file from gaf #########
  # gaf <- read.delim("~/Downloads/PlasmoDB-64_Pfalciparum3D7_GO.gaf", 
  #                   header = F, 
  #                   stringsAsFactors = F,
  #                   comment = "!") %>% 
  #   select(V2, V5) %>% 
  #   rename(geneID = V2, GOterm = V5)
  # 
  # gaf <- aggregate(gaf$GOterm, gaf['geneID'], paste0, collape = "")
  # gaf$GO <- sapply(gaf$x, function(y) paste0(unique(unlist(y)), collapse = ","))
  # gaf <- gaf[,c(1,3)]
  # colnames(gaf) <- NULL
  # 
  # write.table(gaf, "../Data/PlasmoDB-64_Pfalciparum3D7_GO.txt", sep = '\t', row.names = F, quote = F)
  ################
  
  geneID2GO <- readMappings("../Data/PlasmoDB-64_Pfalciparum3D7_GO.txt")
  # parasite genes of interest
  para_in_pos <- unique(as.character(para_genes_pos))
  # to know which genes are interesting within the universe, we do %in% with background genes
  geneList_pos = factor(as.integer(para_bg %in% para_in_pos))
  # make it a named vector, that's what topGO requires
  names(geneList_pos)= para_bg

  if(length(para_in_pos > 0) > 0)
  {
  GOdata_pos <- new("topGOdata",
                ontology = "BP",
                allGenes = geneList_pos,
                annot = annFUN.gene2GO,
                gene2GO = geneID2GO,
                nodeSize = 1)
  # Expected: Under random chance, number of genes that would be expected 
  # to be significantly DE and annotated with that term
  # The column Expected represents the expected number of interesting genes mapped to the 
  # GO term if the interesting genes were randomly distributed over all GO terms.
  
  resultKS_pos=runTest(GOdata_pos, algorithm='weight01', statistic='KS') 
  allGO_pos=usedGO(GOdata_pos)
  all_res_pos=GenTable(GOdata_pos, KS=resultKS_pos, orderBy="KS", topNodes=length(allGO_pos), numChar = 1000)
  
 
  GenesForGOterm_pos <- c()
  myterms_pos = all_res_pos$GO.ID
  mygenes_pos <- genesInTerm(GOdata_pos, myterms_pos)
  for (i in 1:length(myterms_pos))
  {
    myterm_pos <- mygenes_pos[myterms_pos[i]][[1]]
    mygenesforterm_pos <- myterm_pos[myterm_pos %in% para_in_pos]
    genes_pos <- sapply(mygenesforterm_pos,
                    function(x) paste(x, unique(IDconvert[grep(pattern = x, IDconvert$V2),"V10"]), sep = "-"))
    genes_pos <- paste(genes_pos, collapse=' ,')
    GenesForGOterm_pos[i] <- genes_pos
  }

  all_res_pos$GenesForGOterm_pos <- GenesForGOterm_pos
  
  all_res_pos <- all_res_pos %>% 
    filter(Significant > 0)
  }else
  {all_res_pos <- ""}
  
  return(list(pf_neg = all_res_neg, pf_pos = all_res_pos))
}

volcano_plot <- function(de, plot_title)
{
  log10padj.thr = 2
  log2FCthr = 2
  
  de %>% 
    mutate(log.p.adj = -log10(padj)) %>% 
    mutate(col = case_when(
      log.p.adj >= log10padj.thr & abs(log2FoldChange) >= log2FCthr ~ "deepskyblue4",
      log.p.adj < log10padj.thr & abs(log2FoldChange) >= log2FCthr ~ "firebrick",
      log.p.adj >= log10padj.thr & abs(log2FoldChange) < log2FCthr ~ "darkolivegreen",
      TRUE ~ "gray"
    )) %>% 
    # mutate(geneID = ifelse(test = grepl(geneID, pattern = ";"), yes = str_extract(string = geneID, pattern = ".*(?=;)"), no = geneID )) %>% 
    mutate(label = case_when(
      #logPval < 1.3 & abs(fc) < 0.2 ~ "",
      #TRUE ~ geneID
      col == "deepskyblue4" &
        grepl(pattern = "PF3D7", gene) ~ gene_name,
      .default = ""
    )) %>% 
    mutate(shape = case_when(
      grepl(pattern = "PF3D7", gene) ~ 8,
      .default = 16
      )) %>%
    mutate(size = case_when(
      col == "deepskyblue4" ~ 4,
      .default = 3)) -> df
  
# New facet label names for supp variable
# labs <- c("BNT/BNT visit 0 vs 1", "BNT/BNT visit 5 vs 6", 
#           "AZD/BNT visit 0 vs 1", "AZD/BNT visit 5 vs 6")
# names(labs) <- c("contrast: BNT_BNT_visit_0 vs BNT_BNT_visit_1", "contrast: BNT_BNT_visit_5 vs BNT_BNT_visit_6", 
#                       "contrast: AZD_BNT_visit_0 vs AZD_BNT_visit_1", "contrast: AZD_BNT_visit_5 vs AZD_BNT_visit_6")

# vaccine_labeller <- function(variable,value){
#   return(labs[value])
# }

  
  #for(i in 1:length(contrast_list))
    # print(
    # df %>% 
        #filter(contrast == paste0("contrast: ", contrast_list[[i]][1], " vs ", contrast_list[[i]][2], collapse = ""))  %>% 
      
        p <- ggplot(df ,aes(x = log2FoldChange, y = log.p.adj, 
                   colour = col, 
                   label = label, 
                   shape = shape,
                   #size = size
                   )) +
        geom_point(alpha = 0.7, size = 4) +
        #scale_size_identity() +
        scale_colour_identity() +
        scale_shape_identity() +
        theme_bw() +
        geom_hline(yintercept = log10padj.thr, linetype = "dashed", colour = "gray84") +
        geom_vline(xintercept = log2FCthr, linetype = "dashed", colour = "gray84") +
        geom_vline(xintercept = -log2FCthr, linetype = "dashed", colour = "gray84") +
        # facet_wrap(~ contrast, 
        #            ncol = 2,
        #            labeller = as_labeller(labs)) +
        geom_text_repel(size = 2.5, box.padding = 0.5, max.overlaps = 100) +
        theme_bw() +
        theme(strip.background =element_rect(fill="aliceblue", colour = "white"),
              strip.text.x = element_text(size = 11),
              axis.text = element_text(size = 10)) +
        #ylim(-0.75, 100) +
        #xlim(-10, 10) +
        xlab("Log2 Fold Change") +
        ylab("-Log10 p.adjusted") +
         ggtitle(plot_title)
    #)
  ggsave(paste0(plot_title, "_volcano.pdf", collapse = ""))
  
  #return(p1)
}
volcano_plot(de = de, plot_title = "RBC bystander vs RBC infected1")
```


# Pseudobulk DE analysis 1 {.tabset}
```{r}
dual <- readRDS("robjects/dual.integrated.TBremoved.cellbender_plasmo.bg.removed_reintegration.rds")

DefaultAssay(dual) <- "RNA"

dual@meta.data <- dual@meta.data %>% 
  mutate(condition_collapse = case_when(
    (Group == "sporozoite_infected"  | Group == "sporozoite_bystander") ~ "sporozoite_bys_inf",
    (Group == "RBC_infected" | Group == "RBC_bystander") ~ "RBC_bys_inf",
    .default = Group
  )) %>% 
  mutate(cellID = "Cell")

table(dual$Donorid, dual$orig.ident)

dual@meta.data$cond_collapse_donor <- paste(dual@meta.data$condition_collapse,
                                               dual@meta.data$Donorid,
                                               sep = "")

# cts <- AggregateExpression(dual,
#               group.by = c("cellID", "cond_collapse_donor"),
#               assays = "RNA",
#               slot = "counts",
#               return.seurat = F)

cts <- my_PseudobulkExpression(dual,
              group.by = c("cellID", "cond_collapse_donor"),
              assays = "RNA",
              slot = "counts",
              pb.method = "aggregate",
              return.seurat = F)

cts <- cts$RNA

# transpose
# cts.t <- as.data.frame(t(cts))
#
# # split.data.frame
# splitRows <- gsub('_.*', '', rownames(cts.t))
# #splitRows <- sapply(strsplit(rownames(cts.t), split = "D"), function(x) x[1])
# cts.split <- split.data.frame(cts.t,
#                  f = factor(splitRows))
# 
# # fix colnames and transpose
# cts.modified <- lapply(cts.split, function(x){
#   rownames(x) <- gsub("(.+?)(\\_.*)", "\\2", rownames(x))
#   rownames(x) <- gsub("^_", "\\1", rownames(x))
#   t(x)
# })
```


<!-- ```{r} -->

<!-- hs_gtf <- import("../Data/Homo_sapiens.GRCh38.109.gtf") # Ensembl version 109 - June 2, 2023 -->
<!-- hs_gtf_df <- as.data.frame(hs_gtf) -->
<!-- hs_gtf_prot_cod <- hs_gtf_df %>% filter(type == "exon") %>% filter(gene_biotype == "protein_coding") %>% distinct(gene_name, .keep_all = T) -->
<!-- hs_gtf_df <- as.data.frame(hs_gtf) -->
<!-- hs_gtf_prot_cod <- hs_gtf_df %>%  -->
<!--   filter(type == "exon") %>%  -->
<!--   filter(gene_biotype == "protein_coding") %>%  -->
<!--   distinct(gene_name, .keep_all = T) -->

<!-- hs_prot <- hs_gtf_prot_cod %>% pull(gene_name) %>% unique() -->

<!-- # check if all genes are protein-coding -->
<!-- hs_scgenes <- rownames(cts) -->

<!-- length(setdiff(hs_scgenes, hs_prot)) -->
<!-- # 10133 -->
<!-- length(intersect(hs_scgenes, hs_prot)) -->
<!-- # 14399 -->

<!-- use_genes <- intersect(hs_scgenes, hs_prot) -->

<!-- ## mm -->

<!-- mm_gtf <- import("../Data/Mus_musculus.GRCm39.109.gtf") # Ensembl version 109 - June 2, 2023 -->
<!-- mm_gtf_df <- as.data.frame(mm_gtf) -->
<!-- mm_gtf_prot_cod <- mm_gtf_df %>% filter(type == "exon") %>% filter(gene_biotype == "protein_coding") %>% distinct(gene_name, .keep_all = T) -->
<!-- mm_gtf_df <- as.data.frame(mm_gtf) -->
<!-- mm_gtf_prot_cod <- mm_gtf_df %>%  -->
<!--   filter(type == "exon") %>%  -->
<!--   filter(gene_biotype == "protein_coding") %>%  -->
<!--   distinct(gene_name, .keep_all = T) -->

<!-- mm_prot <- mm_gtf_prot_cod %>% pull(gene_name) %>% unique() -->

<!-- # check if all genes are protein-coding -->
<!-- SPZ_untr_4h <- read.csv("..Data/SPZ_untr_4h.csv") -->
<!-- mm_scgenes <- SPZ_untr_4h$GeneSymbol -->

<!-- length(setdiff(mm_scgenes, mm_prot)) -->
<!-- # 10133 -->
<!-- length(intersect(mm_scgenes, mm_prot)) -->
<!-- # 14399 -->

<!-- mm_use_genes <- intersect(mm_scgenes, mm_prot) -->

<!-- ENTREZ_sorted <- mapIds(org.Hs.eg.db, names(iRBC_RBC_geneList_sorted), 'ENTREZID', 'SYMBOL') -->

<!-- ``` -->

<!-- ```{r} -->
<!-- # reduce cts to protein-coding genes -->
<!-- cts <- cts[which(rownames(cts) %in% use_genes),] -->
<!-- ``` -->



## DE - SPZ bys+inf vs RBC bys+inf

```{r}
# SPZ_vs_untr_dual_nothr <- de_function_nothr(condition1 = "control",
#                                  condition2 = "sporozoite_bys_inf")
# SPZ_vs_untr_dual_nothr_geneID <- PF_readable(df = SPZ_vs_untr_dual_nothr)
# 
# DT::datatable(SPZ_vs_untr_dual_nothr_geneID)
##

SPZ_vs_iRBC_dual_nothr <- de_function_nothr(condition1 = "RBC_bys_inf",
                                 condition2 = "sporozoite_bys_inf")
SPZ_vs_iRBC_dual_nothr_geneID <- PF_readable(df = SPZ_vs_iRBC_dual_nothr)

DT::datatable(SPZ_vs_iRBC_dual_nothr_geneID)

##

# iRBC_vs_RBC_dual_nothr <- de_function_nothr(condition1 = "RBC_control",
#                                  condition2 = "RBC_bys_inf")
# iRBC_vs_RBC_dual_nothr_geneID <- PF_readable(df = iRBC_vs_RBC_dual_nothr)
# 
# DT::datatable(iRBC_vs_RBC_dual_nothr_geneID)

#write.csv2(SPZ_vs_untr_dual_nothr_geneID, "SPZ_vs_untr_dual_nothr_geneID.csv", row.names = F, quote = F)
#write.csv2(SPZ_vs_iRBC_dual_nothr_geneID, "SPZ_vs_iRBC_dual_nothr_geneID.csv", row.names = F, quote = F)
#write.csv2(iRBC_vs_RBC_dual_nothr_geneID, "iRBC_vs_RBC_dual_nothr_geneID.csv", row.names = F, quote = F)
```

## GO - SPZ vs iRBC
```{r}
# host and para GO
SPZ_vs_RBC_hostGO <- host_GO(host = "Hs", host_nothr = SPZ_vs_iRBC_dual_nothr_geneID)
SPZ_vs_RBC_plasmoGO <- plasmo_GO(SPZ_vs_iRBC_dual_nothr_geneID)

DT::datatable(SPZ_vs_RBC_hostGO@result)
DT::datatable(SPZ_vs_RBC_plasmoGO[["pf_neg"]])
DT::datatable(SPZ_vs_RBC_plasmoGO[["pf_pos"]])


# write.csv2(SPZ_vs_RBC_hostGO@result, "dual_SPZ_vs_RBC_hostGO.csv", row.names = F, quote = F)
# write.csv2(SPZ_vs_RBC_plasmoGO[[1]], "dual_SPZ_vs_RBC_plasmoGO.csv", row.names = F, quote = F)
```

```{r}
# volcano plot
SPZ_vs_iRBC_volcano <- volcano_plot(de = SPZ_vs_iRBC, plot_title = "RBC vs SPZ1")
SPZ_vs_iRBC_volcano
```

```{r}
#plot_host <- dual_SPZ_vs_RBC_hostGO[rev(order(dual_SPZ_vs_RBC_hostGO$Count)),]
ggplot(dual_SPZ_vs_RBC_hostGO[1:20,], aes(x = Count, y = Description, colour = p.adjust, fill = p.adjust, size = Count)) +
  geom_point() +
  theme_bw() +
  theme(axis.text = element_text(size = 12))

ggsave("dual_SPZ_vs_RBC_hostGO.pdf", width = 8)

#plot_host <- dual_SPZ_vs_RBC_hostGO[rev(order(dual_SPZ_vs_RBC_hostGO$Count)),]
dual_SPZ_vs_RBC_plasmoGO_plot <- dual_SPZ_vs_RBC_plasmoGO %>% 
  filter(!grepl("endonucleolytic", Term)) %>% 
  filter(!grepl("biosynthetic process", Term))
ggplot(dual_SPZ_vs_RBC_plasmoGO_plot[1:20,], aes(x = Significant, y = Term, colour = KS, fill = KS, size = Significant)) +
  geom_point() +
  theme_bw() +
  theme(axis.text = element_text(size = 12))

ggsave("dual_SPZ_vs_RBC_plasmoGO.pdf", width = 8)
```

# Pseudobulk DE analysis 2 {.tabset}
```{r}
#dual <- readRDS("robjects/dual.integrated.TBremoved.cellbender_plasmo.bg.removed_reintegration.rds")

# DefaultAssay(dual) <- "RNA"
# 
# dual@meta.data <- dual@meta.data %>% 
#   mutate(condition_collapse = case_when(
#     (Group == "sporozoite_infected"  | Group == "sporozoite_bystander") ~ "sporozoite_bys_inf",
#     (Group == "RBC_infected" | Group == "RBC_bystander") ~ "RBC_bys_inf",
#     .default = Group
#   )) %>% 
#   mutate(cellID = "Cell")
# 
# table(dual$Donorid, dual$orig.ident)

dual@meta.data$group_donor <- paste(dual@meta.data$Group,
                                               dual@meta.data$Donorid,
                                               sep = "")

# cts <- AggregateExpression(dual,
#               group.by = c("cellID", "cond_collapse_donor"),
#               assays = "RNA",
#               slot = "counts",
#               return.seurat = F)

cts <- my_PseudobulkExpression(dual,
              group.by = c("cellID", "group_donor"),
              assays = "RNA",
              slot = "counts",
              pb.method = "aggregate",
              return.seurat = F)

cts <- cts$RNA

# transpose
# cts.t <- as.data.frame(t(cts))
#
# # split.data.frame
# splitRows <- gsub('_.*', '', rownames(cts.t))
# #splitRows <- sapply(strsplit(rownames(cts.t), split = "D"), function(x) x[1])
# cts.split <- split.data.frame(cts.t,
#                  f = factor(splitRows))
# 
# # fix colnames and transpose
# cts.modified <- lapply(cts.split, function(x){
#   rownames(x) <- gsub("(.+?)(\\_.*)", "\\2", rownames(x))
#   rownames(x) <- gsub("^_", "\\1", rownames(x))
#   t(x)
# })
```



## DE - SPZ_bystander vs RBC_bystander
```{r}
SPZ_bys_vs_RBC_bys_dual_nothr <- de_function_nothr(condition1 = "RBC_bystander",
                                 condition2 = "sporozoite_bystander")
SPZ_bys_vs_RBC_bys_dual_nothr_geneID <- PF_readable(df = SPZ_bys_vs_RBC_bys_dual_nothr)

DT::datatable(SPZ_bys_vs_RBC_bys_dual_nothr_geneID)
```

## GO - SPZ bystander vs RBC bystander - no plasmo enriched gene + GO
```{r, eval = F}
SPZ_bys_vs_RBC_bys_hostGO <- host_GO(host = "Hs", host_nothr = SPZ_bys_vs_RBC_bys_dual_nothr_geneID)
SPZ_bys_vs_RBC_bys_plasmoGO <- plasmo_GO(SPZ_bys_vs_RBC_bys_dual_nothr_geneID)

DT::datatable(SPZ_bys_vs_RBC_bys_hostGO@result)
DT::datatable(SPZ_bys_vs_RBC_bys_plasmoGO[["pf_neg"]])
DT::datatable(SPZ_bys_vs_RBC_bys_plasmoGO[["pf_pos"]])
```


## DE - RBC bystander vs RBC infected
```{r}

RBC_bys_vs_RBC_inf_dual_nothr <- de_function_nothr(condition1 = "RBC_bystander",
                                 condition2 = "RBC_infected")
RBC_bys_vs_RBC_inf_dual_nothr_geneID <- PF_readable(df = RBC_bys_vs_RBC_inf_dual_nothr)

DT::datatable(RBC_bys_vs_RBC_inf_dual_nothr_geneID)


# write.csv2(SPZ_bys_vs_RBC_bys_dual_nothr_geneID, "SPZ_bys_vs_RBC_bys_dual_nothr_geneID.csv", row.names = F, quote = F)
# write.csv2(RBC_bys_vs_RBC_inf_dual_nothr_geneID, "RBC_bys_vs_RBC_inf_dual_nothr_geneID.csv", row.names = F, quote = F)

```

## GO - RBC_infected vs RBC_bystander
```{r}

RBC_bys_vs_RBC_inf_hostGO <- host_GO(host = "Hs", host_nothr = RBC_bys_vs_RBC_inf_dual_nothr_geneID)
RBC_bys_vs_RBC_inf_plasmoGO <- plasmo_GO(RBC_bys_vs_RBC_inf_dual_nothr_geneID)

DT::datatable(RBC_bys_vs_RBC_inf_hostGO@result)
#DT::datatable(RBC_bys_vs_RBC_inf_plasmoGO[["pf_neg"]])
DT::datatable(RBC_bys_vs_RBC_inf_plasmoGO[["pf_pos"]])


# write.csv2(RBC_bys_vs_RBC_inf_hostGO@result, "dual_RBC_bys_vs_RBC_inf_hostGO.csv", row.names = F, quote = F)
# write.csv2(RBC_bys_vs_RBC_inf_plasmoGO[[2]], "dual_RBC_bys_vs_RBC_inf_plasmoGO.csv", row.names = F, quote = F)
```

```{r}
RBC_inf_vs_RBC_bys_volcano <- volcano_plot(de = iRBC_vs_RBC, plot_title = "RBC bystander vs RBC infected1")
```

# Correlation analysis
```{r}
#dual.sce <- as.SingleCellExperiment(dual)
# of.interest.genes.host <- RBC_bys_vs_RBC_inf_dual_nothr_geneID %>%
#   filter(abs(log2FoldChange) > 1.5) %>%
#   filter(padj <= 0.1) %>%
#   mutate(gene = gsub(pattern = "PF3D7_", replacement = "gene-PF3D7-", gene)) %>%
#   filter(!grepl('PF3D7', gene)) %>%
#   pull(gene)
# of.interest.genes.plasmo <- RBC_bys_vs_RBC_inf_dual_nothr_geneID %>%
#   filter(abs(log2FoldChange) > 1.5) %>%
#   filter(padj <= 0.1) %>%
#   mutate(gene = gsub(pattern = "PF3D7_", replacement = "gene-PF3D7-", gene)) %>%
#   filter(grepl('PF3D7', gene)) %>%
#   pull(gene)

of.interest.genes <- RBC_bys_vs_RBC_inf_dual_nothr_geneID %>%
  filter(abs(log2FoldChange) > 1.5) %>%
  filter(padj <= 0.1) %>%
  mutate(gene = gsub(pattern = "PF3D7_", replacement = "gene-PF3D7-", gene)) %>%
  pull(gene)

dual.of.interest.matrix <- dual[of.interest.genes,]@assays$integrated@data %>%
  as.data.frame()
#networks = sclink_net(expr = dual.of.interest.matrix, ncores = 10, lda = seq(0.5, 0.1, -0.05))

system.time(cor.pairs <- correlatePairs(dual.of.interest.matrix, subset.row=rownames(dual.of.interest.matrix), BPPARAM = MulticoreParam(8)))
saveRDS(cor.pairs, "dual.TBplasmoremoved.reanalysed_RBC_bys_vs_RBC_inf_dual_nothr_corpairs.rds")
# cor.pairs

cor.df <- data.frame(gene1 = cor.pairs$gene1,
                        gene2 = cor.pairs$gene2,
                        rho = cor.pairs$rho,
                        p.value = cor.pairs$p.value,
                        FDR = cor.pairs$FDR)

type.cor.df <- cor.df %>% 
  mutate(cor.type = case_when(
    grepl(pattern = "PF3D7", gene1) & !grepl(pattern = "PF3D7", gene2) ~ "hp",
    !grepl(pattern = "PF3D7", gene1) & grepl(pattern = "PF3D7", gene2) ~ "hp",
    grepl(pattern = "PF3D7", gene1) & grepl(pattern = "PF3D7", gene2) ~ "pp",
    !grepl(pattern = "PF3D7", gene1) & !grepl(pattern = "PF3D7", gene2) ~ "hh",
    TRUE ~ "TRUE"
  ))

hp.cor.df <- type.cor.df %>% 
  filter(cor.type == "hp") %>% 
  filter(FDR <= 1e-07) # 1e-07

for(i in 1:nrow(hp.cor.df))
{
  if(grepl(pattern = "PF3D7", hp.cor.df$gene1[i]))
  {
    plasmo_gene <- hp.cor.df$gene1[i]
    hp.cor.df$gene1[i] <- hp.cor.df$gene2[i]
    hp.cor.df$gene2[i] <- plasmo_gene
  }
}

parasite.containing.rows.col1 <- hp.cor.df %>%
  filter(!grepl(pattern = "PF3D7", gene2)) %>%
  filter(grepl(pattern = "PF3D7", gene1))

write.csv2(hp.cor.df, "dual.TBplasmoremoved.reanalysed_RBC_bys_vs_RBC_inf_dual_nothr_corhp.csv", row.names = F, quote = F)

org <- data.frame(Gene = unique(c(hp.cor.df$gene1, hp.cor.df$gene2)))
org <- org %>% 
  mutate(Organism = case_when(
    grepl(pattern = "PF3D7", Gene) ~ "para",
    .default = "host"
  ))

table(org$Organism)

write.csv2(org, "annot_dual.TBplasmoremoved.reanalysed_RBC_bys_vs_RBC_inf_dual_nothr_corhp.csv", row.names = F, quote = F)

# parasite.containing.rows.col1 <- cor.df %>% 
#   filter(grepl(pattern = "PF3D7", gene2)) %>% 
#   filter(!grepl(pattern = "PF3D7", gene1))
```


```{r}
# vis
heatmap.cor <- hp.cor.df %>% 
  select(gene1, gene2, rho) %>% 
  pivot_wider(names_from = gene2, values_from = rho) %>% 
  column_to_rownames("gene1")

heatmap.cor[is.na(heatmap.cor)] <- 0
heatmap <- Heatmap(heatmap.cor,
                   name = "rho",
                   cluster_rows = T,
                   cluster_columns = T,
                   show_column_names = F,
                   height = 21*unit(3.5, "mm"),
                   width = 845*unit(0.18, "mm"),
                   row_names_gp = gpar(fontsize = 6)
                   )
draw(heatmap)

pdf("Heatmap_dual.TBplasmoremoved.reanalysed_RBC_bys_vs_RBC_inf_dual_nothr_corhp.pdf", width = 8)
draw(heatmap)
dev.off()
```

```{r}
#dual.sce <- as.SingleCellExperiment(dual)
# of.interest.genes.host <- RBC_bys_vs_RBC_inf_dual_nothr_geneID %>%
#   filter(abs(log2FoldChange) > 1.5) %>%
#   filter(padj <= 0.1) %>%
#   mutate(gene = gsub(pattern = "PF3D7_", replacement = "gene-PF3D7-", gene)) %>%
#   filter(!grepl('PF3D7', gene)) %>%
#   pull(gene)
# of.interest.genes.plasmo <- RBC_bys_vs_RBC_inf_dual_nothr_geneID %>%
#   filter(abs(log2FoldChange) > 1.5) %>%
#   filter(padj <= 0.1) %>%
#   mutate(gene = gsub(pattern = "PF3D7_", replacement = "gene-PF3D7-", gene)) %>%
#   filter(grepl('PF3D7', gene)) %>%
#   pull(gene)

of.interest.genes <- SPZ_vs_iRBC_dual_nothr_geneID %>%
  filter(abs(log2FoldChange) > 1.5) %>%
  filter(padj <= 0.1) %>%
  mutate(gene = gsub(pattern = "PF3D7_", replacement = "gene-PF3D7-", gene)) %>%
  pull(gene)

dual.subset.conditions <- subset(dual, subset = condition_collapse %in% c("sporozoite_bys_inf", "RBC_bys_inf"))

dual.of.interest.matrix <- dual.subset.conditions[of.interest.genes,]@assays$integrated@data %>%
  as.matrix()
#networks = sclink_net(expr = dual.of.interest.matrix, ncores = 10, lda = seq(0.5, 0.1, -0.05))

system.time(cor.pairs <- correlatePairs(dual.of.interest.matrix, subset.row=rownames(dual.of.interest.matrix), BPPARAM = MulticoreParam(8)))
saveRDS(cor.pairs, "dual.TBplasmoremoved.reanalysed_SPZ_vs_iRBC_dual_nothr_corpairs.rds")
# cor.pairs

cor.df <- data.frame(gene1 = cor.pairs$gene1,
                        gene2 = cor.pairs$gene2,
                        rho = cor.pairs$rho,
                        p.value = cor.pairs$p.value,
                        FDR = cor.pairs$FDR)

type.cor.df <- cor.df %>% 
  mutate(cor.type = case_when(
    grepl(pattern = "PF3D7", gene1) & !grepl(pattern = "PF3D7", gene2) ~ "hp",
    !grepl(pattern = "PF3D7", gene1) & grepl(pattern = "PF3D7", gene2) ~ "hp",
    grepl(pattern = "PF3D7", gene1) & grepl(pattern = "PF3D7", gene2) ~ "pp",
    !grepl(pattern = "PF3D7", gene1) & !grepl(pattern = "PF3D7", gene2) ~ "hh",
    TRUE ~ "TRUE"
  ))

hp.cor.df <- type.cor.df %>% 
  filter(cor.type == "hp") %>% 
  filter(FDR <= 1e-10) # 1e-07

# parasite.containing.rows.col1 <- cor.df %>%
#   filter(grepl(pattern = "PF3D7", gene2)) %>%
#   filter(!grepl(pattern = "PF3D7", gene1))

write.csv2(hp.cor.df, "dual.TBplasmoremoved.reanalysed_SPZ_vs_iRBC_dual_nothr_corhp.csv", row.names = F, quote = F)

org <- data.frame(Gene = unique(c(hp.cor.df$gene1, hp.cor.df$gene2)))
org <- org %>% 
  mutate(Organism = case_when(
    grepl(pattern = "PF3D7", Gene) ~ "para",
    .default = "host"
  ))

table(org$Organism)

write.csv2(org, "annot_dual.TBplasmoremoved.reanalysed_SPZ_vs_iRBC_dual_nothr_corhp.csv", row.names = F, quote = F)


```


```{r}
# vis
for(i in 1:nrow(hp.cor.df))
{
  if(grepl(pattern = "PF3D7", hp.cor.df$gene1[i]))
  {
    plasmo_gene <- hp.cor.df$gene1[i]
    hp.cor.df$gene1[i] <- hp.cor.df$gene2[i]
    hp.cor.df$gene2[i] <- plasmo_gene
  }
}

heatmap.cor <- hp.cor.df %>% 
  select(gene1, gene2, rho) %>% 
  pivot_wider(names_from = gene2, values_from = rho) %>% 
  column_to_rownames("gene1")

heatmap.cor[is.na(heatmap.cor)] <- 0
heatmap <- Heatmap(heatmap.cor,
                   name = "rho",
                   cluster_rows = T,
                   cluster_columns = T,
                   show_column_names = F,
                   height = 146*unit(1, "mm"),
                   width = 761*unit(0.18, "mm"),
                   row_names_gp = gpar(fontsize = 3)
                   )
draw(heatmap)

# pdf("Heatmap_dual.TBplasmoremoved.reanalysed_SPZ_vs_iRBC_dual_nothr_corhp.pdf", width = 8)
# draw(heatmap)
# dev.off()

png("Heatmap_dual.TBplasmoremoved.reanalysed_SPZ_vs_iRBC_dual_nothr_corhp2.png", 
    width = 20, height = 20, units= "cm", res = 300)
draw(heatmap)
dev.off()
```

```{r}
# enlarge a couple of the parts from the heatmap
SPZ_vs_iRBC <- read.csv2("tables/SPZ_vs_iRBC_dual_nothr_geneID.csv")
SPZ_vs_iRBC_corhp <- read.csv2("tables/dual.TBplasmoremoved.reanalysed_SPZ_vs_iRBC_dual_nothr_corhp.csv")

group1 <- c("ISG20", "TNIP", "CCL4L2", "MIR3945HG", "TNF", "TNFAIP3", "IL1B",
            "CCL3L3", "IL1A", "CCL7", "PTX3", "GNG2", "MAP1LC3A", "LAYN", "MT2A",
            "DDX58", "ISG15", "IFIH1", "IL10", "IL1R", "ACOD1", "GDF15",
            "TRGC1", "CSF2", "EGR3", "CCL20", "GCH1", "HERC5", "MCOLN2", "SLAMF1",
            "DUSP2", "IL6", "OASL", "CDKN1A", "CCL5", "INHBA", "LY9", "CCND1",
            "SERPINB7", "IL20")

group1_cor <- SPZ_vs_iRBC_corhp %>% 
  filter(gene1 %in% group1) %>% 
  filter(abs(rho) > 0.4)

heatmap.cor <- group1_cor %>% 
  select(gene1, gene2, rho) %>% 
  mutate(gene2 = gsub("gene-PF3D7-", "PF3D7_", gene2)) %>% 
  pivot_wider(names_from = gene2, values_from = rho) %>% 
  column_to_rownames("gene1")

heatmap.cor[is.na(heatmap.cor)] <- 0
heatmap <- Heatmap(heatmap.cor,
                   name = "rho",
                   col = colorRamp2(c(0.6, 0, -0.6), c("red", "white", "blue")),
                   cluster_rows = T,
                   cluster_columns = T,
                   show_column_names = T,
                   height = nrow(heatmap.cor)*unit(4, "mm"),
                   width = ncol(heatmap.cor)*unit(3, "mm"),
                   row_names_gp = gpar(fontsize = 8),
                   column_names_gp = gpar(fontsize = 8)
                   )
draw(heatmap)

pdf("Heatmap_SPZ_vs_iRBC_group1_corhp.pdf", width = 8)
draw(heatmap)
dev.off()
```


```{r}
group1_cor_rho0.35 <- SPZ_vs_iRBC_corhp %>% 
  filter(gene1 %in% group1) %>% 
  filter(abs(rho) > 0.35)

heatmap.cor <- group1_cor_rho0.35 %>% 
  select(gene1, gene2, rho) %>% 
  mutate(gene2 = gsub("gene-PF3D7-", "PF3D7_", gene2)) %>% 
  pivot_wider(names_from = gene2, values_from = rho) %>% 
  column_to_rownames("gene1")

heatmap.cor[is.na(heatmap.cor)] <- 0
heatmap <- Heatmap(heatmap.cor,
                   name = "rho",
                   col = colorRamp2(c(0.6, 0, -0.6), c("red", "white", "blue")),
                   cluster_rows = T,
                   cluster_columns = T,
                   show_column_names = T,
                   height = nrow(heatmap.cor)*unit(4, "mm"),
                   width = ncol(heatmap.cor)*unit(3, "mm"),
                   row_names_gp = gpar(fontsize = 8),
                   column_names_gp = gpar(fontsize = 8)
                   )
draw(heatmap)

pdf("Heatmap_SPZ_vs_iRBC_group1_corhp_rho0.35.pdf", width = 12)
draw(heatmap)
dev.off()
```


```{r}
group1_cor_rho0.25 <- SPZ_vs_iRBC_corhp %>% 
  filter(gene1 %in% group1) %>% 
  filter(abs(rho) > 0.25)

heatmap.cor <- group1_cor_rho0.25 %>% 
  select(gene1, gene2, rho) %>% 
  mutate(gene2 = gsub("gene-PF3D7-", "PF3D7_", gene2)) %>% 
  pivot_wider(names_from = gene2, values_from = rho) %>% 
  column_to_rownames("gene1")

heatmap.cor[is.na(heatmap.cor)] <- 0
heatmap <- Heatmap(heatmap.cor,
                   name = "rho",
                   col = colorRamp2(c(0.6, 0, -0.6), c("red", "white", "blue")),
                   cluster_rows = T,
                   cluster_columns = T,
                   show_column_names = T,
                   height = nrow(heatmap.cor)*unit(4, "mm"),
                   width = ncol(heatmap.cor)*unit(3, "mm"),
                   row_names_gp = gpar(fontsize = 8),
                   column_names_gp = gpar(fontsize = 8)
                   )
draw(heatmap)

pdf("Heatmap_SPZ_vs_iRBC_group1_corhp_rho0.25.pdf", width = 20)
draw(heatmap)
dev.off()

group1_cor_rho0.25 <- group1_cor_rho0.25 %>% 
  mutate(gene2 = gsub("gene-PF3D7-", "PF3D7_", gene2)) %>% 
  left_join(., SPZ_vs_iRBC, join_by(gene2 == gene)) %>% 
  dplyr::select(gene1, gene2, rho, gene_name)

write.csv2(group1_cor_rho0.25, "Plasmo_names_SPZ_vs_iRBC_group1_corhp_rho0.25.csv", row.names = F, quote = F)
```

