---
title: "Gene sets from bulk RNA-seq"
output:
  html_document:
    df_print: paged
    code_folding: hide
  pdf_document: default
  toc: TRUE
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
knitr::opts_chunk$set(fig.width = 10, fig.height = 8) 
```

```{r}
library(Seurat)
library(patchwork)
library(readxl)
library(tidyverse)
library(ggplot2)
library(ggpubr)
library(reshape2)
library(broom)
library(stringr)
library(ggplot2)
library(SeuratData)
#library(celldex)
#install.packages("cowplot")
#library(cowplot)
#devtools::install_github('cole-trapnell-lab/monocle3')
#library(monocle3)
#remotes::install_github('satijalab/seurat-wrappers')
library(SeuratWrappers)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(ggrepel)
library(data.table)
library(ComplexHeatmap)
library(msigdbr)
library(heatmap3)
library(presto)
library(fgsea)
library(topGO)
library(rtracklayer)
library(circlize)
#library(EnhancedVolcano)
library(clusterProfiler)
library(org.Hs.eg.db)
library(org.Mm.eg.db)
# install.packages("devtools")
# devtools::install_github("saeyslab/nichenetr")
#library(nichenetr)
# mulattoes::install('multtest')
# install.packages('metap')
#library(metap)
#remotes::install_github("satijalab/seurat-data")
```

```{r}
my_PseudobulkExpression <- function(
  object,
  pb.method = 'aggregate',
  assays = NULL,
  features = NULL,
  return.seurat = FALSE,
  group.by = 'ident',
  add.ident = NULL,
  slot = 'data',
  verbose = TRUE,
  ...
) {
  # CheckDots(..., fxns = 'CreateSeuratObject')
  # if (!is.null(x = add.ident)) {
  #   .Deprecated(msg = "'add.ident' is a deprecated argument, please use the 'group.by' argument instead")
  #   group.by <- c('ident', add.ident)
  # }
  # if (!(pb.method %in% c('average', 'aggregate'))) {
  #   stop("'pb.method' must be either 'average' or 'aggregate'")
  # }
  object.assays <- FilterObjects(object = object, classes.keep = 'Assay')
  assays <- assays %||% object.assays
  if (!all(assays %in% object.assays)) {
    assays <- assays[assays %in% object.assays]
    if (length(x = assays) == 0) {
      stop("None of the requested assays are present in the object")
    } else {
      warning("Requested assays that do not exist in object. Proceeding with existing assays only.")
    }
  }
  if (length(x = slot) == 1) {
    slot <- rep_len(x = slot, length.out = length(x = assays))
  } else if (length(x = slot) != length(x = assays)) {
    stop("Number of slots provided does not match number of assays")
  }
  data <- FetchData(object = object, vars = rev(x = group.by))
  data <- data[which(rowSums(x = is.na(x = data)) == 0), , drop = F]
  if (nrow(x = data) < ncol(x = object)) {
    message("Removing cells with NA for 1 or more grouping variables")
    object <- subset(x = object, cells = rownames(x = data))
  }
  for (i in 1:ncol(x = data)) {
    data[, i] <- as.factor(x = data[, i])
  }
  num.levels <- sapply(
    X = 1:ncol(x = data),
    FUN = function(i) {
      length(x = levels(x = data[, i]))
    }
  )
  if (any(num.levels == 1)) {
    message(paste0("The following grouping variables have 1 value and will be ignored: ",
                   paste0(colnames(x = data)[which(num.levels <= 1)], collapse = ", ")))
    group.by <- colnames(x = data)[which(num.levels > 1)]
    data <- data[, which(num.levels > 1), drop = F]
  }
  if (ncol(x = data) == 0) {
    message("All grouping variables have 1 value only. Computing across all cells.")
    category.matrix <- matrix(
      data = 1,
      nrow = ncol(x = object),
      dimnames = list(Cells(x = object), 'all')
    )
    if (pb.method == 'average') {
      category.matrix <- category.matrix / sum(category.matrix)
    }
  } else {
    category.matrix <- model.Matrix(object = as.formula(
      object = paste0(
        '~0+',
        paste0(
          "data[,",
          1:length(x = group.by),
          "]",
          collapse = ":"
        )
      )
    ))
    colsums <- colSums(x = category.matrix)
    category.matrix <- category.matrix[, colsums > 0]
    colsums <- colsums[colsums > 0]
    if (pb.method == 'average') {
      category.matrix <- Sweep(
        x = category.matrix,
        MARGIN = 2,
        STATS = colsums,
        FUN = "/")
    }
    colnames(x = category.matrix) <- sapply(
      X = colnames(x = category.matrix),
      FUN = function(name) {
        name <- gsub(pattern = "data\\[, [1-9]*\\]", replacement = "", x = name)
        return(paste0(rev(x = unlist(x = strsplit(x = name, split = ":"))), collapse = "_"))
      })
  }
  data.return <- list()
  for (i in 1:length(x = assays)) {
    data.use <- GetAssayData(
      object = object,
      assay = assays[i],
      slot = slot[i]
    )
    features.to.avg <- features %||% rownames(x = data.use)
    if (inherits(x = features, what = "list")) {
      features.to.avg <- features[i]
    }
    if (IsMatrixEmpty(x = data.use)) {
      warning(
        "The ", slot[i], " slot for the ", assays[i],
        " assay is empty. Skipping assay.", immediate. = TRUE, call. = FALSE)
      next
    }
    bad.features <- setdiff(x = features.to.avg, y = rownames(x = data.use))
    if (length(x = bad.features) > 0) {
      warning(
        "The following ", length(x = bad.features),
        " features were not found in the ", assays[i], " assay: ",
        paste(bad.features, collapse = ", "), call. = FALSE, immediate. = TRUE)
    }
    features.assay <- intersect(x = features.to.avg, y = rownames(x = data.use))
    if (length(x = features.assay) > 0) {
      data.use <- data.use[features.assay, ]
    } else {
      warning("None of the features specified were found in the ", assays[i],
              " assay.", call. = FALSE, immediate. = TRUE)
      next
    }
    if (slot[i] == 'data') {
      data.use <- expm1(x = data.use)
      if (any(data.use == Inf)) {
        warning("Exponentiation yielded infinite values. `data` may not be log-normed.")
      }
    }
    data.return[[i]] <- as.matrix(x = (data.use %*% category.matrix))
    names(x = data.return)[i] <- assays[[i]]
  }
  if (return.seurat) {
    if (slot[1] == 'scale.data') {
      na.matrix <- data.return[[1]]
      na.matrix[1:length(x = na.matrix)] <- NA
      toRet <- CreateSeuratObject(
        counts = na.matrix,
        project = if (pb.method == "average") "Average" else "Aggregate",
        assay = names(x = data.return)[1],
        check.matrix = FALSE,
        ...
      )
      toRet <- SetAssayData(
        object = toRet,
        assay = names(x = data.return)[1],
        slot = "counts",
        new.data = matrix()
      )
      toRet <- SetAssayData(
        object = toRet,
        assay = names(x = data.return)[1],
        slot = "data",
        new.data = na.matrix
      )
      toRet <- SetAssayData(
        object = toRet,
        assay = names(x = data.return)[1],
        slot = "scale.data",
        new.data = data.return[[1]]
      )
    } else {
      toRet <- CreateSeuratObject(
        counts = data.return[[1]],
        project = if (pb.method == "average") "Average" else "Aggregate",
        assay = names(x = data.return)[1],
        check.matrix = FALSE,
        ...
      )
      toRet <- SetAssayData(
        object = toRet,
        assay = names(x = data.return)[1],
        slot = "data",
        new.data = log1p(x = as.matrix(x = data.return[[1]]))
      )
    }
    #for multimodal data
    if (length(x = data.return) > 1) {
      for (i in 2:length(x = data.return)) {
        if (slot[i] == 'scale.data') {
          na.matrix <- data.return[[i]]
          na.matrix[1:length(x = na.matrix)] <- NA
          toRet[[names(x = data.return)[i]]] <- CreateAssayObject(counts = na.matrix, check.matrix = FALSE)
          toRet <- SetAssayData(
            object = toRet,
            assay = names(x = data.return)[i],
            slot = "counts",
            new.data = matrix()
          )
          toRet <- SetAssayData(
            object = toRet,
            assay = names(x = data.return)[i],
            slot = "data",
            new.data = na.matrix
          )
          toRet <- SetAssayData(
            object = toRet,
            assay = names(x = data.return)[i],
            slot = "scale.data",
            new.data = as.matrix(x = data.return[[i]])
          )
        } else {
          toRet[[names(x = data.return)[i]]] <- CreateAssayObject(counts = data.return[[i]], check.matrix = FALSE)
          toRet <- SetAssayData(
            object = toRet,
            assay = names(x = data.return)[i],
            slot = "data",
            new.data = log1p(x = as.matrix(x = data.return[[i]]))
          )
        }

      }
    }
    if (DefaultAssay(object = object) %in% names(x = data.return)) {
      DefaultAssay(object = toRet) <- DefaultAssay(object = object)
      if (slot[which(DefaultAssay(object = object) %in% names(x = data.return))[1]] != 'scale.data') {
        toRet <- ScaleData(object = toRet, verbose = verbose)
      }
    }
    if ('ident' %in% group.by) {
      first.cells <- c()
      for (i in 1:ncol(x = category.matrix)) {
        first.cells <- c(first.cells, Position(x = category.matrix[,i], f = function(x) {x > 0}))
      }
      Idents(object = toRet) <- Idents(object = object)[first.cells]
    }
    return(toRet)
  } else {
    return(data.return)
  }
}
```

```{r}
immune.combined <- readRDS("robjects/immune.combined.TBremoved.rds")

DefaultAssay(immune.combined) <- "RNA"

immune.combined@meta.data <- immune.combined@meta.data %>% 
  mutate(condition_collapse = case_when(
    (Group == "sporozoite_infected"  | Group == "sporozoite_bystander") ~ "sporozoite_bys_inf",
    (Group == "RBC_infected" | Group == "RBC_bystander") ~ "RBC_bys_inf",
    .default = Group
  ))

table(immune.combined$Donorid, immune.combined$orig.ident)

immune.combined@meta.data$cond_collapse_donor <- paste(immune.combined@meta.data$condition_collapse,
                                               immune.combined@meta.data$Donorid,
                                               sep = "")

# cts <- AggregateExpression(immune.combined,
#               group.by = c("cellID", "cond_collapse_donor"),
#               assays = "RNA",
#               slot = "counts",
#               return.seurat = F)

cts <- my_PseudobulkExpression(immune.combined,
              group.by = c("cellID", "cond_collapse_donor"),
              assays = "RNA",
              slot = "counts",
              pb.method = "aggregate",
              return.seurat = F)

cts <- cts$RNA

# transpose
# cts.t <- as.data.frame(t(cts))
#
# # split.data.frame
# splitRows <- gsub('_.*', '', rownames(cts.t))
# #splitRows <- sapply(strsplit(rownames(cts.t), split = "D"), function(x) x[1])
# cts.split <- split.data.frame(cts.t,
#                  f = factor(splitRows))
# 
# # fix colnames and transpose
# cts.modified <- lapply(cts.split, function(x){
#   rownames(x) <- gsub("(.+?)(\\_.*)", "\\2", rownames(x))
#   rownames(x) <- gsub("^_", "\\1", rownames(x))
#   t(x)
# })
```

```{r}

hs_gtf <- import("../Data/Homo_sapiens.GRCh38.109.gtf") # Ensembl version 109 - June 2, 2023
hs_gtf_df <- as.data.frame(hs_gtf)
hs_gtf_prot_cod <- hs_gtf_df %>% filter(type == "exon") %>% filter(gene_biotype == "protein_coding") %>% distinct(gene_name, .keep_all = T)
hs_gtf_df <- as.data.frame(hs_gtf)
hs_gtf_prot_cod <- hs_gtf_df %>% 
  filter(type == "exon") %>% 
  filter(gene_biotype == "protein_coding") %>% 
  distinct(gene_name, .keep_all = T)

hs_prot <- hs_gtf_prot_cod %>% pull(gene_name) %>% unique()

# check if all genes are protein-coding
hs_scgenes <- rownames(cts)

length(setdiff(hs_scgenes, hs_prot))
# 10133
length(intersect(hs_scgenes, hs_prot))
# 14399

use_genes <- intersect(hs_scgenes, hs_prot)

## mm

mm_gtf <- import("../Data/Mus_musculus.GRCm39.109.gtf") # Ensembl version 109 - June 2, 2023
mm_gtf_df <- as.data.frame(mm_gtf)
mm_gtf_prot_cod <- mm_gtf_df %>% filter(type == "exon") %>% filter(gene_biotype == "protein_coding") %>% distinct(gene_name, .keep_all = T)
mm_gtf_df <- as.data.frame(mm_gtf)
mm_gtf_prot_cod <- mm_gtf_df %>% 
  filter(type == "exon") %>% 
  filter(gene_biotype == "protein_coding") %>% 
  distinct(gene_name, .keep_all = T)

mm_prot <- mm_gtf_prot_cod %>% pull(gene_name) %>% unique()

# check if all genes are protein-coding
SPZ_untr_4h <- read.csv("../Data/SPZ_untr_4h.csv")
mm_scgenes <- SPZ_untr_4h$GeneSymbol

length(setdiff(mm_scgenes, mm_prot))
# 10133
length(intersect(mm_scgenes, mm_prot))
# 14399

mm_use_genes <- intersect(mm_scgenes, mm_prot)

#ENTREZ_sorted <- mapIds(org.Hs.eg.db, names(iRBC_RBC_geneList_sorted), 'ENTREZID', 'SYMBOL')

```

```{r}
# reduce cts to protein-coding genes
cts <- cts[which(rownames(cts) %in% use_genes),]
```

```{r}

# Monocytes activated (type I IFN signature)
# Run DE analysis
# 1. Get counts matrix

de_function <- function(condition1, condition2, padj_thr_sc, logFC_thr)
{
  padj_thr_sc <- padj_thr_sc
  
  counts <- cts %>% 
    as.data.frame() %>% 
    dplyr::select(c(starts_with(condition1), starts_with(condition2)))

  # 2 Generate metadata
  colData <- data.frame(sample = colnames(counts))
  colData$condition = sapply(str_split(colData$sample, "D"), function(x) x[1])
  colData <- colData %>% 
    tibble::column_to_rownames("sample")
  
  # more info for metadata
  
  # DEseq2 object
  dds <- DESeqDataSetFromMatrix(countData = counts,
                               colData = colData,
                               design = ~condition)
  
  
  # filter dds
  keep <- rowSums(counts(dds)) >= 10
  dds <- dds[keep,]
  
  dds$condition <- relevel(dds$condition, ref = condition1)
  
  dds <- DESeq(dds)
  
  comparison <- resultsNames(dds)[2]

  res <- results(dds, name = comparison)
  
  # Turn the DESeq2 results object into a tibble for use with tidyverse functions
  res_tbl <- res %>%
    data.frame() %>%
    rownames_to_column(var = "gene") %>%
    as_tibble()
  
  
  
  #write.csv(res_tbl, "iRBC_vs_RBC_scRNAseq.csv", row.names = F, quote = F)
  ####
  
  # Subset the significant results
  sig_res <- dplyr::filter(res_tbl, padj <= padj_thr_sc) %>%
    filter(abs(log2FoldChange) >= logFC_thr) %>% 
    dplyr::arrange(padj)
  
  # Check significant genes output
  #sig_res
  
  ## Extract normalized counts from dds object
  normalized_counts <- counts(dds, normalized = TRUE)
  
  ## Extract normalized counts for significant genes only
  sig_counts <- normalized_counts[rownames(normalized_counts) %in% sig_res$gene, ]
  
  
  return(list(dds,sig_counts, sig_res))
  
}

de_function_nothr <- function(condition1, condition2)
{
  #padj_thr_sc <- padj_thr_sc
  
  counts <- cts %>% 
    as.data.frame() %>% 
    dplyr::select(c(starts_with(condition1), starts_with(condition2)))

  # 2 Generate metadata
  colData <- data.frame(sample = colnames(counts))
  colData$condition = sapply(str_split(colData$sample, "D"), function(x) x[1])
  colData <- colData %>% 
    tibble::column_to_rownames("sample")
  
  # more info for metadata
  
  # DEseq2 object
  dds <- DESeqDataSetFromMatrix(countData = counts,
                               colData = colData,
                               design = ~condition)
  
  
  # filter dds
  keep <- rowSums(counts(dds)) >= 10
  dds <- dds[keep,]
  
  dds$condition <- relevel(dds$condition, ref = condition1)
  
  dds <- DESeq(dds)
  
  comparison <- resultsNames(dds)[2]

  res <- results(dds, name = comparison)
  
  # Turn the DESeq2 results object into a tibble for use with tidyverse functions
  res_tbl <- res %>%
    data.frame() %>%
    rownames_to_column(var = "gene") %>%
    as_tibble()
  
  
  
  #write.csv(res_tbl, "iRBC_vs_RBC_scRNAseq.csv", row.names = F, quote = F)
  ####
  
  # # Subset the significant results
  # sig_res <- dplyr::filter(res_tbl, padj <= padj_thr_sc) %>%
  #   filter(abs(log2FoldChange) >= logFC_thr) %>% 
  #   dplyr::arrange(padj)
  # 
  # # Check significant genes output
  # #sig_res
  # 
  # ## Extract normalized counts from dds object
  # normalized_counts <- counts(dds, normalized = TRUE)
  # 
  # ## Extract normalized counts for significant genes only
  # sig_counts <- normalized_counts[rownames(normalized_counts) %in% sig_res$gene, ]
  # 
  return(res_tbl)
  #return(list(dds,sig_counts, sig_res))
  
}

convertMouseGeneList <- function(x){
require("biomaRt")
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl", host = "dec2021.archive.ensembl.org")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl", host = "dec2021.archive.ensembl.org")
genesV2 = getLDS(attributes = c("mgi_symbol"), filters = "mgi_symbol", values = x , mart = mouse, attributesL = c("hgnc_symbol"), martL = human, uniqueRows=T)
humanx <- unique(genesV2[, 2])
return(humanx)
}

bulkRNAseq_comparison <- function(file, condition1, condition2, padj_thr_bulk, padj_thr_sc, logFC_thr)
{
  # bulk
  bulk_table <- read.csv(paste0("../Data/", file, collapse = ''))
  table_genes <- bulk_table %>% 
    filter(abs(log2FoldChange) >= logFC_thr) %>% 
    filter(padj <= padj_thr_bulk) 
  
  bulk_pos <- table_genes %>% 
    filter(log2FoldChange >= 0)
  
  bulk_neg <- table_genes %>% 
    filter(log2FoldChange < 0)
  
  table_genes_pos <- bulk_pos %>% 
    pull(GeneSymbol)
  
  table_genes_neg <- bulk_neg %>% 
    pull(GeneSymbol)
  
  # scRNAseq
  scRNA_pseudobulk_list <- de_function(condition1 = condition1, 
                                        condition2 = condition2,
                                       padj_thr_sc = padj_thr_sc,
                                       logFC_thr = logFC_thr
                                        #celltype = celltype,
                                        #geneset = table_genes_v
                                       )
  
  dds <- scRNA_pseudobulk_list[[1]]
  
  sig.genes_table <- scRNA_pseudobulk_list[[2]]
  
  human_sig_res <- scRNA_pseudobulk_list[[3]]
  
  if(length(table_genes_pos) > 0)
  {
    humangenes_pos <- convertMouseGeneList(table_genes_pos)
    
    sig.genes_pvalue_pos <- scRNA_pseudobulk_list[[3]] %>% 
    filter(log2FoldChange >= logFC_thr) %>% 
    left_join(data.frame(gene = humangenes_pos),.) %>% 
    na.omit() %>% 
    dplyr::arrange(padj)
    
    sig.genes_table_pos <- sig.genes_table %>% 
    as.data.frame() %>% 
    rownames_to_column("Gene") %>% 
    filter(Gene %in% sig.genes_pvalue_pos$gene) %>% 
    column_to_rownames("Gene")
  }else{
    sig.genes_table_pos = 0
    sig.genes_pvalue_pos = 0
  }
   
  
  humangenes_neg <- convertMouseGeneList(table_genes_neg)

  sig.genes_pvalue_neg <- scRNA_pseudobulk_list[[3]] %>% 
    filter(log2FoldChange <= -logFC_thr) %>% 
    left_join(data.frame(gene = humangenes_neg),.) %>% 
    na.omit()%>% 
    dplyr::arrange(padj)
  
  # getting genes common to bulk data
  
sig.genes_table_neg <- sig.genes_table %>% 
    as.data.frame() %>% 
    rownames_to_column("Gene") %>% 
    filter(Gene %in% sig.genes_pvalue_neg$gene) %>% 
    column_to_rownames("Gene")
  

  vsd <- assay(vst(dds, 
           nsub = sum( rowMeans( counts(dds, normalized=TRUE)) > 5 )))
  z <- t(scale(t(vsd)))
  z_pos <- z[rownames(z) %in% rownames(sig.genes_table_pos),]
  z_neg <- z[rownames(z) %in% rownames(sig.genes_table_neg),]

  #return(list(bulktable_genes = table_genes, z = z, sig.genes_exp = sig.genes_table, sig_genes_pvalue = sig.genes_pvalue))
  
  return(list(bulktable_genes = table_genes, z_pos = z_pos, z_neg = z_neg, human_sig_res = human_sig_res, sig.genes_exp_pos = sig.genes_table_pos, sig.genes_exp_neg = sig.genes_table_neg, sig_genes_pvalue_pos = sig.genes_pvalue_pos, sig_genes_pvalue_neg = sig.genes_pvalue_neg))
}

heat_colors <- rev(brewer.pal(11, "PuOr"))

gsea <- function(category, deg)
{
  m_df<- msigdbr(species = "Homo sapiens", category = category)
  fgsea_sets<- m_df %>% split(x = .$gene_symbol, f = .$gs_name)

  gene_logFC_rank <- deg[,c("gene", "log2FoldChange")] %>% 
    arrange(-log2FoldChange) %>% 
    deframe(.)
    
  fgseaRes<- fgsea(fgsea_sets, stats = gene_logFC_rank, nperm = 1000)
  
  fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))
  
  return(list(fgsea_sets = fgsea_sets, gene_logFC_rank = gene_logFC_rank, fgseaResTidy = fgseaResTidy))
}
  
```

# All cells

## Sporozoite_infected vs control (base)

```{r}

SPZ_GFPpos_vs_untr <- bulkRNAseq_comparison(file = "SPZ-GFPpos_untr_df_4h.csv",
                                            condition1 = "control",
                                            condition2 = "sporozoite_infected",
                                            #celltype = "Activated monocytes",
                                            padj_thr_bulk = 0.1,
                                            padj_thr_sc = 0.1,
                                            logFC_thr = 2)
DT::datatable(SPZ_GFPpos_vs_untr[["bulktable_genes"]])

DT::datatable(SPZ_GFPpos_vs_untr[["sig_genes_pvalue_pos"]])  

DT::datatable(SPZ_GFPpos_vs_untr[["sig_genes_pvalue_neg"]])  


z_pos <- SPZ_GFPpos_vs_untr[["z_pos"]] %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("gene") %>%
  left_join(data.frame(gene = SPZ_GFPpos_vs_untr[["sig_genes_pvalue_pos"]][,"gene"]),.) %>%
  tibble::column_to_rownames("gene")

z_neg <- SPZ_GFPpos_vs_untr[["z_neg"]] %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("gene") %>%
  left_join(data.frame(gene = SPZ_GFPpos_vs_untr[["sig_genes_pvalue_neg"]][,"gene"]),.) %>%
  tibble::column_to_rownames("gene")

Heatmap(z_pos[1:60,], name = "z-score", 
        width = ncol(z_pos)*unit(5, "mm"), 
        height = 60*unit(2.5, "mm"),
        col = heat_colors, 
        cluster_rows= T, cluster_columns = T, 
        show_column_dend = T, show_row_dend = F, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize = 5),
        column_names_side = "top", column_names_rot = 45, 
        column_names_gp = gpar(fontsize = 6),
        #row_km = 2, 
        column_km = 2, border = TRUE, 
        column_title = NULL, row_title = NULL, 
        border_gp = gpar(col = "darkgrey"))

Heatmap(z_neg, name = "z-score", 
        width = ncol(z_neg)*unit(5, "mm"), 
        height = nrow(z_neg)*unit(2.5, "mm"),
        col = heat_colors, 
        cluster_rows= T, cluster_columns = T, 
        show_column_dend = T, show_row_dend = F, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize = 5),
        column_names_side = "top", column_names_rot = 45, 
        column_names_gp = gpar(fontsize = 6),
        #row_km = 2, 
        column_km = 2, border = TRUE, 
        column_title = NULL, row_title = NULL, 
        border_gp = gpar(col = "darkgrey"))

## fgsea

# C2

fgsea_SPZ_GFPpos_vs_untr_C2 <- gsea(category = "C2", deg = rbind(SPZ_GFPpos_vs_untr[["sig_genes_pvalue_pos"]], SPZ_GFPpos_vs_untr[["sig_genes_pvalue_neg"]]))

DT::datatable(fgsea_SPZ_GFPpos_vs_untr_C2[["fgseaResTidy"]])

ggplot(fgsea_SPZ_GFPpos_vs_untr_C2[["fgseaResTidy"]] %>% filter(pval < 0.008) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= NES)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") +
  theme_minimal()

plotEnrichment(fgsea_SPZ_GFPpos_vs_untr_C2[["fgsea_sets"]][["REACTOME_IRF3_MEDIATED_INDUCTION_OF_TYPE_I_IFN"]],
               fgsea_SPZ_GFPpos_vs_untr_C2[["gene_logFC_rank"]]) + labs(title="REACTOME_IRF3_MEDIATED_INDUCTION_OF_TYPE_I_IFN")

plotEnrichment(fgsea_SPZ_GFPpos_vs_untr_C2[["fgsea_sets"]][["REACTOME_ACTIVATION_OF_IRF3_IRF7_MEDIATED_BY_TBK1_IKK_EPSILON"]],
               fgsea_SPZ_GFPpos_vs_untr_C2[["gene_logFC_rank"]]) + labs(title="REACTOME_ACTIVATION_OF_IRF3_IRF7_MEDIATED_BY_TBK1_IKK_EPSILON")

# C5

fgsea_SPZ_GFPpos_vs_untr_C5 <- gsea(category = "C5", deg = rbind(SPZ_GFPpos_vs_untr[["sig_genes_pvalue_pos"]], SPZ_GFPpos_vs_untr[["sig_genes_pvalue_neg"]]))

DT::datatable(fgsea_SPZ_GFPpos_vs_untr_C5[["fgseaResTidy"]])

ggplot(fgsea_SPZ_GFPpos_vs_untr_C5[["fgseaResTidy"]] %>% filter(pval < 0.008) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= NES)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") +
  theme_minimal()

plotEnrichment(fgsea_SPZ_GFPpos_vs_untr_C5[["fgsea_sets"]][["GOBP_RESPONSE_TO_WOUNDING"]],
               fgsea_SPZ_GFPpos_vs_untr_C5[["gene_logFC_rank"]]) + labs(title="GOBP_RESPONSE_TO_WOUNDING")


```

# Sporozoite bystander vs untreated

```{r}

SPZ_GFPneg_vs_untr <- bulkRNAseq_comparison(file = "SPZ-GFPneg_untr_df_4h.csv",
                                            condition1 = "control",
                                            condition2 = "sporozoite_bystander",
                                            #celltype = "Activated monocytes",
                                            padj_thr_bulk = 0.1,
                                            padj_thr_sc = 0.1,
                                            logFC_thr = 2)

DT::datatable(SPZ_GFPneg_vs_untr[["bulktable_genes"]])

DT::datatable(SPZ_GFPneg_vs_untr[["sig_genes_pvalue_pos"]])  

DT::datatable(SPZ_GFPneg_vs_untr[["sig_genes_pvalue_neg"]])  


z_pos <- SPZ_GFPneg_vs_untr[["z_pos"]] %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("gene") %>%
  left_join(data.frame(gene = SPZ_GFPneg_vs_untr[["sig_genes_pvalue_pos"]][,"gene"]),.) %>%
  tibble::column_to_rownames("gene")

z_neg <- SPZ_GFPneg_vs_untr[["z_neg"]] %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("gene") %>%
  left_join(data.frame(gene = SPZ_GFPneg_vs_untr[["sig_genes_pvalue_neg"]][,"gene"]),.) %>%
  tibble::column_to_rownames("gene")

Heatmap(z_pos[1:60,], name = "z-score", 
        width = ncol(z_pos)*unit(5, "mm"), 
        height = 60*unit(2.5, "mm"),
        col = heat_colors, 
        cluster_rows= T, cluster_columns = T, 
        show_column_dend = T, show_row_dend = F, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize = 5),
        column_names_side = "top", column_names_rot = 45, 
        column_names_gp = gpar(fontsize = 6),
        #row_km = 2, 
        column_km = 2, border = TRUE, 
        column_title = NULL, row_title = NULL, 
        border_gp = gpar(col = "darkgrey"))

Heatmap(z_neg, name = "z-score", 
        width = ncol(z_neg)*unit(5, "mm"), 
        height = nrow(z_neg)*unit(2.5, "mm"),
        col = heat_colors, 
        cluster_rows= T, cluster_columns = T, 
        show_column_dend = T, show_row_dend = F, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize = 5),
        column_names_side = "top", column_names_rot = 45, 
        column_names_gp = gpar(fontsize = 6),
        #row_km = 2, 
        column_km = 2, border = TRUE, 
        column_title = NULL, row_title = NULL, 
        border_gp = gpar(col = "darkgrey"))

## fgsea

# C2

fgsea_SPZ_GFPneg_vs_untr_C2 <- gsea(category = "C2", deg = rbind(SPZ_GFPneg_vs_untr[["sig_genes_pvalue_pos"]], SPZ_GFPneg_vs_untr[["sig_genes_pvalue_neg"]]))

DT::datatable(fgsea_SPZ_GFPneg_vs_untr_C2[["fgseaResTidy"]])

ggplot(fgsea_SPZ_GFPneg_vs_untr_C2[["fgseaResTidy"]] %>% filter(pval < 0.008) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= NES)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") +
  theme_minimal()

plotEnrichment(fgsea_SPZ_GFPneg_vs_untr_C2[["fgsea_sets"]][["REACTOME_IRF3_MEDIATED_INDUCTION_OF_TYPE_I_IFN"]],
               fgsea_SPZ_GFPneg_vs_untr_C2[["gene_logFC_rank"]]) + labs(title="REACTOME_IRF3_MEDIATED_INDUCTION_OF_TYPE_I_IFN")

plotEnrichment(fgsea_SPZ_GFPneg_vs_untr_C2[["fgsea_sets"]][["REACTOME_ACTIVATION_OF_IRF3_IRF7_MEDIATED_BY_TBK1_IKK_EPSILON"]],
               fgsea_SPZ_GFPneg_vs_untr_C2[["gene_logFC_rank"]]) + labs(title="REACTOME_ACTIVATION_OF_IRF3_IRF7_MEDIATED_BY_TBK1_IKK_EPSILON")

# C5

fgsea_SPZ_GFPneg_vs_untr_C5 <- gsea(category = "C5", deg = rbind(SPZ_GFPneg_vs_untr[["sig_genes_pvalue_pos"]], SPZ_GFPneg_vs_untr[["sig_genes_pvalue_neg"]]))

DT::datatable(fgsea_SPZ_GFPneg_vs_untr_C5[["fgseaResTidy"]])

ggplot(fgsea_SPZ_GFPneg_vs_untr_C5[["fgseaResTidy"]] %>% filter(pval < 0.008) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= NES)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") +
  theme_minimal()

plotEnrichment(fgsea_SPZ_GFPneg_vs_untr_C5[["fgsea_sets"]][["GOBP_RESPONSE_TO_WOUNDING"]],
               fgsea_SPZ_GFPneg_vs_untr_C5[["gene_logFC_rank"]]) + labs(title="GOBP_RESPONSE_TO_WOUNDING")
```

# SPZ infected vs SPZ bystander

```{r}

SPZ_GFPpos_vs_SPZ_GFPneg <- bulkRNAseq_comparison(file = "SPZ-GFPpos_SPZ-GFPneg_df.csv",
                                            condition1 = "sporozoite_bystander",
                                            condition2 = "sporozoite_infected",
                                            #celltype = "Activated monocytes",
                                            padj_thr_bulk = 0.1,
                                            padj_thr_sc = 0.1,
                                            logFC_thr = 2)

DT::datatable(SPZ_GFPpos_vs_SPZ_GFPneg[["bulktable_genes"]])

DT::datatable(SPZ_GFPpos_vs_SPZ_GFPneg[["sig_genes_pvalue_pos"]])  

DT::datatable(SPZ_GFPpos_vs_SPZ_GFPneg[["sig_genes_pvalue_neg"]])  


# z_pos <- SPZ_GFPpos_vs_SPZ_GFPneg[["z_pos"]] %>% 
#   as.data.frame() %>% 
#   tibble::rownames_to_column("gene") %>%
#   left_join(data.frame(gene = SPZ_GFPpos_vs_SPZ_GFPneg[["sig_genes_pvalue_pos"]][,"gene"]),.) %>%
#   tibble::column_to_rownames("gene")
# 
# z_neg <- SPZ_GFPpos_vs_SPZ_GFPneg[["z_neg"]] %>% 
#   as.data.frame() %>% 
#   tibble::rownames_to_column("gene") %>%
#   left_join(data.frame(gene = SPZ_GFPpos_vs_SPZ_GFPneg[["sig_genes_pvalue_neg"]][,"gene"]),.) %>%
#   tibble::column_to_rownames("gene")

# Heatmap(z_pos[1:60,], name = "z-score", 
#         width = ncol(z_pos)*unit(5, "mm"), 
#         height = 60*unit(2.5, "mm"),
#         col = heat_colors, 
#         cluster_rows= T, cluster_columns = T, 
#         show_column_dend = T, show_row_dend = F, 
#         show_row_names = T, 
#         row_names_gp = gpar(fontsize = 5),
#         column_names_side = "top", column_names_rot = 45, 
#         column_names_gp = gpar(fontsize = 6),
#         #row_km = 2, 
#         column_km = 2, border = TRUE, 
#         column_title = NULL, row_title = NULL, 
#         border_gp = gpar(col = "darkgrey"))
# 
# Heatmap(z_neg, name = "z-score", 
#         width = ncol(z_neg)*unit(5, "mm"), 
#         height = nrow(z_neg)*unit(2.5, "mm"),
#         col = heat_colors, 
#         cluster_rows= T, cluster_columns = T, 
#         show_column_dend = T, show_row_dend = F, 
#         show_row_names = T, 
#         row_names_gp = gpar(fontsize = 5),
#         column_names_side = "top", column_names_rot = 45, 
#         column_names_gp = gpar(fontsize = 6),
#         #row_km = 2, 
#         column_km = 2, border = TRUE, 
#         column_title = NULL, row_title = NULL, 
#         border_gp = gpar(col = "darkgrey"))

```

# RBC bystander vs RBC control

```{r}
# RBC bystander vs untreated
# iRBC_GFPneg_vs_RBC_untr <- bulkRNAseq_comparison(file = "iRBC-GFPneg_RBC_df_4h.csv",
#                                             condition1 = "RBC_control",
#                                             condition2 = "RBC_bystander",
#                                             #celltype = "Activated monocytes",
#                                             padj_thr_bulk = 0.05,
#                                             padj_thr_sc = 0.1,
#                                             logFC_thr = 1.5)
# DT::datatable(iRBC_GFPneg_vs_RBC_untr[[1]])
# 
# z <- iRBC_GFPneg_vs_RBC_untr[[2]] %>% 
#   as.data.frame() %>% 
#   tibble::rownames_to_column("gene") %>% 
#   left_join(iRBC_GFPneg_vs_RBC_untr[[4]][,"gene"],.) %>% 
#   tibble::column_to_rownames("gene")
# 
# DT::datatable(iRBC_GFPneg_vs_RBC_untr[["sig_genes_pvalue"]]) 
# 
# Heatmap(z, name = "z-score", 
#         width = ncol(z)*unit(5, "mm"), 
#         height = nrow(z)*unit(2.5, "mm"),
#         col = heat_colors, 
#         cluster_rows= T, cluster_columns = T, 
#         show_column_dend = T, show_row_dend = F, 
#         show_row_names = T, 
#         row_names_gp = gpar(fontsize = 5),
#         column_names_side = "top", column_names_rot = 45, 
#         column_names_gp = gpar(fontsize = 6),
#         #row_km = 2, 
#         column_km = 2, border = TRUE, 
#         column_title = NULL, row_title = NULL, 
#         border_gp = gpar(col = "darkgrey"))

```

# RBC infected vs RBC control

```{r}
# RBC bystander vs untreated

iRBC_GFPpos_vs_RBC_untr <- bulkRNAseq_comparison(file = "iRBC-GFPpos_RBC_df_4h.csv",
                                            condition1 = "RBC_control",
                                            condition2 = "RBC_infected",
                                           # celltype = "Activated monocytes",
                                            padj_thr_bulk = 0.1,
                                            padj_thr_sc = 0.1,
                                            logFC_thr = 1.5)
DT::datatable(iRBC_GFPpos_vs_RBC_untr[[1]])

# z_pos <- SPZ_GFPpos_vs_SPZ_GFPneg[["z_pos"]] %>% 
#   as.data.frame() %>% 
#   tibble::rownames_to_column("gene") %>%
#   left_join(data.frame(gene = SPZ_GFPpos_vs_SPZ_GFPneg[["sig_genes_pvalue_pos"]][,"gene"]),.) %>%
#   tibble::column_to_rownames("gene")
# 
# z_neg <- SPZ_GFPpos_vs_SPZ_GFPneg[["z_neg"]] %>% 
#   as.data.frame() %>% 
#   tibble::rownames_to_column("gene") %>%
#   left_join(data.frame(gene = SPZ_GFPpos_vs_SPZ_GFPneg[["sig_genes_pvalue_neg"]][,"gene"]),.) %>%
#   tibble::column_to_rownames("gene")

# Heatmap(z_pos[1:60,], name = "z-score", 
#         width = ncol(z_pos)*unit(5, "mm"), 
#         height = 60*unit(2.5, "mm"),
#         col = heat_colors, 
#         cluster_rows= T, cluster_columns = T, 
#         show_column_dend = T, show_row_dend = F, 
#         show_row_names = T, 
#         row_names_gp = gpar(fontsize = 5),
#         column_names_side = "top", column_names_rot = 45, 
#         column_names_gp = gpar(fontsize = 6),
#         #row_km = 2, 
#         column_km = 2, border = TRUE, 
#         column_title = NULL, row_title = NULL, 
#         border_gp = gpar(col = "darkgrey"))
# 
# Heatmap(z_neg, name = "z-score", 
#         width = ncol(z_neg)*unit(5, "mm"), 
#         height = nrow(z_neg)*unit(2.5, "mm"),
#         col = heat_colors, 
#         cluster_rows= T, cluster_columns = T, 
#         show_column_dend = T, show_row_dend = F, 
#         show_row_names = T, 
#         row_names_gp = gpar(fontsize = 5),
#         column_names_side = "top", column_names_rot = 45, 
#         column_names_gp = gpar(fontsize = 6),
#         #row_km = 2, 
#         column_km = 2, border = TRUE, 
#         column_title = NULL, row_title = NULL, 
#         border_gp = gpar(col = "darkgrey"))

```

# RBC infected vs RBC bystander

```{r}

iRBC_GFPpos_vs_iRBC_GFPneg <- bulkRNAseq_comparison(file = "iRBC-GFPpos_iRBC-GFPneg_df_4h.csv",
                                            condition1 = "RBC_bystander",
                                            condition2 = "RBC_infected",
                                            #celltype = "Activated monocytes",
                                            padj_thr_bulk = 0.1,
                                            padj_thr_sc = 0.1,
                                            logFC_thr = 1.5)
DT::datatable(iRBC_GFPpos_vs_iRBC_GFPneg[[1]])

# z <- iRBC_GFPpos_vs_iRBC_GFPneg[[2]] %>% 
#   as.data.frame() %>% 
#   tibble::rownames_to_column("gene") %>% 
#   left_join(iRBC_GFPpos_vs_iRBC_GFPneg[[4]][,"gene"],.) %>% 
#   tibble::column_to_rownames("gene")
# 
# DT::datatable(iRBC_GFPpos_vs_iRBC_GFPneg[["sig_genes_pvalue"]]) 
# 
# Heatmap(z, name = "z-score", 
#         width = ncol(z)*unit(5, "mm"), 
#         height = nrow(z)*unit(2.5, "mm"),
#         col = heat_colors, 
#         cluster_rows= T, cluster_columns = T, 
#         show_column_dend = T, show_row_dend = F, 
#         show_row_names = T, 
#         row_names_gp = gpar(fontsize = 5),
#         column_names_side = "top", column_names_rot = 45, 
#         column_names_gp = gpar(fontsize = 6),
#         #row_km = 2, 
#         column_km = 2, border = TRUE, 
#         column_title = NULL, row_title = NULL, 
#         border_gp = gpar(col = "darkgrey"))

```

# SPZ vs iRBC

```{r}
SPZ_vs_iRBC <- bulkRNAseq_comparison(file = "SPZ-iRBC_df_4h.csv",
                                            condition1 = "RBC_infected",
                                            condition2 = "sporozoite_infected",
                                            #celltype = "Activated monocytes",
                                            padj_thr_bulk = 0.1,
                                            padj_thr_sc = 0.1,
                                            logFC_thr = 1.5)

DT::datatable(SPZ_vs_iRBC[["bulktable_genes"]])

DT::datatable(SPZ_vs_iRBC[["sig_genes_pvalue_pos"]])  

DT::datatable(SPZ_vs_iRBC[["sig_genes_pvalue_neg"]])  


z_pos <- SPZ_vs_iRBC[["z_pos"]] %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("gene") %>%
  left_join(data.frame(gene = SPZ_vs_iRBC[["sig_genes_pvalue_pos"]][,"gene"]),.) %>%
  tibble::column_to_rownames("gene")

z_neg <- SPZ_vs_iRBC[["z_neg"]] %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("gene") %>%
  left_join(data.frame(gene = SPZ_vs_iRBC[["sig_genes_pvalue_neg"]][,"gene"]),.) %>%
  tibble::column_to_rownames("gene")

#png("SPZ_vs_iRBC_heatmap.png", res = 300, width = 15, height = 20, units = "cm")
Heatmap(z_pos, name = "z-score", 
        width = ncol(z_pos)*unit(5, "mm"), 
        height = 60*unit(2.5, "mm"),
        col = heat_colors, 
        cluster_rows= T, cluster_columns = T, 
        show_column_dend = T, show_row_dend = F, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize = 8),
        column_names_side = "top", column_names_rot = 45, 
        column_names_gp = gpar(fontsize = 6),
        #row_km = 2, 
        column_km = 2, border = TRUE, 
        column_title = NULL, row_title = NULL, 
        border_gp = gpar(col = "darkgrey"))
#dev.off()

Heatmap(z_neg, name = "z-score", 
        width = ncol(z_neg)*unit(5, "mm"), 
        height = nrow(z_neg)*unit(2.5, "mm"),
        col = heat_colors, 
        cluster_rows= T, cluster_columns = T, 
        show_column_dend = T, show_row_dend = F, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize = 5),
        column_names_side = "top", column_names_rot = 45, 
        column_names_gp = gpar(fontsize = 6),
        #row_km = 2, 
        column_km = 2, border = TRUE, 
        column_title = NULL, row_title = NULL, 
        border_gp = gpar(col = "darkgrey"))

# C2

fgsea_SPZ_vs_iRBC_C2 <- gsea(category = "C2", deg = rbind(SPZ_vs_iRBC[["sig_genes_pvalue_pos"]], SPZ_vs_iRBC[["sig_genes_pvalue_neg"]]))

DT::datatable(fgsea_SPZ_vs_iRBC_C2[["fgseaResTidy"]])

ggplot(fgsea_SPZ_vs_iRBC_C2[["fgseaResTidy"]] %>% filter(pval < 0.008) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= NES)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") +
  theme_minimal()

plotEnrichment(fgsea_SPZ_vs_iRBC_C2[["fgsea_sets"]][["REACTOME_IRF3_MEDIATED_INDUCTION_OF_TYPE_I_IFN"]],
               fgsea_SPZ_vs_iRBC_C2[["gene_logFC_rank"]]) + labs(title="REACTOME_IRF3_MEDIATED_INDUCTION_OF_TYPE_I_IFN")

plotEnrichment(fgsea_SPZ_vs_iRBC_C2[["fgsea_sets"]][["REACTOME_ACTIVATION_OF_IRF3_IRF7_MEDIATED_BY_TBK1_IKK_EPSILON"]],
               fgsea_SPZ_vs_iRBC_C2[["gene_logFC_rank"]]) + labs(title="REACTOME_ACTIVATION_OF_IRF3_IRF7_MEDIATED_BY_TBK1_IKK_EPSILON")

# C5

fgsea_SPZ_vs_iRBC_C5 <- gsea(category = "C5", deg = rbind(SPZ_vs_iRBC[["sig_genes_pvalue_pos"]], SPZ_vs_iRBC[["sig_genes_pvalue_neg"]]))

DT::datatable(fgsea_SPZ_vs_iRBC_C5[["fgseaResTidy"]])

ggplot(fgsea_SPZ_vs_iRBC_C5[["fgseaResTidy"]] %>% filter(pval < 0.008) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= NES)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") +
  theme_minimal()

plotEnrichment(fgsea_SPZ_vs_iRBC_C5[["fgsea_sets"]][["GOBP_RESPONSE_TO_WOUNDING"]],
               fgsea_SPZ_vs_iRBC_C5[["gene_logFC_rank"]]) + labs(title="GOBP_RESPONSE_TO_WOUNDING")

```

## Sporozoite_bys_inf vs control (base)

```{r}

SPZ_vs_untr <- bulkRNAseq_comparison(file = "SPZ_untr_4h.csv",
                                            condition1 = "control",
                                            condition2 = "sporozoite_bys_inf",
                                            #celltype = "Activated monocytes",
                                            padj_thr_bulk = 0.1,
                                            padj_thr_sc = 0.1,
                                            logFC_thr = 1.5)

DT::datatable(SPZ_vs_untr[["bulktable_genes"]])

DT::datatable(SPZ_vs_untr[["sig_genes_pvalue_pos"]])  

DT::datatable(SPZ_vs_untr[["sig_genes_pvalue_neg"]])  


z_pos <- SPZ_vs_untr[["z_pos"]] %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("gene") %>%
  left_join(data.frame(gene = SPZ_vs_untr[["sig_genes_pvalue_pos"]][,"gene"]),.) %>%
  tibble::column_to_rownames("gene")

z_neg <- SPZ_vs_untr[["z_neg"]] %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("gene") %>%
  left_join(data.frame(gene = SPZ_vs_untr[["sig_genes_pvalue_neg"]][,"gene"]),.) %>%
  tibble::column_to_rownames("gene")

Heatmap(z_pos[1:60,], name = "z-score", 
        width = ncol(z_pos)*unit(5, "mm"), 
        height = 60*unit(2.5, "mm"),
        col = heat_colors, 
        cluster_rows= T, cluster_columns = T, 
        show_column_dend = T, show_row_dend = F, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize = 5),
        column_names_side = "top", column_names_rot = 45, 
        column_names_gp = gpar(fontsize = 6),
        #row_km = 2, 
        column_km = 2, border = TRUE, 
        column_title = NULL, row_title = NULL, 
        border_gp = gpar(col = "darkgrey"))

Heatmap(z_neg, name = "z-score", 
        width = ncol(z_neg)*unit(5, "mm"), 
        height = nrow(z_neg)*unit(2.5, "mm"),
        col = heat_colors, 
        cluster_rows= T, cluster_columns = T, 
        show_column_dend = T, show_row_dend = F, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize = 5),
        column_names_side = "top", column_names_rot = 45, 
        column_names_gp = gpar(fontsize = 6),
        #row_km = 2, 
        column_km = 2, border = TRUE, 
        column_title = NULL, row_title = NULL, 
        border_gp = gpar(col = "darkgrey"))

```



## Sporozoite_bys_inf vs iRBC_bys_inf (base)

```{r}

iRBC_vs_SPZ <- bulkRNAseq_comparison(file = "iRBC_SPZ_4h.csv",
                                            condition1 = "sporozoite_bys_inf",
                                            condition2 = "RBC_bys_inf",
                                            #celltype = "Activated monocytes",
                                            padj_thr_bulk = 0.1,
                                            padj_thr_sc = 0.1,
                                            logFC_thr = 1.5)

DT::datatable(SPZ_vs_iRBC[["bulktable_genes"]])

DT::datatable(SPZ_vs_iRBC[["sig_genes_pvalue_pos"]])  

DT::datatable(SPZ_vs_iRBC[["sig_genes_pvalue_neg"]])  


z_pos <- SPZ_vs_iRBC[["z_pos"]] %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("gene") %>%
  left_join(data.frame(gene = SPZ_vs_iRBC[["sig_genes_pvalue_pos"]][,"gene"]),.) %>%
  tibble::column_to_rownames("gene")

z_neg <- SPZ_vs_iRBC[["z_neg"]] %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("gene") %>%
  left_join(data.frame(gene = SPZ_vs_iRBC[["sig_genes_pvalue_neg"]][,"gene"]),.) %>%
  tibble::column_to_rownames("gene")

Heatmap(z_pos, name = "z-score", 
        width = ncol(z_pos)*unit(5, "mm"), 
        height = nrow(z_pos)*unit(2.5, "mm"),
        col = heat_colors, 
        cluster_rows= T, cluster_columns = T, 
        show_column_dend = T, show_row_dend = F, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize = 5),
        column_names_side = "top", column_names_rot = 45, 
        column_names_gp = gpar(fontsize = 6),
        #row_km = 2, 
        column_km = 2, border = TRUE, 
        column_title = NULL, row_title = NULL, 
        border_gp = gpar(col = "darkgrey"))

Heatmap(z_neg, name = "z-score", 
        width = ncol(z_neg)*unit(5, "mm"), 
        height = nrow(z_neg)*unit(2.5, "mm"),
        col = heat_colors, 
        cluster_rows= T, cluster_columns = T, 
        show_column_dend = T, show_row_dend = F, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize = 5),
        column_names_side = "top", column_names_rot = 45, 
        column_names_gp = gpar(fontsize = 6),
        #row_km = 2, 
        column_km = 2, border = TRUE, 
        column_title = NULL, row_title = NULL, 
        border_gp = gpar(col = "darkgrey"))

# res_tbl from de_function() for SPZ_vs_iRBC
res_tbl <- res_tbl %>% column_to_rownames("gene")

# Plot Volcano plots:
# Exchange DEseq2 results object as required:
enhvol = EnhancedVolcano(res_tbl,
                lab = rownames(res_tbl),
                x = 'log2FoldChange',
                y = 'padj',
                pCutoff=0.05,
                FCcutoff=1,
                cutoffLineType="blank",
                col=c('black', 'black', 'black', 'red3'),
                colAlpha=0.6,
                gridlines.major = FALSE,
                gridlines.minor = FALSE,
                #selectLab = c(''),
                #selectLab = c('Tnfa','Il1rn', 'Cxcl2', 'Socs3', 'Hbegf', 'Junb', 'IL10', 'Lilr4b', 'Atf4',
                 #          'Clec4e', 'Procr', 'Cxcl1', 'Bhlhe40', 'Mdm2', 'CD40', 'Nos2'),
                #legendVisible=FALSE,
                border='full',
                #transcriptPointSize=1.5,
                axisLabSize=20,
                xlim=c(-5,5.5),
                #ylim=c(-0.1,3)
)

pdf("SPZ_vs_iRBC_volcano_human.pdf", height = 10)
enhvol
dev.off()

```

## RBC_bys_inf vs RBC_control (base)

```{r}

iRBC_vs_RBC <- bulkRNAseq_comparison(file = "iRBC_RBC_4h.csv",
                                            condition1 = "RBC_control",
                                            condition2 = "RBC_bys_inf",
                                            #celltype = "Activated monocytes",
                                            padj_thr_bulk = 0.1,
                                            padj_thr_sc = 0.1,
                                            logFC_thr = 1.5)

DT::datatable(iRBC_vs_RBC[["bulktable_genes"]])

#DT::datatable(iRBC_vs_RBC[["sig_genes_pvalue_pos"]])  

DT::datatable(iRBC_vs_RBC[["sig_genes_pvalue_neg"]])  


z_pos <- iRBC_vs_RBC[["z_pos"]] %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("gene") %>%
  left_join(data.frame(gene = iRBC_vs_RBC[["sig_genes_pvalue_pos"]][,"gene"]),.) %>%
  tibble::column_to_rownames("gene")

z_neg <- iRBC_vs_RBC[["z_neg"]] %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("gene") %>%
  left_join(data.frame(gene = iRBC_vs_RBC[["sig_genes_pvalue_neg"]][,"gene"]),.) %>%
  tibble::column_to_rownames("gene")

Heatmap(z_pos, name = "z-score", 
        width = ncol(z_pos)*unit(5, "mm"), 
        height = nrow(z_pos)*unit(2.5, "mm"),
        col = heat_colors, 
        cluster_rows= T, cluster_columns = T, 
        show_column_dend = T, show_row_dend = F, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize = 5),
        column_names_side = "top", column_names_rot = 45, 
        column_names_gp = gpar(fontsize = 6),
        #row_km = 2, 
        column_km = 2, border = TRUE, 
        column_title = NULL, row_title = NULL, 
        border_gp = gpar(col = "darkgrey"))

Heatmap(z_neg, name = "z-score", 
        width = ncol(z_neg)*unit(5, "mm"), 
        height = nrow(z_neg)*unit(2.5, "mm"),
        col = heat_colors, 
        cluster_rows= T, cluster_columns = T, 
        show_column_dend = T, show_row_dend = F, 
        show_row_names = T, 
        row_names_gp = gpar(fontsize = 5),
        column_names_side = "top", column_names_rot = 45, 
        column_names_gp = gpar(fontsize = 6),
        #row_km = 2, 
        column_km = 2, border = TRUE, 
        column_title = NULL, row_title = NULL, 
        border_gp = gpar(col = "darkgrey"))

```

```{r}
# enrichment test

#table(In_network = all_host_genes %in% network_genes, Cell_type = all_host_genes %in% ensIDs[[i]]$GENEID)
SPZ_untr_4h <- read.csv("../Data/SPZ_untr_4h.csv")
all_genes <- unique(union(rownames(dds), convertMouseGeneList(SPZ_untr_4h$GeneSymbol)))

# SPZ vs untr
tab <- table(In_human = all_genes %in% SPZ_vs_untr$human_sig_res$gene,
              In_mouse = all_genes %in% convertMouseGeneList(SPZ_vs_untr$bulktable_genes$GeneSymbol))
tab
fisher.test(tab, alternative = "greater")

# iRBC vs SPZ
tab <- table(In_human = all_genes %in% iRBC_vs_SPZ$human_sig_res$gene,
              In_mouse = all_genes %in% convertMouseGeneList(iRBC_vs_SPZ$bulktable_genes$GeneSymbol))
tab
fisher.test(tab, alternative = "greater")

# iRBC vs RBC
tab <- table(In_human = all_genes %in% iRBC_vs_RBC$human_sig_res$gene,
              In_mouse = all_genes %in% convertMouseGeneList(iRBC_vs_RBC$bulktable_genes$GeneSymbol))
tab
fisher.test(tab, alternative = "greater")

```

```{r}
saveRDS(SPZ_vs_untr, "SPZ_vs_untr.rds")
saveRDS(iRBC_vs_SPZ, "iRBC_vs_SPZ.rds")
saveRDS(iRBC_vs_RBC, "iRBC_vs_RBC.rds")

```

```{r}
SPZ_vs_untr <- readRDS("robjects/SPZ_vs_untr.rds")
iRBC_vs_SPZ <- readRDS("robjects/iRBC_vs_SPZ.rds")
iRBC_vs_RBC <- readRDS("robjects/iRBC_vs_RBC.rds")

```


```{r}
# create matrix forheatmap

# 1. genes

# SPZ_vs_untr_log2FC = data.frame(gene = c(SPZ_vs_untr[["sig_genes_pvalue_pos"]][,"gene"], SPZ_vs_untr[["sig_genes_pvalue_neg"]][,"gene"]),
#                          SPZ_vs_untr_log2FC = c(SPZ_vs_untr[["sig_genes_pvalue_pos"]][,"log2FoldChange"], SPZ_vs_untr[["sig_genes_pvalue_neg"]][,"log2FoldChange"]))
# 
# iRBC_vs_SPZ_log2FC = data.frame(gene = c(iRBC_vs_SPZ[["sig_genes_pvalue_pos"]]$gene, iRBC_vs_SPZ[["sig_genes_pvalue_neg"]]$gene),
#                          iRBC_vs_SPZ_log2FC = c(iRBC_vs_SPZ[["sig_genes_pvalue_pos"]]$log2FoldChange, iRBC_vs_SPZ[["sig_genes_pvalue_neg"]]$log2FoldChange))
# 
# iRBC_vs_RBC_log2FC = data.frame(gene = c(iRBC_vs_RBC[["sig_genes_pvalue_pos"]]$gene, iRBC_vs_RBC[["sig_genes_pvalue_neg"]]$gene),
#                          iRBC_vs_RBC_log2FC = c(iRBC_vs_RBC[["sig_genes_pvalue_pos"]]$log2FoldChange, iRBC_vs_RBC[["sig_genes_pvalue_neg"]]$log2FoldChange))
# 
# DEG_all_comparisons <- list(SPZ_vs_untr_log2FC, iRBC_vs_SPZ_log2FC) %>% #iRBC_vs_RBC_log2FC
#   reduce(full_join, by = "gene") %>% 
#   column_to_rownames("gene") %>% 
#   mutate(iRBC_vs_SPZ_log2FC = -1*iRBC_vs_SPZ_log2FC) %>% 
#   mutate(SPZ_vs_iRBC_logFC = iRBC_vs_SPZ_log2FC) %>% 
#   dplyr::select(-iRBC_vs_SPZ_log2FC)
#   
# 

top.n = 20

geneList <- unique(c(SPZ_vs_untr[["sig_genes_pvalue_pos"]]$gene,
              SPZ_vs_untr[["sig_genes_pvalue_neg"]]$gene,
              iRBC_vs_SPZ[["sig_genes_pvalue_pos"]]$gene,
              iRBC_vs_SPZ[["sig_genes_pvalue_neg"]]$gene
              ))

SPZ_vs_untr_log2FC = SPZ_vs_untr$human_sig_res %>% 
  dplyr::select(gene, log2FoldChange) %>% 
  filter(gene %in% geneList) %>% 
  mutate(SPZ_vs_untr = log2FoldChange) %>% 
  dplyr::select(-log2FoldChange) %>% 
  arrange(-abs(SPZ_vs_untr)) 

SPZ_vs_untr_log2FC_top10 <- SPZ_vs_untr_log2FC %>% 
  top_n(n = top.n)

SPZ_vs_iRBC_log2FC = iRBC_vs_SPZ$human_sig_res %>% 
  dplyr::select(gene, log2FoldChange) %>% 
  filter(gene %in% geneList) %>% 
  mutate(SPZ_vs_iRBC = -1*log2FoldChange) %>% 
  dplyr::select(-log2FoldChange) %>% 
  arrange(-abs(SPZ_vs_iRBC)) 

SPZ_vs_iRBC_log2FC_top10 <- SPZ_vs_iRBC_log2FC %>% 
  top_n(n = top.n)

iRBC_vs_RBC_log2FC = iRBC_vs_RBC$human_sig_res %>% 
  dplyr::select(gene, log2FoldChange) %>% 
  filter(gene %in% geneList) %>% 
  mutate(iRBC_vs_RBC = log2FoldChange) %>% 
  dplyr::select(-log2FoldChange)%>% 
  arrange(-abs(iRBC_vs_RBC)) 

iRBC_vs_RBC_log2FC_top10 <- iRBC_vs_RBC_log2FC %>% 
  top_n(n = top.n)

DEG_all_comparisons <- list(SPZ_vs_untr_log2FC_top10, SPZ_vs_iRBC_log2FC_top10, iRBC_vs_RBC_log2FC_top10) %>%
  purrr::reduce(full_join, by = "gene") %>% 
  column_to_rownames("gene")

# top 10 log2FC genes and their log2FC from all conditions

SPZ_vs_untr_allcontrasts <- DEG_all_comparisons %>% 
  rownames_to_column("gene") %>% 
  dplyr::select(gene) %>% 
  left_join(., SPZ_vs_untr_log2FC)

SPZ_vs_iRBC_allcontrasts <- DEG_all_comparisons %>% 
  rownames_to_column("gene") %>% 
  dplyr::select(gene) %>% 
  left_join(., SPZ_vs_iRBC_log2FC)

iRBC_vs_RBC_allcontrasts <- DEG_all_comparisons %>% 
  rownames_to_column("gene") %>% 
  dplyr::select(gene) %>% 
  left_join(., iRBC_vs_RBC_log2FC)

DEG_all_comparisons <- list(SPZ_vs_untr_allcontrasts, SPZ_vs_iRBC_allcontrasts, iRBC_vs_RBC_allcontrasts) %>%
  purrr::reduce(full_join, by = "gene") %>% 
  column_to_rownames("gene") %>% 
  arrange(-SPZ_vs_untr)

#DEG_all_comparisons[is.na(DEG_all_comparisons)] <- 0

# genes_to_show <- c(iRBC_vs_SPZ[["sig_genes_pvalue_pos"]]$gene,
#               iRBC_vs_SPZ[["sig_genes_pvalue_neg"]]$gene)
# mark_at = which(rownames(DEG_all_comparisons) %in% genes_to_show)
# genes_on_heatmap <- rownames(DEG_all_comparisons[mark_at,])
# ha = rowAnnotation(gene = anno_mark(at = mark_at, labels = genes_on_heatmap))

p <- Heatmap(DEG_all_comparisons, 
        #right_annotation = ha,
        #col = viridis_pal(option = "B")(8),
        #col = colorRampPalette(rev(brewer.pal(n = 5, name ="Spectral")))(50),
        col = colorRamp2(c(10, 0, -10), c("blue", "white", "red")),
        rect_gp = gpar(col = "white", lwd = 0.5),
        width = ncol(DEG_all_comparisons)*unit(7, "mm"),
        height = nrow(DEG_all_comparisons)*unit(3, "mm"),
        name = "log2FC",
        na_col = "gray",
        heatmap_legend_param = list(legend_gp = gpar(fontsize = 5)),
        cluster_rows = F, 
        cluster_columns = F,
        show_row_dend = F, 
        show_row_names = T,
        column_title = "Diff. expressed genes",
        column_title_gp = gpar(fontsize = 12),
        column_names_gp = gpar(fontsize = 8),
        column_names_rot = 45,
        row_names_gp = gpar(fontsize = 5))

pdf("Heatmap_genes_on_overlap_mouse_human_top20inAllContrasts.pdf")
draw(p)
dev.off()

```

```{r}
SPZ_vs_untr_human <- de_function(condition1 = "control",
                                 condition2 = "sporozoite_bys_inf",
                                 padj_thr_sc = 0.1,
                                 logFC_thr = 1.5)

SPZ_vs_iRBC_human <- de_function(condition1 = "RBC_bys_inf",
                                 condition2 = "sporozoite_bys_inf",
                                 padj_thr_sc = 0.1,
                                 logFC_thr = 1.5)

iRBC_vs_RBC_human <- de_function(condition1 = "RBC_control",
                                 condition2 = "RBC_bys_inf",
                                 padj_thr_sc = 0.1,
                                 logFC_thr = 1.5)
# without thresholds
SPZ_vs_untr_human_nothr <- de_function_nothr(condition1 = "control",
                                 condition2 = "sporozoite_bys_inf")

SPZ_vs_iRBC_human_nothr <- de_function_nothr(condition1 = "RBC_bys_inf",
                                 condition2 = "sporozoite_bys_inf")

iRBC_vs_RBC_human_nothr <- de_function_nothr(condition1 = "RBC_control",
                                 condition2 = "RBC_bys_inf")



SPZ_vs_untr_log2FC = SPZ_vs_untr_human_nothr %>% 
  dplyr::select(gene, log2FoldChange) %>% 
  filter(gene %in% geneList) %>% 
  mutate(SPZ_vs_untr = log2FoldChange) %>% 
  dplyr::select(-log2FoldChange)

SPZ_vs_iRBC_log2FC = SPZ_vs_iRBC_human_nothr %>% 
  dplyr::select(gene, log2FoldChange) %>% 
  filter(gene %in% geneList) %>% 
  mutate(SPZ_vs_iRBC = log2FoldChange) %>% 
  dplyr::select(-log2FoldChange)

iRBC_vs_RBC_log2FC = iRBC_vs_RBC_human_nothr %>% 
  dplyr::select(gene, log2FoldChange) %>% 
  filter(gene %in% geneList) %>% 
  mutate(iRBC_vs_RBC = log2FoldChange) %>% 
  dplyr::select(-log2FoldChange)

DEG_all_comparisons <- list(SPZ_vs_untr_log2FC, SPZ_vs_iRBC_log2FC, iRBC_vs_RBC_log2FC) %>%
  purrr::reduce(inner_join, by = "gene") %>% 
  column_to_rownames("gene")

SPZ_vs_untr_pval <- SPZ_vs_untr_human_nothr %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>% 
  dplyr::select(gene, padj) %>% 
  dplyr::rename(SPZ_vs_untr = "padj")

SPZ_vs_iRBC_pval <- SPZ_vs_iRBC_human_nothr %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>% 
  dplyr::select(gene, padj) %>% 
  dplyr::rename(SPZ_vs_iRBC = "padj")

iRBC_vs_RBC_pval <- iRBC_vs_RBC_human_nothr %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>% 
  dplyr::select(gene, padj) %>% 
  dplyr::rename(iRBC_vs_RBC = "padj")

PVAL_all_comparisons <- list(SPZ_vs_untr_pval, SPZ_vs_iRBC_pval,  iRBC_vs_RBC_pval) %>% 
  purrr::reduce(inner_join, by = "gene") %>% 
  filter(gene %in% rownames(DEG_all_comparisons)) %>% 
  column_to_rownames("gene")

pval_hclust <- hclust(dist(PVAL_all_comparisons))
gene_dend <- as.dendrogram(pval_hclust)
plot(gene_dend)
#gene_dend = dendextend::rotate(gene_dend,1:50)

cluster_rows = gene_dend

# genes_to_show <- c(iRBC_vs_SPZ[["sig_genes_pvalue_pos"]]$gene,
#               iRBC_vs_SPZ[["sig_genes_pvalue_neg"]]$gene)

genes_to_show = c("CDC42EP4", "GADD45A", "NR4A2", "IRF4", "TNFRSF10D", 
             "FRMD6", "OTUD7B", "PPARGC1B", "KLF4", "DUSP1", "IL10",
             "SLCO2B1", "INHBA", "PTGS2", "GEM")

mark_at = which(rownames(DEG_all_comparisons) %in% genes_to_show)
genes_on_heatmap <- rownames(DEG_all_comparisons[mark_at,])
ha = rowAnnotation(gene = anno_mark(at = mark_at,
                                    labels = genes_on_heatmap,
                                    lines_gp = gpar(size = 2),
                                    labels_gp = gpar(fontsize = 5)))
p <- Heatmap(DEG_all_comparisons, 
        right_annotation = ha,
        #col = viridis_pal(option = "B")(8),
        #col = colorRampPalette(rev(brewer.pal(n = 5, name ="Spectral")))(50),
        col = colorRamp2(c(10, 0, -10), c("red", "white", "blue")),
 cell_fun = function(j, i, x, y, w, h, fill) {
#     if(PVAL_all_comparisons[i, j] <= 0.0001) {
#         grid.text("...", x, y)
#     } else if(PVAL_all_comparisons[i, j] <= 0.001) {
#         grid.text("..", x, y)
#     }  else 
if(PVAL_all_comparisons[i, j] <= 0.01) {
        grid.text("-", x, y)

}},
        rect_gp = gpar(col = "white", lwd = 0.5),
        width = ncol(DEG_all_comparisons)*unit(7, "mm"),
        #height = nrow(DEG_all_comparisons)*unit(0.3, "mm"),
        name = "log2FC",
        na_col = "gray",
        heatmap_legend_param = list(legend_gp = gpar(fontsize = 5)),
        #cluster_rows = cluster_rows, 
        cluster_rows = T,
        cluster_columns = F,
        show_row_dend = F, 
        show_row_names = F,
        column_title = "Diff. expressed genes",
        column_title_gp = gpar(fontsize = 12),
        column_names_gp = gpar(fontsize = 8),
        column_names_rot = 45)

svg("Heatmap_genes_on_overlap_mouse_human.svg")
draw(p)
dev.off()

write.csv(SPZ_vs_iRBC_human_nothr, "SPZ_vs_iRBC_human_nothr.csv", quote = F, row.names = F)
write.csv(SPZ_vs_untr_human_nothr, "SPZ_vs_untr_human_nothr.csv", quote = F, row.names = F)
write.csv(iRBC_vs_RBC_human_nothr, "iRBC_vs_RBC_human_nothr.csv", quote = F, row.names = F)
```

```{r gseGO}

SPZ_vs_untr_human_nothr_geneList <- SPZ_vs_untr_human_nothr$log2FoldChange
names(SPZ_vs_untr_human_nothr_geneList) <- SPZ_vs_untr_human_nothr$gene

SPZ_vs_untr_human_nothr_geneList_sorted <- sort(SPZ_vs_untr_human_nothr_geneList, decreasing = T)
names(SPZ_vs_untr_human_nothr_geneList_sorted) <- mapIds(org.Hs.eg.db, names(SPZ_vs_untr_human_nothr_geneList_sorted), 'ENTREZID', 'SYMBOL')

SPZ_vs_untr_human_nothr_ego <- gseGO(geneList     = SPZ_vs_untr_human_nothr_geneList_sorted,
              OrgDb        = org.Hs.eg.db,
              eps = 0,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE)

write.csv2(SPZ_vs_untr_human_nothr_ego@result, "SPZ_vs_untr_human_nothr_gseGO.csv", row.names = F, quote = F)

# enrichGO
SPZ_vs_untr_human_nothr_geneList <- SPZ_vs_untr_human_nothr %>% 
  filter(padj <= 0.1 & abs(log2FoldChange) >= 1.5) %>% 
  dplyr::select(gene, log2FoldChange)
SPZ_vs_untr_human_nothr_geneList_FC <- SPZ_vs_untr_human_nothr_geneList$log2FoldChange
names(SPZ_vs_untr_human_nothr_geneList_FC) <- SPZ_vs_untr_human_nothr_geneList$gene

SPZ_vs_untr_human_nothr_geneList_sorted <- sort(SPZ_vs_untr_human_nothr_geneList_FC, decreasing = T)
names(SPZ_vs_untr_human_nothr_geneList_sorted) <- mapIds(org.Hs.eg.db, names(SPZ_vs_untr_human_nothr_geneList_sorted), 'ENTREZID', 'SYMBOL')

SPZ_vs_untr_human_nothr_ego1 <- enrichGO(gene     = names(SPZ_vs_untr_human_nothr_geneList_sorted),
                                         universe = mapIds(org.Hs.eg.db,SPZ_vs_iRBC_human_nothr$gene, 'ENTREZID', 'SYMBOL'),
              OrgDb        = org.Hs.eg.db,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.01,
              #keyType = "SYMBOL",
              qvalueCutoff = 0.05,
              readable      = TRUE)
##################
SPZ_vs_iRBC_human_nothr_geneList <- SPZ_vs_iRBC_human_nothr$log2FoldChange
names(SPZ_vs_iRBC_human_nothr_geneList) <- SPZ_vs_iRBC_human_nothr$gene

SPZ_vs_iRBC_human_nothr_geneList_sorted <- sort(SPZ_vs_iRBC_human_nothr_geneList, decreasing = T)
names(SPZ_vs_iRBC_human_nothr_geneList_sorted) <- mapIds(org.Hs.eg.db, names(SPZ_vs_iRBC_human_nothr_geneList_sorted), 'ENTREZID', 'SYMBOL')

SPZ_vs_iRBC_human_nothr_ego <- gseGO(geneList     = SPZ_vs_iRBC_human_nothr_geneList_sorted,
              OrgDb        = org.Hs.eg.db,
              eps = 0,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE)

write.csv2(SPZ_vs_iRBC_human_nothr_ego@result, "SPZ_vs_iRBC_human_nothr_gseGO.csv", row.names = F, quote = F)

# enrichGO
SPZ_vs_iRBC_human_nothr_geneList <- SPZ_vs_iRBC_human_nothr %>% 
  filter(padj <= 0.1 & abs(log2FoldChange) >= 1.5) %>% 
  dplyr::select(gene, log2FoldChange)
SPZ_vs_iRBC_human_nothr_geneList_FC <- SPZ_vs_iRBC_human_nothr_geneList$log2FoldChange
names(SPZ_vs_iRBC_human_nothr_geneList_FC) <- SPZ_vs_iRBC_human_nothr_geneList$gene

SPZ_vs_iRBC_human_nothr_geneList_sorted <- sort(SPZ_vs_iRBC_human_nothr_geneList_FC, decreasing = T)
# names(SPZ_vs_iRBC_human_nothr_geneList_sorted) <- mapIds(org.Hs.eg.db, names(SPZ_vs_iRBC_human_nothr_geneList_sorted), 'ENTREZID', 'SYMBOL')

SPZ_vs_iRBC_human_nothr_ego1 <- enrichGO(gene     = names(SPZ_vs_iRBC_human_nothr_geneList_sorted),
                                         universe = SPZ_vs_iRBC_human_nothr$gene,
              OrgDb        = org.Hs.eg.db,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.01,
              keyType = "SYMBOL",
              qvalueCutoff = 0.05,
              readable      = TRUE)
#######################

iRBC_vs_RBC_human_nothr_geneList <- iRBC_vs_RBC_human_nothr$log2FoldChange
names(iRBC_vs_RBC_human_nothr_geneList) <- iRBC_vs_RBC_human_nothr$gene

iRBC_vs_RBC_human_nothr_geneList_sorted <- sort(iRBC_vs_RBC_human_nothr_geneList, decreasing = T)
names(iRBC_vs_RBC_human_nothr_geneList_sorted) <- mapIds(org.Hs.eg.db, names(iRBC_vs_RBC_human_nothr_geneList_sorted), 'ENTREZID', 'SYMBOL')

iRBC_vs_RBC_human_nothr_ego <- gseGO(geneList     = iRBC_vs_RBC_human_nothr_geneList_sorted,
              OrgDb        = org.Hs.eg.db,
              eps = 0,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE)

write.csv2(iRBC_vs_RBC_human_nothr_ego@result, "iRBC_vs_RBC_human_nothr_gseGO.csv", row.names = F, quote = F)

# enrichGO
iRBC_vs_RBC_human_nothr_geneList <- iRBC_vs_RBC_human_nothr %>% 
  filter(padj <= 0.1 & abs(log2FoldChange) >= 1.5) %>% 
  dplyr::select(gene, log2FoldChange)
iRBC_vs_RBC_human_nothr_geneList_FC <- iRBC_vs_RBC_human_nothr_geneList$log2FoldChange
names(iRBC_vs_RBC_human_nothr_geneList_FC) <- iRBC_vs_RBC_human_nothr_geneList$gene

iRBC_vs_RBC_human_nothr_geneList_sorted <- sort(iRBC_vs_RBC_human_nothr_geneList_FC, decreasing = T)
# names(iRBC_vs_RBC_human_nothr_geneList_sorted) <- mapIds(org.Hs.eg.db, names(iRBC_vs_RBC_human_nothr_geneList_sorted), 'ENTREZID', 'SYMBOL')

iRBC_vs_RBC_human_nothr_ego1 <- enrichGO(gene     = names(iRBC_vs_RBC_human_nothr_geneList_sorted),
                                         universe = iRBC_vs_RBC_human_nothr$gene,
              OrgDb        = org.Hs.eg.db,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.01,
              keyType = "SYMBOL",
              qvalueCutoff = 0.05,
              readable      = TRUE)
#######################

saveRDS(iRBC_vs_RBC_human_nothr_ego, "iRBC_vs_RBC_human_nothr_ego.rds")
saveRDS(SPZ_vs_iRBC_human_nothr_ego, "SPZ_vs_iRBC_human_nothr_ego.rds")
saveRDS(SPZ_vs_untr_human_nothr_ego, "SPZ_vs_untr_human_nothr_ego.rds")

# mouse
SPZ_vs_untr_mouse_nothr <- read.csv("../Data/SPZ_untr_4h.csv")
SPZ_vs_untr_mouse_nothr_geneList <- SPZ_vs_untr_mouse_nothr$log2FoldChange
names(SPZ_vs_untr_mouse_nothr_geneList) <- SPZ_vs_untr_mouse_nothr$GeneSymbol

SPZ_vs_untr_mouse_nothr_geneList_sorted <- sort(SPZ_vs_untr_mouse_nothr_geneList, decreasing = T)
names(SPZ_vs_untr_mouse_nothr_geneList_sorted) <- mapIds(org.Mm.eg.db, names(SPZ_vs_untr_mouse_nothr_geneList_sorted), 'ENTREZID', 'SYMBOL')

SPZ_vs_untr_mouse_nothr_ego <- gseGO(geneList     = SPZ_vs_untr_mouse_nothr_geneList_sorted,
              OrgDb        = org.Mm.eg.db,
              eps = 0,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE)

write.csv2(SPZ_vs_untr_mouse_nothr_ego@result, "SPZ_vs_untr_mouse_nothr_gseGO.csv", row.names = F, quote = F)

# enrichGO
SPZ_vs_untr_mouse_nothr_geneList <- SPZ_vs_untr_mouse_nothr %>% 
  filter(padj <= 0.1 & abs(log2FoldChange) >= 1.5) %>% 
  dplyr::select(GeneSymbol, log2FoldChange)
SPZ_vs_untr_mouse_nothr_geneList_FC <- SPZ_vs_untr_mouse_nothr_geneList$log2FoldChange
names(SPZ_vs_untr_mouse_nothr_geneList_FC) <- SPZ_vs_untr_mouse_nothr_geneList$GeneSymbol

SPZ_vs_untr_mouse_nothr_geneList_sorted <- sort(SPZ_vs_untr_mouse_nothr_geneList_FC, decreasing = T)
# names(SPZ_vs_untr_mouse_nothr_geneList_sorted) <- mapIds(org.Hs.eg.db, names(SPZ_vs_untr_mouse_nothr_geneList_sorted), 'ENTREZID', 'SYMBOL')

SPZ_vs_untr_mouse_nothr_ego1 <- enrichGO(gene     = names(SPZ_vs_untr_mouse_nothr_geneList_sorted),
                                         universe = SPZ_vs_untr_mouse_nothr$GeneSymbol,
              OrgDb        = org.Mm.eg.db,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.01,
              keyType = "SYMBOL",
              qvalueCutoff = 0.05,
              readable      = TRUE)
#######################

SPZ_vs_iRBC_mouse_nothr <- read.csv("../Data/iRBC_SPZ_4h.csv")
SPZ_vs_iRBC_mouse_nothr_geneList <- -1*SPZ_vs_iRBC_mouse_nothr$log2FoldChange
names(SPZ_vs_iRBC_mouse_nothr_geneList) <- SPZ_vs_iRBC_mouse_nothr$GeneSymbol

SPZ_vs_iRBC_mouse_nothr_geneList_sorted <- sort(SPZ_vs_iRBC_mouse_nothr_geneList, decreasing = T)
names(SPZ_vs_iRBC_mouse_nothr_geneList_sorted) <- mapIds(org.Mm.eg.db, names(SPZ_vs_iRBC_mouse_nothr_geneList_sorted), 'ENTREZID', 'SYMBOL')

SPZ_vs_iRBC_mouse_nothr_ego <- gseGO(geneList     = SPZ_vs_iRBC_mouse_nothr_geneList_sorted,
              OrgDb        = org.Mm.eg.db,
              eps = 0,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE)

write.csv2(SPZ_vs_iRBC_mouse_nothr_ego@result, "SPZ_vs_iRBC_mouse_nothr_gseGO.csv", row.names = F, quote = F)

# enrichGO
SPZ_vs_iRBC_mouse_nothr_geneList <- SPZ_vs_iRBC_mouse_nothr %>% 
  filter(padj <= 0.1 & abs(log2FoldChange) >= 1.5) %>% 
  dplyr::select(GeneSymbol, log2FoldChange)
SPZ_vs_iRBC_mouse_nothr_geneList_FC <- SPZ_vs_iRBC_mouse_nothr_geneList$log2FoldChange
names(SPZ_vs_iRBC_mouse_nothr_geneList_FC) <- SPZ_vs_iRBC_mouse_nothr_geneList$GeneSymbol

SPZ_vs_iRBC_mouse_nothr_geneList_sorted <- sort(SPZ_vs_iRBC_mouse_nothr_geneList_FC, decreasing = T)
# names(SPZ_vs_iRBC_mouse_nothr_geneList_sorted) <- mapIds(org.Hs.eg.db, names(SPZ_vs_iRBC_mouse_nothr_geneList_sorted), 'ENTREZID', 'SYMBOL')

SPZ_vs_iRBC_mouse_nothr_ego1 <- enrichGO(gene     = names(SPZ_vs_iRBC_mouse_nothr_geneList_sorted),
                                         universe = SPZ_vs_iRBC_mouse_nothr$GeneSymbol,
              OrgDb        = org.Mm.eg.db,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.01,
              keyType = "SYMBOL",
              qvalueCutoff = 0.05,
              readable      = TRUE)
#######################

iRBC_vs_RBC_mouse_nothr <- read.csv("../Data/iRBC_RBC_4h.csv")
iRBC_vs_RBC_mouse_nothr_geneList <- iRBC_vs_RBC_mouse_nothr$log2FoldChange
names(iRBC_vs_RBC_mouse_nothr_geneList) <- iRBC_vs_RBC_mouse_nothr$GeneSymbol

iRBC_vs_RBC_mouse_nothr_geneList_sorted <- sort(iRBC_vs_RBC_mouse_nothr_geneList, decreasing = T)
names(iRBC_vs_RBC_mouse_nothr_geneList_sorted) <- mapIds(org.Mm.eg.db, names(iRBC_vs_RBC_mouse_nothr_geneList_sorted), 'ENTREZID', 'SYMBOL')

iRBC_vs_RBC_mouse_nothr_ego <- gseGO(geneList     = iRBC_vs_RBC_mouse_nothr_geneList_sorted,
              OrgDb        = org.Mm.eg.db,
              eps = 0,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.05,
              verbose      = FALSE)

write.csv2(iRBC_vs_RBC_mouse_nothr_ego@result, "iRBC_vs_RBC_mouse_nothr_gseGO.csv", row.names = F, quote = F)

# enrichGO
iRBC_vs_RBC_mouse_nothr_geneList <- iRBC_vs_RBC_mouse_nothr %>% 
  filter(padj <= 0.1 & abs(log2FoldChange) >= 0.5) %>% 
  dplyr::select(GeneSymbol, log2FoldChange)
iRBC_vs_RBC_mouse_nothr_geneList_FC <- iRBC_vs_RBC_mouse_nothr_geneList$log2FoldChange
names(iRBC_vs_RBC_mouse_nothr_geneList_FC) <- iRBC_vs_RBC_mouse_nothr_geneList$GeneSymbol

iRBC_vs_RBC_mouse_nothr_geneList_sorted <- sort(iRBC_vs_RBC_mouse_nothr_geneList_FC, decreasing = T)
# names(iRBC_vs_RBC_mouse_nothr_geneList_sorted) <- mapIds(org.Hs.eg.db, names(iRBC_vs_RBC_mouse_nothr_geneList_sorted), 'ENTREZID', 'SYMBOL')

iRBC_vs_RBC_mouse_nothr_ego1 <- enrichGO(gene     = names(iRBC_vs_RBC_mouse_nothr_geneList_sorted),
                                         universe = iRBC_vs_RBC_mouse_nothr$GeneSymbol,
              OrgDb        = org.Mm.eg.db,
              ont          = "BP",
              minGSSize    = 100,
              maxGSSize    = 500,
              pvalueCutoff = 0.01,
              keyType = "SYMBOL",
              qvalueCutoff = 0.05,
              readable      = TRUE)
#######################

saveRDS(iRBC_vs_RBC_mouse_nothr_ego, "iRBC_vs_RBC_mouse_nothr_ego.rds")
saveRDS(SPZ_vs_iRBC_mouse_nothr_ego, "SPZ_vs_iRBC_mouse_nothr_ego.rds")
saveRDS(SPZ_vs_untr_mouse_nothr_ego, "SPZ_vs_untr_mouse_nothr_ego.rds")
## common ones between human and mouse

## readRDS
iRBC_vs_RBC_mouse_nothr_ego <- readRDS("robjects/iRBC_vs_RBC_mouse_nothr_ego.rds")
SPZ_vs_iRBC_mouse_nothr_ego <- readRDS("robjects/SPZ_vs_iRBC_mouse_nothr_ego.rds")
SPZ_vs_untr_mouse_nothr_ego <- readRDS("robjects/SPZ_vs_untr_mouse_nothr_ego.rds")

iRBC_vs_RBC_human_nothr_ego <- readRDS("robjects/iRBC_vs_RBC_human_nothr_ego.rds")
SPZ_vs_iRBC_human_nothr_ego <- readRDS("robjects/SPZ_vs_iRBC_human_nothr_ego.rds")
SPZ_vs_untr_human_nothr_ego <- readRDS("robjects/SPZ_vs_untr_human_nothr_ego.rds")
############

SPZ_vs_untr_common <- SPZ_vs_untr_human_nothr_ego@result[,c("ID", "Description", "NES", "p.adjust")] %>% 
  inner_join(., SPZ_vs_untr_mouse_nothr_ego@result[,c("ID", "Description", "NES", "p.adjust")], by = c("ID", "Description")) %>% 
  dplyr::rename(NES_human = NES.x, padj_human = p.adjust.x,
                NES_mouse = NES.y, padj_mouse = p.adjust.y)

SPZ_vs_iRBC_common <- SPZ_vs_iRBC_human_nothr_ego@result[,c("ID", "Description", "NES", "p.adjust")] %>% 
  inner_join(., SPZ_vs_iRBC_mouse_nothr_ego@result[,c("ID", "Description", "NES", "p.adjust")], by = c("ID", "Description")) %>% 
  dplyr::rename(NES_human = NES.x, padj_human = p.adjust.x,
                NES_mouse = NES.y, padj_mouse = p.adjust.y)

iRBC_vs_RBC_common <- iRBC_vs_RBC_human_nothr_ego@result[,c("ID", "Description", "NES", "p.adjust")] %>% 
  inner_join(., iRBC_vs_RBC_mouse_nothr_ego@result[,c("ID", "Description", "NES", "p.adjust")], by = c("ID", "Description")) %>% 
  dplyr::rename(NES_human = NES.x, padj_human = p.adjust.x,
                NES_mouse = NES.y, padj_mouse = p.adjust.y)

write.csv2(SPZ_vs_untr_common, "SPZ_vs_untr_common_gseGOBP.csv", row.names = F, quote = F)
write.csv2(SPZ_vs_iRBC_common, "SPZ_vs_iRBC_common_gseGOBP.csv", row.names = F, quote = F)
write.csv2(iRBC_vs_RBC_common, "iRBC_vs_RBC_common_gseGOBP.csv", row.names = F, quote = F)

# enrichGO
SPZ_vs_untr_common <- SPZ_vs_untr_human_nothr_ego1@result[,c("Description", "p.adjust")] %>% 
  inner_join(., SPZ_vs_untr_mouse_nothr_ego1@result[,c("Description", "p.adjust")], by = c("Description"))

SPZ_vs_iRBC_common <- SPZ_vs_iRBC_human_nothr_ego1@result[,c("Description", "p.adjust")] %>% 
  inner_join(., SPZ_vs_iRBC_mouse_nothr_ego1@result[,c("Description", "p.adjust")], by = c("Description"))

iRBC_vs_RBC_common <- iRBC_vs_RBC_human_nothr_ego1@result[,c("Description", "p.adjust")] %>% 
  inner_join(., iRBC_vs_RBC_mouse_nothr_ego1@result[,c("Description", "p.adjust")], by = c("Description"))

#hclust based on pvalue in human

SPZ_vs_untr_common_pvalue <- SPZ_vs_untr_common %>% 
  dplyr::select(Description, padj_human) %>% 
  #column_to_rownames("Description") %>% 
  dplyr::rename(SPZ_vs_untr = padj_human)

SPZ_vs_iRBC_common_pvalue <- SPZ_vs_iRBC_common %>% 
  dplyr::select(Description, padj_human) %>% 
  #column_to_rownames("Description") %>% 
  dplyr::rename(SPZ_vs_iRBC = padj_human)

iRBC_vs_RBC_common_pvalue <- iRBC_vs_RBC_common %>% 
  dplyr::select(Description, padj_human) %>% 
  #column_to_rownames("Description") %>% 
  dplyr::rename(iRBC_vs_RBC = padj_human)

PVAL_gseGO_comparisons <- list(SPZ_vs_untr_common_pvalue, SPZ_vs_iRBC_common_pvalue,  iRBC_vs_RBC_common_pvalue) %>%
  purrr::reduce(left_join, by = "Description") %>%
  column_to_rownames("Description") %>% 
  filter(SPZ_vs_untr <= 0.001 & SPZ_vs_iRBC <= 0.001 ) %>% 
  arrange(SPZ_vs_iRBC) %>% 
  filter(!is.na(SPZ_vs_iRBC))

#PVAL_gseGO_comparisons[is.na(PVAL_gseGO_comparisons)] <- 1

pval_hclust <- hclust(dist(PVAL_gseGO_comparisons))
gene_dend <- as.dendrogram(pval_hclust)
plot(gene_dend)
#gene_dend = dendextend::rotate(gene_dend,1:50)

cluster_rows = gene_dend

SPZ_vs_untr_common_NES <- SPZ_vs_untr_common %>% 
  dplyr::select(Description, NES_human) %>% 
  #column_to_rownames("Description") %>% 
  dplyr::rename(SPZ_vs_untr = NES_human)

SPZ_vs_iRBC_common_NES <- SPZ_vs_iRBC_common %>% 
  dplyr::select(Description, NES_human) %>% 
  #column_to_rownames("Description") %>% 
  dplyr::rename(SPZ_vs_iRBC = NES_human)

iRBC_vs_RBC_common_NES <- iRBC_vs_RBC_common %>% 
  dplyr::select(Description, NES_human) %>% 
  #column_to_rownames("Description") %>% 
  dplyr::rename(iRBC_vs_RBC = NES_human)

NES_gseGO_comparisons <- list(SPZ_vs_untr_common_NES, SPZ_vs_iRBC_common_NES,  iRBC_vs_RBC_common_NES) %>%
  purrr::reduce(full_join, by = "Description") %>%
  filter(Description %in% rownames(PVAL_gseGO_comparisons)) %>% 
 column_to_rownames("Description") %>% 
  arrange(-SPZ_vs_untr) %>% 
  filter(SPZ_vs_untr >= 1.77)

p <- Heatmap(-log10(PVAL_gseGO_comparisons), 
        #right_annotation = ha,
        #col = viridis_pal(option = "B")(8),
        #col = colorRampPalette(rev(brewer.pal(n = 5, name ="Spectral")))(50),
        col = colorRamp2(c(15, 0), c("blue", "red")),
# cell_fun = function(j, i, x, y, w, h, fill) {
#     # if(PVAL_gseGO_comparisons[i, j] <= 0.0001) {
#     #     grid.text("...", x, y)
#     # } else if(PVAL_gseGO_comparisons[i, j] <= 0.001) {
#     #     grid.text("..", x, y)
#     # }  else 
#   if(PVAL_gseGO_comparisons[i, j] <= 0.01) {
#         grid.text("-", x, y)
# 
# }},
        rect_gp = gpar(col = "white", lwd = 0.5),
        width = ncol(PVAL_gseGO_comparisons)*unit(7, "mm"),
        height = nrow(PVAL_gseGO_comparisons)*unit(1.5, "mm"),
        name = "-log10(padj)",
        na_col = "gray",
        heatmap_legend_param = list(legend_gp = gpar(fontsize = 5)),
        #cluster_rows = cluster_rows, 
        cluster_rows = F,
        cluster_columns = F,
        show_row_dend = F, 
        show_row_names = T,
        column_title = "GSEA GO:BP",
        column_title_gp = gpar(fontsize = 12),
        column_names_gp = gpar(fontsize = 8),
        column_names_rot = 45,
row_names_gp = gpar(fontsize = 5))

pdf("Heatmap_gseaGOBP_overlap_mouse_human_onlySPZ_vs_iRBC.pdf", height = 12)
draw(p)
dev.off()

p <- Heatmap(NES_gseGO_comparisons, 
        col = colorRamp2(c(max(NES_gseGO_comparisons, na.rm = T), min(NES_gseGO_comparisons, na.rm = T)), c("red", "white")),
        rect_gp = gpar(col = "white", lwd = 0.5),
        width = ncol(NES_gseGO_comparisons)*unit(7, "mm"),
        height = nrow(NES_gseGO_comparisons)*unit(1.5, "mm"),
        name = "NES",
        na_col = "gray",
        heatmap_legend_param = list(legend_gp = gpar(fontsize = 5)),
        #cluster_rows = cluster_rows, 
        cluster_rows = F,
        cluster_columns = F,
        show_row_dend = F, 
        show_row_names = T,
        column_title = "GSEA GO:BP",
        column_title_gp = gpar(fontsize = 12),
        column_names_gp = gpar(fontsize = 8),
        column_names_rot = 45,
row_names_gp = gpar(fontsize = 5))

pdf("Heatmap_gseaGOBP_overlap_mouse_human_NES_onlySPZ_vs_iRBC.pdf", height = 12)
draw(p)
dev.off()

svg("Heatmap_gseaGOBP_overlap_mouse_human_NES_onlySPZ_vs_iRBC.svg", height = 12)
draw(p)
dev.off()
```