---
title: "scRNA-seq of Plasmodium sporozoite-infected monocytes - GSEA"
output:
  html_document:
    df_print: paged
    code_folding: hide
  pdf_document: default
  toc: TRUE
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
knitr::opts_chunk$set(fig.width = 10, fig.height = 8) 

```

```{r}
library(Seurat)
library(patchwork)
library(readxl)
library(tidyverse)
library(ggplot2)
library(ggpubr)
library(reshape2)
library(broom)
library(stringr)
library(ggplot2)
library(SeuratData)
#library(celldex)
#install.packages("cowplot")
library(cowplot)
#devtools::install_github('cole-trapnell-lab/monocle3')
library(monocle3)
#remotes::install_github('satijalab/seurat-wrappers')
library(SeuratWrappers)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(ggrepel)
library(data.table)
library(msigdbr)
#library(devtools)
#install_github("arc85/singleseqgset")
#library(singleseqgset)
library(heatmap3)
#install_github('immunogenomics/presto')
library(presto)
library(fgsea)
# mulattoes::install('multtest')
# install.packages('metap')
#library(metap)
#remotes::install_github("satijalab/seurat-data")
```

```{r}
immune.combined <- readRDS("immune.combined_all_libs_preprocessed.rds")

immune.combined@meta.data <- immune.combined@meta.data %>% 
  mutate(act_inact = case_when(
    str_detect(predicted.celltype, "Monocytes activated") ~ "Activated monocytes",
    .default = predicted.celltype
  ))

# table_for_celltype_condition_percent_plot1 <- table(immune.combined@meta.data$predicted.celltype, immune.combined@meta.data$Group) %>% 
#   as.data.frame()
#   
# table_for_celltype_condition_percent_plot <- immune.combined@meta.data %>% 
#   group_by(Group, predicted.celltype) %>% 
#   summarise(n = n()) %>% 
#   mutate(freq = 100*n / sum(n)) %>% 
#   select(-n) %>% 
#   pivot_wider(names_from = Group, values_from = freq) %>% 
#   column_to_rownames("predicted.celltype")
#   
# write.csv2(table_for_celltype_condition_percent_plot, "Condition_celltype_percent.csv", row.names = T, quote = F)
```

```{r}

genes <- wilcoxauc(immune.combined, 'act_inact')
#dplyr::count(genes, group)

m_df<- msigdbr(species = "Homo sapiens", category = "C2")
fgsea_sets<- m_df %>% split(x = .$gene_symbol, f = .$gs_name)

```

# Activated monocytes

```{r}
DT::datatable(genes %>%
  dplyr::filter(group == "Activated monocytes") %>%
  arrange(desc(logFC), desc(auc)) %>% 
  filter(logFC >= 1) )

DT::datatable(genes %>%
  dplyr::filter(group == "Activated monocytes") %>%
  arrange(desc(logFC), desc(auc)) %>% 
  filter(logFC < -0.5) )

act.mono.genes<- genes %>%
  dplyr::filter(group == "Activated monocytes") %>%
  arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)

ranks<- deframe(act.mono.genes)

#head(ranks)

fgseaRes<- fgsea(fgsea_sets, stats = ranks, nperm = 1000)

fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))


fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES, -nMoreExtreme) %>% 
  arrange(padj) 

# only plot the top 20 pathways
# ggplot(fgseaResTidy %>% filter(pval < 0.008) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
#   geom_col(aes(fill= NES)) +
#   coord_flip() +
#   labs(x="Pathway", y="Normalized Enrichment Score",
#        title="Hallmark pathways NES from GSEA") + 
#   theme_minimal()

png("IRF3.png", res = 300, height = 15, width = 25, units = "cm")
plotEnrichment(fgsea_sets[["REACTOME_IRF3_MEDIATED_INDUCTION_OF_TYPE_I_IFN"]],
               ranks) + labs(title="REACTOME_IRF3_MEDIATED_INDUCTION_OF_TYPE_I_IFN")
dev.off()

png("IRF3_IRF7.png", res = 300, height = 15, width = 25, units = "cm")
plotEnrichment(fgsea_sets[["REACTOME_ACTIVATION_OF_IRF3_IRF7_MEDIATED_BY_TBK1_IKK_EPSILON"]],
               ranks) + labs(title="REACTOME_ACTIVATION_OF_IRF3_IRF7_MEDIATED_BY_TBK1_IKK_EPSILON")
dev.off()


```



```{r}

genes <- wilcoxauc(immune.combined, 'act_inact')
#dplyr::count(genes, group)

m_df<- msigdbr(species = "Homo sapiens", category = "C5")
fgsea_sets<- m_df %>% split(x = .$gene_symbol, f = .$gs_name)

```

# Activated monocytes

```{r}
DT::datatable(genes %>%
  dplyr::filter(group == "Activated monocytes") %>%
  arrange(desc(logFC), desc(auc)) %>% 
  filter(logFC >= 1) )

DT::datatable(genes %>%
  dplyr::filter(group == "Activated monocytes") %>%
  arrange(desc(logFC), desc(auc)) %>% 
  filter(logFC < -0.5) )

act.mono.genes<- genes %>%
  dplyr::filter(group == "Activated monocytes") %>%
  arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)

ranks<- deframe(act.mono.genes)

#head(ranks)

fgseaRes<- fgsea(fgsea_sets, stats = ranks, nperm = 1000)

fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))


fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES, -nMoreExtreme) %>% 
  arrange(padj) 

# only plot the top 20 pathways
# ggplot(fgseaResTidy %>% filter(pval < 0.008) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
#   geom_col(aes(fill= NES)) +
#   coord_flip() +
#   labs(x="Pathway", y="Normalized Enrichment Score",
#        title="Hallmark pathways NES from GSEA") + 
#   theme_minimal()
png("wounding.png", res = 300, height = 15, width = 25, units = "cm")
plotEnrichment(fgsea_sets[["GOBP_RESPONSE_TO_WOUNDING"]],
               ranks) + labs(title="GOBP_RESPONSE_TO_WOUNDING")
dev.off()


```

```{r}
genes <- wilcoxauc(immune.combined, 'predicted.celltype')
#dplyr::count(genes, group)

m_df<- msigdbr(species = "Homo sapiens", category = "C5")
fgsea_sets<- m_df %>% split(x = .$gene_symbol, f = .$gs_name)

```


# Monocytes activated (Nfkb)
```{r}

DT::datatable(genes %>%
  dplyr::filter(group == "Monocytes activated (Nfkb)") %>%
  arrange(desc(logFC), desc(auc)) %>% 
  filter(logFC >= 1) )

DT::datatable(genes %>%
  dplyr::filter(group == "Monocytes activated (Nfkb)") %>%
  arrange(desc(logFC), desc(auc)) %>% 
  filter(logFC < -0.5) )

act.mono.nfkb.genes<- genes %>%
  dplyr::filter(group == "Monocytes activated (Nfkb)") %>%
  arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)


ranks<- deframe(act.mono.nfkb.genes)

#head(ranks)

fgseaRes<- fgsea(fgsea_sets, stats = ranks, nperm = 1000)

fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))


fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES, -nMoreExtreme) %>% 
  arrange(padj)

# only plot the top 20 pathways
ggplot(fgseaResTidy %>% filter(pval < 0.008) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= NES)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_minimal()

plotEnrichment(fgsea_sets[["GOBP_RESPONSE_TO_WOUNDING"]],
               ranks) + labs(title="GOBP_RESPONSE_TO_WOUNDING")

# How to read the figure ?
# 
# The X-axis is all your genes in the expriment (~ 20,000 in this case) pre-ranked by your metric. each black bar is the gene in this gene set(pathway). You have an idea where are the genes located in the pre-ranked list.
# 
# Enrichement Score (ES) is calcuated by some metric that ES is positive if the gene set is located in the top of the pre-ranked gene list. ES is negative if the gene set is located in the bottom of the pre-ranked gene list.
# 

```

# Monocytes activated (type I IFN signature)
```{r}

DT::datatable(genes %>%
  dplyr::filter(group == "Monocytes activated (type I IFN signature)") %>%
  arrange(desc(logFC), desc(auc)) %>% 
  filter(logFC >= 1) )

DT::datatable(genes %>%
  dplyr::filter(group == "Monocytes activated (type I IFN signature)") %>%
  arrange(desc(logFC), desc(auc)) %>% 
  filter(logFC < -0.5) )

act.mono.ifn.genes <- genes %>%
  dplyr::filter(group == "Monocytes activated (type I IFN signature)") %>%
  arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)


ranks<- deframe(act.mono.ifn.genes)

#head(ranks)

fgseaRes<- fgsea(fgsea_sets, stats = ranks, nperm = 1000)

fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))


fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES, -nMoreExtreme) %>% 
  arrange(padj)

# only plot the top 20 pathways
ggplot(fgseaResTidy %>% filter(pval < 0.008) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= NES)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_minimal()

plotEnrichment(fgsea_sets[["GOBP_RESPONSE_TO_WOUNDING"]],
               ranks) + labs(title="GOBP_RESPONSE_TO_WOUNDING")

# How to read the figure ?
# 
# The X-axis is all your genes in the expriment (~ 20,000 in this case) pre-ranked by your metric. each black bar is the gene in this gene set(pathway). You have an idea where are the genes located in the pre-ranked list.
# 
# Enrichement Score (ES) is calcuated by some metric that ES is positive if the gene set is located in the top of the pre-ranked gene list. ES is negative if the gene set is located in the bottom of the pre-ranked gene list.
# 

```


# Monocytes activated (Nfkb + type I IFN)
```{r}

DT::datatable(genes %>%
  dplyr::filter(group == "Monocytes activated (Nfkb + type I IFN)") %>%
  arrange(desc(logFC), desc(auc)) %>% 
  filter(logFC >= 1) )

DT::datatable(genes %>%
  dplyr::filter(group == "Monocytes activated (Nfkb + type I IFN)") %>%
  arrange(desc(logFC), desc(auc)) %>% 
  filter(logFC < -0.5) )

act.mono.nfkb.ifn.genes<- genes %>%
  dplyr::filter(group == "Monocytes activated (Nfkb + type I IFN)") %>%
  arrange(desc(auc)) %>% 
  dplyr::select(feature, auc)


ranks<- deframe(act.mono.nfkb.ifn.genes)

#head(ranks)

fgseaRes<- fgsea(fgsea_sets, stats = ranks, nperm = 1000)

fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))


fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES, -nMoreExtreme) %>% 
  arrange(padj)

# only plot the top 20 pathways
ggplot(fgseaResTidy %>% filter(pval < 0.008) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= NES)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_minimal()

plotEnrichment(fgsea_sets[["GOBP_RESPONSE_TO_WOUNDING"]],
               ranks) + labs(title="GOBP_RESPONSE_TO_WOUNDING")

# How to read the figure ?
# 
# The X-axis is all your genes in the expriment (~ 20,000 in this case) pre-ranked by your metric. each black bar is the gene in this gene set(pathway). You have an idea where are the genes located in the pre-ranked list.
# 
# Enrichement Score (ES) is calcuated by some metric that ES is positive if the gene set is located in the top of the pre-ranked gene list. ES is negative if the gene set is located in the bottom of the pre-ranked gene list.
# 

```