---
title: "Single cell RNA-seq of Plasmodium sporozoite-infected monocytes - integration"
output:
  html_document:
    df_print: paged
    code_folding: hide
  pdf_document: default
  toc: TRUE
---

```{r, warnings = F, message = F}
library(Seurat)
library(patchwork)
library(readxl)
library(tidyverse)
library(ggplot2)
library(ggpubr)
library(reshape2)
library(broom)
library(stringr)
library(ggplot2)
library(SeuratData)
#library(celldex)
#install.packages("cowplot")
library(cowplot)
#devtools::install_github('cole-trapnell-lab/monocle3')
library(monocle3)
#remotes::install_github('satijalab/seurat-wrappers')
library(SeuratWrappers)
#BiocManager::install('multtest', force = T)
library(multtest)
# install.packages('metap')
#library(metap)
#remotes::install_github("satijalab/seurat-data")
```

```{r}

# EXP1_l2 <- readRDS("Jan27analysis/EXP1_l2_res0.9.rds")
# EXP2_l1 <- readRDS("Jan27analysis/EXP2_l1_res0.9.rds")
# EXP2_l2 <- readRDS("Jan27analysis/EXP2_l2_res0.9.rds")
# EXP2_l3 <- readRDS("Jan27analysis/EXP2_l3_res0.9.rds")

EXP1_l1 <- readRDS(list.files(pattern = "^EXP1_l1.*\\.rds$"))
EXP1_l2 <- readRDS(list.files(pattern = "^EXP1_l2.*\\.rds$"))
EXP2_l1 <- readRDS(list.files(pattern = "^EXP2_l1.*\\.rds$"))
EXP2_l2 <- readRDS(list.files(pattern = "^EXP2_l2_reseq.*\\.rds$"))
EXP2_l3 <- readRDS(list.files(pattern = "^EXP2_l3.*\\.rds$"))

exp.list <- list(EXP1_l1 = EXP1_l1,
  EXP1_l2 = EXP1_l2,
  EXP2_l1 = EXP2_l1,
  EXP2_l2 = EXP2_l2,
  EXP2_l3 = EXP2_l3
            )

```

```{r}

# barplot of how many cells of each condition are in each lane

meta.mat <- pseudo.integration@meta.data

ggplot(meta.mat, aes(x = factor(orig.ident))) + 
  geom_bar(stat = "count", position = position_dodge(), aes(fill = Group)) +
  scale_fill_viridis_d() +
  theme_bw()

ggsave("cellnumber_per_condition_in_lanes.png", dpi = 300, width = 25, height = 15, units = "cm")

ggplot(meta.mat, aes(x = factor(orig.ident))) + 
  geom_bar(stat = "count", position = position_dodge(), aes(fill = Group)) +
  scale_fill_viridis_d() +
  theme_bw()

 ggplot(meta.mat, aes(x = factor(orig.ident), fill = Group)) + 
  geom_bar(position = "fill") + 
   geom_text(
    aes(label=signif(..count.. / tapply(..count.., ..x.., sum)[as.character(..x..)], digits=3)),
    stat="count",
    position=position_fill(vjust=0.5)) +
  scale_fill_viridis_d() +
  theme_bw()

ggsave("cellnumber_per_condition_in_lanes_percent.png", dpi = 300, width = 25, height = 15, units = "cm")
  
```


```{r}
# merge analysis
pseudo.integration <- merge(EXP1_l1,
                            c(EXP1_l2, EXP2_l1, EXP2_l2, EXP2_l3),
                            project = "Malaria2022")

pseudo.integration <- NormalizeData(pseudo.integration)
pseudo.integration <- ScaleData(pseudo.integration, verbose = FALSE)
pseudo.integration <- FindVariableFeatures(pseudo.integration, selection.method = "vst", nfeatures = 2000)
pseudo.integration <- RunPCA(pseudo.integration, npcs = 30, verbose = FALSE)
pseudo.integration <- RunUMAP(pseudo.integration, reduction = "pca", dims = 1:30)
pseudo.integration <- FindNeighbors(pseudo.integration, reduction = "pca", dims = 1:30)
pseudo.integration <- FindClusters(pseudo.integration, resolution = 0.5)

DimPlot(pseudo.integration, reduction = "umap", group.by = "orig.ident")
```

```{r}
EXP1_l1_meta <- EXP1_l1@meta.data %>% 
  rownames_to_column("Cell") %>% 
  select("Cell", "nFeature_RNA")
EXP1_l1_umap <- EXP1_l1@reductions[["umap"]]@cell.embeddings %>% 
  as.data.frame() %>% 
  rownames_to_column("Cell") %>%
  left_join(., EXP1_l1_meta) %>% 
  mutate(lib = "EXP1_l1")


EXP1_l2_meta <- EXP1_l2@meta.data %>% 
  rownames_to_column("Cell") %>% 
  select("Cell", "nFeature_RNA")
EXP1_l2_umap <- EXP1_l2@reductions[["umap"]]@cell.embeddings %>% 
  as.data.frame() %>% 
  rownames_to_column("Cell") %>%
  left_join(., EXP1_l2_meta) %>% 
  mutate(lib = "EXP1_l2")



EXP2_l1_meta <- EXP2_l1@meta.data %>% 
  rownames_to_column("Cell") %>% 
  select("Cell", "nFeature_RNA")
EXP2_l1_umap <- EXP2_l1@reductions[["umap"]]@cell.embeddings %>% 
  as.data.frame() %>% 
  rownames_to_column("Cell") %>%
  left_join(., EXP2_l1_meta) %>% 
  mutate(lib = "EXP2_l1")


EXP2_l2_meta <- EXP2_l2@meta.data %>% 
  rownames_to_column("Cell") %>% 
  select("Cell", "nFeature_RNA")
EXP2_l2_umap <- EXP2_l2@reductions[["umap"]]@cell.embeddings %>% 
  as.data.frame() %>% 
  rownames_to_column("Cell") %>%
  left_join(., EXP2_l2_meta) %>% 
  mutate(lib = "EXP2_l2")


EXP2_l3_meta <- EXP2_l3@meta.data %>% 
  rownames_to_column("Cell") %>% 
  select("Cell", "nFeature_RNA")
EXP2_l3_umap <- EXP2_l3@reductions[["umap"]]@cell.embeddings %>% 
  as.data.frame() %>% 
  rownames_to_column("Cell") %>%
  left_join(., EXP2_l3_meta) %>% 
  mutate(lib = "EXP2_l3")

umap <- rbind(EXP1_l1_umap,
              EXP1_l2_umap,
              EXP2_l1_umap,
              EXP2_l2_umap,
              EXP2_l3_umap)

myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
sc <- scale_colour_gradientn(colours = myPalette(100), limits=c(100, 6000))

ggplot(umap, aes(UMAP_1, UMAP_2, colour = nFeature_RNA)) +
  geom_point(size = 0.3, alpha = 0.8) +
  facet_wrap( ~ lib, ncol = 2, scales = "free") +
  sc +
  theme_bw()
ggsave("UMAP_nFeature_RNA_libwise.png", height = 9, width = 10, dpi = 300)

EXP1_l1_sporozoite_infected <- subset(EXP1_l1, subset = Group == "sporozoite_infected")
DimPlot(EXP1_l1_sporozoite_infected, group.by = "nFeature_RNA") + ggtitle("EXP1_l1_sporozoite_infected")

EXP1_l1_RBC_bystander <- subset(EXP1_l1, subset = Group == "RBC_bystander")
DimPlot(EXP1_l1_RBC_bystander, group.by = "nFeature_RNA") + ggtitle("EXP1_l1_RBC_bystander") + NoLegend()

EXP1_l1_RBC_control <- subset(EXP1_l1, subset = Group == "RBC_control")
DimPlot(EXP1_l1_RBC_control, group.by = "nFeature_RNA") + ggtitle("EXP1_l1_RBC_control") + NoLegend()

EXP1_l1_control <- subset(EXP1_l1, subset = Group == "control")
DimPlot(EXP1_l1_control, group.by = "nFeature_RNA") + ggtitle("EXP1_l1_control") + NoLegend()
```


```{r}
# normalize and identify variable features for each dataset independently
exp.list <- lapply(X = exp.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = exp.list)
# retrieve anchors

immune.anchors <- FindIntegrationAnchors(object.list = exp.list, anchor.features = features)

# this command creates an 'integrated' data assay
immune.combined <- IntegrateData(anchorset = immune.anchors)
```

```{r}
# Let’s set the assay to RNA and visualize the datasets before integration.

DefaultAssay(immune.combined) <- "RNA"

#Let’s do normalization, HVG finding, scaling, PCA, and UMAP on the un-integrated (RNA) assay:

immune.combined <- NormalizeData(immune.combined, verbose = F)
immune.combined <- FindVariableFeatures(immune.combined, selection.method = "vst", nfeatures = 2000, verbose = F)
immune.combined <- ScaleData(immune.combined, verbose = F)
immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = F)
immune.combined <- RunUMAP(immune.combined, reduction = "pca", dims = 1:30, verbose = F)

## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
## To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
## This message will be shown once per session

# UMAP plot of the datasets before integration shows clear separation. Note that we can use patchwork syntax with Seurat plotting functions:

DimPlot(immune.combined,reduction = "umap", group.by = "orig.ident") + plot_annotation(title = "Before Integration")

```


```{r}

DefaultAssay(immune.combined) <- "integrated"

# Run the standard workflow for visualization and clustering
immune.combined <- ScaleData(immune.combined, verbose = FALSE)
immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)
immune.combined <- RunUMAP(immune.combined, reduction = "pca", dims = 1:30)
immune.combined <- FindNeighbors(immune.combined, reduction = "pca", dims = 1:30)
immune.combined <- FindClusters(immune.combined, resolution = 0.5)

saveRDS(immune.combined, "immune.combined_all_libs_preprocessed.rds")

# Visualization
p1 <- DimPlot(immune.combined, reduction = "umap", group.by = "orig.ident") + plot_annotation(title = "after integration")
#p2 <- DimPlot(immune.combined, reduction = "umap", label = TRUE, repel = TRUE)
p1 #+ p2
DimPlot(immune.combined, reduction = "umap", group.by = "hash.ID")+ plot_annotation(title = "after integration")
DimPlot(immune.combined, reduction = "umap", group.by = "Group")+ plot_annotation(title = "after integration")

```


```{r}


# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the 'RNA' assay

VlnPlot(immune.combined, features=c("nCount_RNA", "nFeature_RNA", "percent.mt"), group.by = "orig.ident", pt.size = 0.5) 

ggplot(immune.combined@meta.data, aes(x = nFeature_RNA, fill = factor(orig.ident))) + geom_histogram(binwidth = 50, position = "identity") + theme_bw() + facet_wrap(~ orig.ident)
ggsave("after_integration_nFeatures_per_lane.png", dpi = 300, width = 10)
#ggplot(mono@meta.data, aes(x = nFeature_RNA)) + geom_histogram(binwidth = 20) + xlim(c(0, 600))

ggplot(immune.combined@meta.data, aes(x = nCount_RNA, fill = factor(orig.ident))) + geom_histogram(binwidth = 200, position = "identity") + theme_bw() + facet_wrap(~ orig.ident)
ggsave("after_integration_nCount_per_lane.png", dpi = 300, width = 10)#ggplot(mono@meta.data, aes(x = nCount_RNA)) + geom_histogram(binwidth = 50) + xlim(c(0, 5000))

ggplot(immune.combined@meta.data, aes(x = percent.mt, fill = factor(orig.ident))) + geom_histogram(binwidth = 0.1, position = "identity") + theme_bw() + facet_wrap(~ orig.ident)
ggsave("after_integration_precent.mt_per_lane.png", dpi = 300, width = 10)
```


```{r}
Idents(immune.combined) <- "predicted.celltype"
png("integrated_celltypes_by_condition.png", width = 40, height = 30, units = "cm", res = 300)
DimPlot(immune.combined, reduction = "umap", split.by = "Group", ncol = 2,
        order = c("control", "RBC_control",
                  "sporozoite_bystander", "RBC_bystander",
                  "sporozoite_infected", "RBC_infected"))
dev.off()


```

<!-- ```{r} -->
<!-- # get markers -->
<!-- # sudo apt-get install libfftw3-dev -->
<!-- # sudo apt-get install libsndfile1-dev -->

<!-- DefaultAssay(immune.combined) <- "RNA" -->
<!-- nk.markers <- FindConservedMarkers(immune.combined, ident.1 = 0, grouping.var = "orig.ident", verbose = FALSE) -->
<!-- head(nk.markers) -->
<!-- FeaturePlot(immune.combined, features = rownames(nk.markers)[1:15], min.cutoff = "q9") -->
<!-- ``` -->

```{r}
# For performing differential expression after integration, we switch back to the original
# data
#DefaultAssay(immune.combined) <- "RNA"
#markers <- FindConservedMarkers(immune.combined, grouping.var = "seurat_clusters", verbose = FALSE)
#head(markers)

allmarkers <- FindAllMarkers(immune.combined)
mono.allmarkers.top20 <- allmarkers %>%
    group_by(cluster) %>%
    slice_max(n = 20, order_by = avg_log2FC)
mono.allmarkers.top20

set.seed(123)
FeaturePlot(immune.combined, features = sample(mono.allmarkers.top20$gene, 10))
FeaturePlot(immune.combined, features = c("CD14", "FCGR3A"))
FeaturePlot(immune.combined, features = c("CD3D", "SELL", "CREM", "CD8A", "GNLY", "CD79A", "FCGR3A",
    "CCL2", "PPBP"), min.cutoff = "q9")
png("annotation.png", width = 30, height = 30, units = "cm", res = 300)
DimPlot(immune.combined, reduction = "umap", label = "T")
dev.off()
DimPlot(immune.combined, reduction = "umap", label = "F", group.by = "orig.ident")
DimPlot(immune.combined, reduction = "umap", label = "F", group.by = "Group")


mono.allmarkers.top20 %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10
#DoHeatmap(immune.combined, features = top10$gene) + NoLegend() + theme(text = element_text(size = 5))

```

```{r}

DoHeatmap(immune.combined, features = top10$gene) + NoLegend() + theme(text = element_text(size = 5))
# pdf("IntegratedDataHeatmap.pdf")
# DoHeatmap(immune.combined, features = top10$gene) + NoLegend() + theme(text = element_text(size = 5))
# dev.off()
```

### Dotplot to view marker genes in each cluster
```{r}
# DotPlot function

myDotPlot <- function(
  object,
  assay = NULL,
  features,
  cols = c("lightgrey", "darkblue", "salmon", "goldenrod1", "lightgreen"),
  col.min = -2.5,
  col.max = 2.5,
  dot.min = 0,
  dot.scale = 3,
  text.size = 6,
  idents = NULL,
  group.by = NULL,
  split.by = NULL,
  cluster.idents = FALSE,
  scale = TRUE,
  scale.by = 'radius',
  scale.min = NA,
  scale.max = NA
) {
  assay <- assay %||% DefaultAssay(object = object)
  DefaultAssay(object = object) <- assay
  brewer.pal.info = c("lightgreen", "darkblue", "salmon", "goldenrod1")
  split.colors <- !is.null(x = split.by) && !any(cols %in% rownames(x = brewer.pal.info))
  scale.func <- switch(
    EXPR = scale.by,
    'size' = scale_size,
    'radius' = scale_radius,
    stop("'scale.by' must be either 'size' or 'radius'")
  )
  feature.groups <- NULL
  if (is.list(features) | any(!is.na(names(features)))) {
    feature.groups <- unlist(x = sapply(
      X = 1:length(features),
      FUN = function(x) {
        return(rep(x = names(x = features)[x], each = length(features[[x]])))
      }
    ))
    if (any(is.na(x = feature.groups))) {
      warning(
        "Some feature groups are unnamed.",
        call. = FALSE,
        immediate. = TRUE
      )
    }
    features <- unlist(x = features)
    names(x = feature.groups) <- features
  }
  cells <- unlist(x = CellsByIdentities(object = object, idents = idents))

  data.features <- FetchData(object = object, vars = features, cells = cells)
  data.features$id <- if (is.null(x = group.by)) {
    Idents(object = object)[cells, drop = TRUE]
  } else {
    object[[group.by, drop = TRUE]][cells, drop = TRUE]
  }
  if (!is.factor(x = data.features$id)) {
    data.features$id <- factor(x = data.features$id)
  }
  id.levels <- levels(x = data.features$id)
  data.features$id <- as.vector(x = data.features$id)
  if (!is.null(x = split.by)) {
    splits <- object[[split.by, drop = TRUE]][cells, drop = TRUE]
    if (split.colors) {
      if (length(x = unique(x = splits)) > length(x = cols)) {
        stop("Not enough colors for the number of groups")
      }
      cols <- cols[1:length(x = unique(x = splits))]
      names(x = cols) <- unique(x = splits)
    }
    data.features$id <- paste(data.features$id, splits, sep = '_')
    unique.splits <- unique(x = splits)
    id.levels <- paste0(rep(x = id.levels, each = length(x = unique.splits)), "_", rep(x = unique(x = splits), times = length(x = id.levels)))
  }
  data.plot <- lapply(
    X = unique(x = data.features$id),
    FUN = function(ident) {
      data.use <- data.features[data.features$id == ident, 1:(ncol(x = data.features) - 1), drop = FALSE]
      avg.exp <- apply(
        X = data.use,
        MARGIN = 2,
        FUN = function(x) {
          return(mean(x = expm1(x = x)))
        }
      )
      pct.exp <- apply(X = data.use, MARGIN = 2, FUN = PercentAbove, threshold = 0)
      return(list(avg.exp = avg.exp, pct.exp = pct.exp))
    }
  )
  names(x = data.plot) <- unique(x = data.features$id)
  if (cluster.idents) {
    mat <- do.call(
      what = rbind,
      args = lapply(X = data.plot, FUN = unlist)
    )
    mat <- scale(x = mat)
    id.levels <- id.levels[hclust(d = dist(x = mat))$order]
  }
  data.plot <- lapply(
    X = names(x = data.plot),
    FUN = function(x) {
      data.use <- as.data.frame(x = data.plot[[x]])
      data.use$features.plot <- rownames(x = data.use)
      data.use$id <- x
      return(data.use)
    }
  )
  data.plot <- do.call(what = 'rbind', args = data.plot)
  if (!is.null(x = id.levels)) {
    data.plot$id <- factor(x = data.plot$id, levels = id.levels)
  }
  ngroup <- length(x = levels(x = data.plot$id))
  if (ngroup == 1) {
    scale <- FALSE
    warning(
      "Only one identity present, the expression values will be not scaled",
      call. = FALSE,
      immediate. = TRUE
    )
  } else if (ngroup < 5 & scale) {
    warning(
      "Scaling data with a low number of groups may produce misleading results",
      call. = FALSE,
      immediate. = TRUE
    )
  }
  avg.exp.scaled <- sapply(
    X = unique(x = data.plot$features.plot),
    FUN = function(x) {
      data.use <- data.plot[data.plot$features.plot == x, 'avg.exp']
      if (scale) {
        data.use <- scale(x = data.use)
        data.use <- MinMax(data = data.use, min = col.min, max = col.max)
      } else {
        data.use <- log1p(x = data.use)
      }
      return(data.use)
    }
  )
  avg.exp.scaled <- as.vector(x = t(x = avg.exp.scaled))
  if (split.colors) {
    avg.exp.scaled <- as.numeric(x = cut(x = avg.exp.scaled, breaks = 20))
  }
  data.plot$avg.exp.scaled <- avg.exp.scaled
  data.plot$features.plot <- factor(
    x = data.plot$features.plot,
    levels = features
  )
  data.plot$pct.exp[data.plot$pct.exp < dot.min] <- NA
  data.plot$pct.exp <- data.plot$pct.exp * 100
  if (split.colors) {
    splits.use <- vapply(
      X = as.character(x = data.plot$id),
      FUN = gsub,
      FUN.VALUE = character(length = 1L),
      pattern =  paste0(
        '^((',
        paste(sort(x = levels(x = object), decreasing = TRUE), collapse = '|'),
        ')_)'
      ),
      replacement = '',
      USE.NAMES = FALSE
    )
    data.plot$colors <- mapply(
      FUN = function(color, value) {
        return(colorRampPalette(colors = c('lightgrey', color))(20)[value])
      },
      color = cols[splits.use],
      value = avg.exp.scaled
    )
  }
  color.by <- ifelse(test = split.colors, yes = 'colors', no = 'avg.exp.scaled')
  if (!is.na(x = scale.min)) {
    data.plot[data.plot$pct.exp < scale.min, 'pct.exp'] <- scale.min
  }
  if (!is.na(x = scale.max)) {
    data.plot[data.plot$pct.exp > scale.max, 'pct.exp'] <- scale.max
  }
  if (!is.null(x = feature.groups)) {
    data.plot$feature.groups <- factor(
      x = feature.groups[data.plot$features.plot],
      levels = unique(x = feature.groups)
    )
  }
  plot <- ggplot(data = data.plot, mapping = aes_string(x = 'features.plot', y = 'id')) +
    geom_point(mapping = aes_string(size = 'pct.exp', color = color.by)) +
    scale.func(range = c(0, dot.scale), limits = c(scale.min, scale.max)) +
    theme(axis.title.x = element_blank(), axis.title.y = element_blank(), 
          axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1)) +
    guides(size = guide_legend(title = 'Percent Expressed')) +
    labs(
      x = 'Features',
      y = ifelse(test = is.null(x = split.by), yes = 'Identity', no = 'Split Identity')
    ) +
    theme_cowplot() +
    theme(axis.text=element_text(size=text.size)) +
    coord_flip()
  
  if (!is.null(x = feature.groups)) {
    plot <- plot + facet_grid(
      facets = ~feature.groups,
      scales = "free_x",
      space = "free_x",
      switch = "y"
    ) + theme(
      panel.spacing = unit(x = 1, units = "lines"),
      strip.background = element_blank()
    )
  }
  if (split.colors) {
    plot <- plot + scale_color_identity()
  } else if (length(x = cols) == 1) {
    plot <- plot + scale_color_distiller(palette = cols)
  } else {
    plot <- plot + scale_color_gradient(low = cols[1], high = cols[2])
  }
  if (!split.colors) {
    plot <- plot + guides(color = guide_colorbar(title = 'Average Expression'))
  }
  return(plot)
}

```

<!-- ```{r} -->
<!-- # assign annotation to clusters -->
<!-- immune.combined <- RenameIdents(immune.combined, `0` = "CD14 Mono", `1` = "CD4 Naive T", `2` = "CD4 Memory T", -->
<!--     `3` = "CD16 Mono", `4` = "B", `5` = "CD8 T", `6` = "NK", `7` = "T activated", `8` = "DC", `9` = "B Activated", -->
<!--     `10` = "Mk", `11` = "pDC", `12` = "Eryth", `13` = "Mono/Mk Doublets", `14` = "HSPC") -->
<!-- DimPlot(immune.combined, label = TRUE) -->
<!-- ``` -->


```{r}
# Idents(immune.combined) <- factor(Idents(immune.combined), levels = c("HSPC", "Mono/Mk Doublets",
#     "pDC", "Eryth", "Mk", "DC", "CD14 Mono", "CD16 Mono", "B Activated", "B", "CD8 T", "NK", "T activated",
#     "CD4 Naive T", "CD4 Memory T"))
# markers.to.plot <- c("CD3D", "CREM", "HSPH1", "SELL", "GIMAP5", "CACYBP", "GNLY", "NKG7", "CCL5",
#     "CD8A", "MS4A1", "CD79A", "MIR155HG", "NME1", "FCGR3A", "VMO1", "CCL2", "S100A9", "HLA-DQA1",
#     "GPR183", "PPBP", "GNG11", "HBA2", "HBB", "TSPAN13", "IL3RA", "IGJ", "PRSS57")

myDotPlot(immune.combined, features = unique(top10$gene), dot.scale = 2, split.by = "orig.ident", text.size = 3) + RotatedAxis()
myDotPlot(immune.combined, features = unique(top10$gene), dot.scale = 2, split.by = "Group", text.size = 3) + RotatedAxis()

png("integrated_dotplot_groupbyCluster2.png", res = 300, width = 40, height = 40, units = "cm")
myDotPlot(immune.combined, features = unique(top10$gene), dot.scale = 4, text.size = 9, group.by = "seurat_clusters",) + RotatedAxis()
dev.off()
```

```{r}

# immune.combined <- RenameIdents(immune.combined, 
#                                 `0` = "Classical monocytes", 
#                                 `1` = "Monocytes activated (Nfkb)", 
#                                 `2` = "Intermediate monocytes",
#                                 `3` = "Monocytes activated (type I IFN signature)",
#                                 `4` = "Monocytes activated (Nfkb + type I IFN)", 
#                                 `5` = "Classical monocytes II",
#                                 `6` = "Classical Monocytes III",
#                                 `7` = "Non-classical monocytes",
#                                 `8` = "(CD8?) T-cells",
#                                 `9` = "DCs",
#                                 `10` = "Activated DCs?", 
#                                 `11` = "??",
#                                 `12` = "Non-classcial monocytes II", 
#                                 `13` = "?",
#                                 `14` = "B-cells")

immune.combined <- RenameIdents(immune.combined, 
                                `0` = "Classical monocytes", 
                                `1` = "Monocytes activated (Nfkb)", 
                                `2` = "Intermediate monocytes",
                                `3` = "Monocytes activated (type I IFN signature)",
                                `4` = "Monocytes activated (Nfkb + type I IFN)", 
                                `5` = "Classical monocytes II",
                                `6` = "Classical Monocytes III",
                                `7` = "?",
                                `8` = "Non-classical monocytes",
                                `9` = "(CD8?) T cells",
                                `10` = "DCs", 
                                `11` = "Activated DCs?",
                                `12` = "(CD8?) T cells II", 
                                `13` = "B cells",
                                `14` = "Metallotheionein expressing",
                                `15` = "(CD8?) T cells III")
immune.combined@meta.data$predicted.celltype <- immune.combined@active.ident

DimPlot(immune.combined, label = TRUE)
top10 %>%
    group_by(cluster) %>%
    top_n(n = 2, wt = avg_log2FC) -> top2
myDotPlot(immune.combined, features = unique(top2$gene),  dot.scale = 5, text.size = 7) + RotatedAxis()
```

```{r}

#QC coloured by clusters

# three clusters
table(immune.combined$RNA_snn_res.0.9)
cluster.set <- unique(immune.combined$RNA_snn_res.0.9)
# overall nFeature
sum(rowSums(immune.combined[['integrated']]@data) != 0)

# nFeature for each cluster
nF <- sapply(X = cluster.set, function(c) {
  cells.c <- WhichCells(object = immune.combined, expression = RNA_snn_res.0.9 == c)
  #nFeature.c <- sum(rowSums(mono[['RNA']]@counts[, cells.c ]) != 0) / length(cells.c) # normalised by number of cells in a cluster
  df.cells <- immune.combined@meta.data[which(rownames(immune.combined@meta.data) %in% cells.c),]
  nFeature.c <- sum(df.cells$nFeature_RNA) / length(cells.c)
  return(nFeature.c)
}
)

#nC <- sapply(cluster.set, function(x) sum(x == mono$RNA_snn_res.0.9))

nC <- sapply(X = cluster.set, function(c) {
  cells.c <- WhichCells(object = immune.combined, expression = RNA_snn_res.0.9 == c)
  df.cells <- immune.combined@meta.data[which(rownames(immune.combined@meta.data) %in% cells.c),]
  nCount.c <- sum(df.cells$nCount_RNA) / length(cells.c)
  return(nCount.c)
}
)

# mt <- sapply(X = cluster.set, function(c) {
#   cells.c <- WhichCells(object = mono, expression = RNA_snn_res.0.9 == c)
#   mt.df <- mono[colnames(mono)%in%cells.c,]
#   mt <- sum(mono$percent.mt[ cells.c ] != 0) / length(cells.c) # normalised by number of cells in a cluster
#   return(mt)
# }
# )

# First the cluster annotation and the tsne embeddings are merged
label.df <- cbind(as.data.frame(immune.combined@meta.data), as.data.frame(immune.combined@reductions$umap@cell.embeddings)) 

# using dplyr across to calculate the mean mitochondrial percentage and 
# the median tsne values per cluster
label.df <- label.df %>% dplyr::group_by(seurat_clusters) %>% dplyr::summarise(dplyr::across(percent.mt, ~ mean(.), .names = "mean_{.col}"), dplyr::across(contains("UMAP"), ~ median(.)))

qc_df <- data.frame(seurat_clusters = cluster.set, nFeature_RNA = nF, nCount_RNA = nC) %>% 
  left_join(., label.df)

ggplot(qc_df, aes(x = nFeature_RNA, y = nCount_RNA, size = nCount_RNA, colour = mean_percent.mt)) + #, label = seurat_clusters
  geom_point(alpha = 0.7) +
  #geom_text(check_overlap = T, hjust = 0, nudge_x = 1.5, nudge_y = 1.5, colour = "black") +
  ggrepel::geom_text_repel(aes(label = seurat_clusters), size = 3) +
  theme_bw()


```


```{r}
celltype.set <- unique(immune.combined$predicted.celltype)
# overall nFeature
sum(rowSums(immune.combined[['integrated']]@data) != 0)

# nFeature for each cluster
nF <- sapply(X = celltype.set, function(c) {
  cells.c <- WhichCells(object = immune.combined, expression = predicted.celltype == c)
  #nFeature.c <- sum(rowSums(immune.combined[['RNA']]@counts[, cells.c ]) != 0) / length(cells.c) # normalised by number of cells in a cluster
  df.cells <- immune.combined@meta.data[which(rownames(immune.combined@meta.data) %in% cells.c),]
  nFeature.c <- sum(df.cells$nFeature_RNA) / length(cells.c)
  return(nFeature.c)
}
)

#nC <- sapply(cluster.set, function(x) sum(x == immune.combined$RNA_snn_res.0.9))

nC <- sapply(X = celltype.set, function(c) {
  cells.c <- WhichCells(object = immune.combined, expression = predicted.celltype == c)
  df.cells <- immune.combined@meta.data[which(rownames(immune.combined@meta.data) %in% cells.c),]
  nCount.c <- sum(df.cells$nCount_RNA) / length(cells.c)
  return(nCount.c)
}
)

# mt <- sapply(X = cluster.set, function(c) {
#   cells.c <- WhichCells(object = immune.combined, expression = RNA_snn_res.0.9 == c)
#   mt.df <- immune.combined[colnames(immune.combined)%in%cells.c,]
#   mt <- sum(immune.combined$percent.mt[ cells.c ] != 0) / length(cells.c) # normalised by number of cells in a cluster
#   return(mt)
# }
# )

# First the cluster annotation and the tsne embeddings are merged
label.df.ct <- cbind(as.data.frame(immune.combined@meta.data), as.data.frame(immune.combined@reductions$umap@cell.embeddings)) 

# using dplyr across to calculate the mean mitochondrial percentage and 
# the median tsne values per cluster
label.df.ct <- label.df.ct %>% 
  dplyr::group_by(predicted.celltype) %>% 
  dplyr::summarise(dplyr::across(percent.mt, ~ mean(.), .names = "mean_{.col}"), 
                   dplyr::across(contains("umap"), ~ median(.)))

qc_df.ct <- data.frame(predicted.celltype = celltype.set, nFeature_RNA = nF, nCount_RNA = nC) %>% 
  left_join(., label.df.ct)

ggplot(qc_df.ct, aes(x = nFeature_RNA, y = nCount_RNA, size = nCount_RNA, colour = mean_percent.mt)) + #, label = seurat_clusters
  geom_point(alpha = 0.7) +
  #geom_text(check_overlap = T, hjust = 0, nudge_x = 1.5, nudge_y = 1.5, colour = "black") +
  ggrepel::geom_text_repel(aes(label = predicted.celltype), size = 2, max.overlap = 100) +
  theme_bw()


```

<!-- ```{r} -->
<!-- # annotation -->
<!-- imm.query = exp.list[["EXP2_l2"]] -->
<!-- transfer.anchors <- FindTransferAnchors(reference = immune.combined, query = imm.query, -->
<!--     dims = 1:30, reference.reduction = "pca") -->
<!-- predictions <- TransferData(anchorset = immune.anchors, refdata = immune.combined$celltype, -->
<!--     dims = 1:30) -->
<!-- imm.query <- AddMetaData(imm.query, metadata = predictions) -->


<!-- MapQuery(anchorset = immune.anchors, -->
<!--          query = exp.list, -->
<!--          reference = immune.combined -->
<!--          ) -->


<!-- ``` -->

```{r}
# pseudo time analysis with monocle3
DefaultAssay(immune.combined) <- "RNA"
cds <- as.cell_data_set(immune.combined)
fData(cds)
rownames(fData(cds))[1:10]
fData(cds)$gene_short_name <- rownames(fData(cds))
head(fData(cds))
head(counts(cds))


#Retrieve clustering information from Surat object
#1. Assign partitions

recreate.partitions <- c(rep(1, length(cds@colData@rownames)))
names(recreate.partitions) <- cds@colData@rownames
recreate.partitions <- as.factor(recreate.partitions)
recreate.partitions

cds@clusters@listData[["UMAP"]][["partitions"]] <- recreate.partitions

# Assign cluster information

list.cluster <- immune.combined@active.ident
cds@clusters@listData[["UMAP"]][["clusters"]] <- list.cluster

# Assign UMAP coordinates

cds@int_colData@listData[["reducedDims"]]@listData[["UMAP"]] <- immune.combined@reductions$umap@cell.embeddings


# Learn Trajectory

cds <- learn_graph(cds, use_partition = F)

# Plot
cluster.before.traj <-plot_cells(cds, color_cells_by = "cluster", label_groups_by_cluster = T,
           group_label_size = 3, graph_label_size = 3) + theme(legend.position = "right")
cluster.before.traj


# Plot trajectory
#png("integrated_pseudotime.png")
plot_cells(cds, color_cells_by = "cluster", label_groups_by_cluster = F,
           label_branch_points = T, label_roots = T, label_leaves = F,
           group_label_size = 5,graph_label_size = 3)

# Pseudonym time cells in Pseudotime

cds <- order_cells(cds, reduction_method = "UMAP", root_cells = colnames(cds[, clusters(cds) == "Classical monocytes"])) #, root_cells = colnames(cds[, clusters(cds) == 0]).
# If annotation is not done, use cluster numbers (always use active.ident of object)
png("integrated_pseudotime_groupByPseudotime.png")
plot_cells(cds, color_cells_by = "pseudotime", label_groups_by_cluster = F,
           label_branch_points = F, label_roots = F, label_leaves = F,  graph_label_size = 3)
dev.off()
## Cells aren't colored in a way that allows them to be grouped.

# Cells ordered by Monocle3 Pseudotime

head(pseudotime(cds), 10)
cds$monocle3_pseudotime <- pseudotime(cds)
data.pseudo <- as.data.frame(colData(cds))

ggplot(data.pseudo, aes(monocle3_pseudotime, ident, fill = seurat_clusters)) + geom_boxplot()

ggplot(data.pseudo, aes(monocle3_pseudotime, reorder(ident, monocle3_pseudotime), fill = seurat_clusters)) + geom_boxplot()

# Find genes that change as a function of pseudotime
cds <- estimate_size_factors(cds)
cds@rowRanges@elementMetadata@listData[["gene_short_name"]] <- rownames(fData(cds)) # maybe
deg <- graph_test(cds, neighbor_graph = "principal_graph")
genes.pseudotime <- deg %>% arrange(q_value) %>% filter(status == "OK") %>%  pull(gene_short_name)

#FeaturePlot(immune.combined, features = genes.pseudotime)

# Add pseudotime values into the seuratobject

immune.combined$pseudotime <- pseudotime(cds)
FeaturePlot(immune.combined, features = "pseudotime", split.by = "Group")

df_meta <- immune.combined@meta.data %>% 
  rownames_to_column("Cell")
df <- immune.combined@reductions[["umap"]]@cell.embeddings %>% 
  as.data.frame() %>% 
  rownames_to_column("Cell") %>%
  left_join(., df_meta)

myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
sc <- scale_colour_gradientn(colours = myPalette(100), limits=c(0, 12))

ggplot(df, aes(UMAP_1, UMAP_2, colour = pseudotime)) +
  geom_point(size = 0.3, alpha = 0.8) +
  facet_wrap( ~ factor(Group,
                       levels = c("control", "RBC_control",
                       "sporozoite_bystander", "RBC_bystander",
                       "sporozoite_infected", "RBC_infected")), 
              ncol = 2, scales = "free") +
  sc +
  theme_classic() +
  theme(strip.background =element_rect(fill="aliceblue", colour = "lightblue"))
ggsave("pseudotime_split.by.condition.png", dpi = 300, width = 20, height = 20, units = "cm")

#my_genes <- row.names(subset(fData(cds), gene_short_name %in% genes.pseudotime))
my_genes <- c("CD14", "FCGR3A", "IL1B", "MS4A7", "CXCL10", "IGHM")
cds_subset <- cds[my_genes,]
plot_genes_in_pseudotime(cds_subset, color_cells_by = "monocle3_pseudotime" )

```


```{r}

# RBC_infected_cells <- row.names(subset(pData(cds),
#             Group == "RBC_infected"))
#RBC_infected <- cds[,RBC_infected_cells]
RBC_inf = cds[,rownames(immune.combined@meta.data[immune.combined@meta.data$Group == "RBC_infected",])]
plot_cells(RBC_inf, color_cells_by = "cluster", label_groups_by_cluster = F,
           label_branch_points = F, label_roots = F, label_leaves = F,
           group_label_size = 5,graph_label_size = 3)
imm_comb_RBC_inf <- subset(immune.combined, subset = Group == "RBC_infected")
FeaturePlot(imm_comb_RBC_inf, features = "pseudotime")

my_genes <- c("CD14", "FCGR3A", "IL1B", "MS4A7", "CXCL10", "IGHM")
RBC_inf_subset <- RBC_inf[my_genes,]
plot_genes_in_pseudotime(RBC_inf_subset, color_cells_by = "monocle3_pseudotime" )

```

```{r}


RBC_bys = cds[,rownames(immune.combined@meta.data[immune.combined@meta.data$Group == "RBC_bystander",])]
plot_cells(RBC_bys, color_cells_by = "cluster", label_groups_by_cluster = F,
           label_branch_points = F, label_roots = F, label_leaves = F,
           group_label_size = 5,graph_label_size = 3)
imm_comb_RBC_bys <- subset(immune.combined, subset = Group == "RBC_bystander")
FeaturePlot(imm_comb_RBC_bys, features = "pseudotime")

my_genes <- c("CD14", "FCGR3A", "IL1B", "MS4A7", "CXCL10", "IGHM")
RBC_bys_subset <- RBC_bys[my_genes,]
plot_genes_in_pseudotime(RBC_bys_subset, color_cells_by = "monocle3_pseudotime" )


```

```{r}


SPZ_bys = cds[,rownames(immune.combined@meta.data[immune.combined@meta.data$Group == "sporozoite_bystander",])]
plot_cells(SPZ_bys, color_cells_by = "cluster", label_groups_by_cluster = F,
           label_branch_points = F, label_roots = F, label_leaves = F,
           group_label_size = 5,graph_label_size = 3)
imm_comb_SPZ_bys <- subset(immune.combined, subset = Group == "sporozoite_bystander")
FeaturePlot(imm_comb_SPZ_bys, features = "pseudotime")

my_genes <- c("CD14", "FCGR3A", "IL1B", "MS4A7", "CXCL10", "IGHM")
SPZ_bys_subset <- SPZ_bys[my_genes,]
plot_genes_in_pseudotime(SPZ_bys_subset, color_cells_by = "monocle3_pseudotime" )


```
