---
title: "Single cell RNA-seq of Plasmodium sporozoite-infected monocytes - Normalisation"
output:
  pdf_document: default
  html_document:
    df_print: paged
  toc: TRUE
---

<!-- ```{r, echo = F} -->
<!-- # knitr::opts_chunk$set(warning = F, message = F) -->
<!-- # knitr::opts_chunk$set(fig.pos = "!H", out.extra = "") -->
<!-- #knitr::opts_chunk$set(tidy.opts=list(width.cutoff=80),tidy=TRUE) -->

<!-- ``` -->

<!-- ```{r load_libraries, echo = F, message = F, warning = F} -->
<!-- library(Seurat) -->
<!-- library(patchwork) -->
<!-- library(readxl) -->
<!-- library(tidyverse) -->
<!-- library(ggplot2) -->
<!-- library(ggpubr) -->
<!-- library(reshape2) -->
<!-- library(broom) -->
<!-- library(stringr) -->
<!-- remotes::install_github('chris-mcginnis-ucsf/DoubletFinder') -->
<!-- library(DoubletFinder) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- load("malaria_monocyte_demultiplex.RData") -->
<!-- ``` -->
# Step 3: Normalisation
```{r}
# After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method “LogNormalize” that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. Normalized values are stored in pbmc[["RNA"]]@data.

mono <- NormalizeData(mono.demul, normalization.method = "LogNormalize", scale.factor = 10000)

# For clarity, in this previous line of code (and in future commands), we provide the default values for certain parameters in the function call. However, this isn’t required and the same behavior can be achieved with:
```

```{r}
apply(mono@assays$RNA@data,1,mean) -> gene.expression
sort(gene.expression, decreasing = TRUE) -> gene.expression
head(gene.expression, n=50)
```

<!-- ```{r} -->
<!-- # We can already see that there may be some issues to address in this data. Malat1 is a nuclear expressed transcript which tends to persist when cells have lysed and the cytoplasm has gone. It is generally highly expressed anyway, but cells with a very high level might indicate a problem. -->
<!-- #  -->
<!-- # We can also see high amounts of ribosomal proteins. Again, these are generally highly expressed but their presence in specific subsets might also be of concern in regards to the accuracy of quantitation in the data. -->
<!-- #  -->
<!-- # We can look in various ways at how well the data have been normalised. We can pick out a specific gene: -->

<!-- ggplot(mapping = aes(mono@assays$RNA@data["GAPDH",])) +  -->
<!--   geom_histogram(binwidth = 1, fill="yellow", colour="black") +  -->
<!--   ggtitle("GAPDH expression") -->

<!-- # So even for a so-called housekeeping gene we still see a significant proportion of dropout cells, and expression values which spread over 3 orders of magnitude. -->
<!-- #  -->
<!-- # We can also go a bit wider and pick the first 100 cells and look at the distributions of their expression values. -->

<!-- as.tibble( -->
<!--   mono@assays$RNA@data[,1:100] -->
<!-- ) %>% -->
<!--   pivot_longer( -->
<!--     cols=everything(), -->
<!--     names_to="cell", -->
<!--     values_to="expression" -->
<!--   ) %>% -->
<!--   ggplot(aes(x=expression, group=cell)) + -->
<!--   geom_density() + -->
<!--   coord_cartesian(ylim=c(0,0.6), xlim=c(0,3)) -->
<!-- ``` -->

<!-- ```{r} -->

<!-- # So we can see that this simplistic normalisation doesn’t actually normalise the quantitative data very well because it’s so biased by the proportion of zero values in the dataset. This simplistic normalisation therefore doesn’t do a great job in this instance. -->
<!-- #  -->
<!-- # We can try the normalisation again, this time using a centered log ratio transformation - more similar to the sort of size factor based normalisation which is used for many RNA-Seq experiments. The margin=2 option means that it normalises per cell instead of per gene -->

<!-- NormalizeData(mono, normalization.method = "CLR", margin = 2) -> mono -->

<!-- as.tibble( -->
<!--   mono@assays$RNA@data[,1:100] -->
<!-- ) %>% -->
<!--   pivot_longer( -->
<!--     cols=everything(), -->
<!--     names_to="cell", -->
<!--     values_to="expression" -->
<!--   ) %>% -->
<!--   ggplot(aes(x=expression, group=cell)) + -->
<!--   geom_density() + -->
<!--   coord_cartesian(ylim=c(0,0.6), xlim=c(0,3)) -->

<!-- # This method clearly gives us much more well matched distributions, so from a quantitative point of view this is going to be easier to compare between samples, so we’ll stick to this. -->
<!-- #  -->
<!-- # We can also look at some overall metrics. Here we can compare the quantitative value at the 95th percentile to the mean expression -->

<!-- tibble( -->
<!--   pc95 = apply(mono[["RNA"]]@data,2,quantile,0.95), -->
<!--   measured = apply(mono[["RNA"]]@data,2,function(x)(100*sum(x!=0))/length(x)) -->
<!-- ) -> normalisation.qc -->

<!-- normalisation.qc %>%  -->
<!--   ggplot(aes(x=measured,y=pc95))+ -->
<!--   geom_point()+ -->
<!--   ggtitle("Normalisation of data") -->

<!-- # We can see that the CLR normalisation works pretty well with the 95th percentile being largely stable for cells with more than 5% measured genes (otherwise the value will be zero by definition). -->
<!-- ``` -->


<!-- ```{r cell_cycle} -->
<!-- #cc.genes.updated.2019 -->

<!-- #We can use these to try to predict the cell cycle of each cell. -->

<!-- CellCycleScoring(mono, s.features = cc.genes.updated.2019$s.genes, g2m.features = cc.genes.updated.2019$g2m.genes, set.ident = TRUE) -> mono -->

<!-- # We should now have a bunch of new QC metrics to give the SCORE for S and G2M -->
<!-- mono[[]] -->

<!-- #We can look at the spread of the cells in different states. -->

<!-- as_tibble(mono[[]]) %>% -->
<!--   ggplot(aes(Phase)) + geom_bar() -->

<!-- as_tibble(mono[[]]) %>% -->
<!--   ggplot(aes(x=S.Score, y=G2M.Score, color=Phase)) +  -->
<!--   geom_point() + -->
<!--   coord_cartesian(xlim=c(-0.15,0.15), ylim=c(-0.15,0.15)) + -->
<!--   theme_bw() -->

<!-- # Although the tool has made predictions of the stage for each cell, there isn’t a huge separation between the groups it’s picked so we have some hope that this will have a relatively minor influence on the overall expression patterns we see. We’ll pick this up later once we’ve clustered the data and we can see what the content of the different clusters looks like. -->
<!-- ``` -->


```{r}

# Identification of highly variable features (feature selection)
# 
# We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). We and others have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.
# 
# Our procedure in Seurat is described in detail here, and improves on previous versions by directly modeling the mean-variance relationship inherent in single-cell data, and is implemented in the FindVariableFeatures() function. By default, we return 2,000 features per dataset. These will be used in downstream analysis, like PCA.

mono <- FindVariableFeatures(mono, selection.method = "vst", nfeatures = 2000)

# as_tibble(HVFInfo(mono),rownames = "Gene") -> variance.data
# variance.data %>% 
#   mutate(hypervariable=Gene %in% VariableFeatures(mono)
# ) -> variance.data
# 
# head(variance.data, n=10)
# 
# # We can plot out a graph of the variance vs mean and highlight the selected genes so we can see whether we think we’re likely to capture what we need.

# variance.data %>% 
#   ggplot(aes(log(mean),log(variance),color=hypervariable)) + 
#   geom_point() + 
#   scale_color_manual(values=c("black","red")) +
#   theme_bw()

```

```{r}

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(mono), 10)
top10

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(mono)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)

plot2

```

```{r}
# Scale data
# Next, we apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The ScaleData() function:
# 
#     Shifts the expression of each gene, so that the mean expression across cells is 0
#     Scales the expression of each gene, so that the variance across cells is 1
#         This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
#     The results of this are stored in pbmc[["RNA"]]@scale.data

all.genes <- rownames(mono)
mono <- ScaleData(mono, features = all.genes)

```

```{r}
#save.image("malaria_monocyte_normalised.RData")

```